{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_elof_genes(row):\n",
    "    if (row['pLI'] > 0.9\n",
    "            and row['oe_lof_upper'] < 0.35\n",
    "            and row['pHaplo'] > 0.86):\n",
    "        return True\n",
    "    else:\n",
    "        return False \n",
    "\n",
    "def extract_g2pLoF_genes(x):\n",
    "    if 'absent gene product' in x:\n",
    "        return 'PASS'\n",
    "    else:\n",
    "        return '.'\n",
    "\n",
    "def extract_definitive(x):\n",
    "    if x == 'definitive':\n",
    "        return 'PASS'\n",
    "    else:\n",
    "        return '.'\n",
    "\n",
    "def extract_strong(x):\n",
    "    if (x == 'definitive' or x == 'strong'):\n",
    "        return 'PASS'\n",
    "    else:\n",
    "        return '.'\n",
    "\n",
    "def extract_moderate(x):\n",
    "    if (x == 'definitive' or x == 'strong' or x == 'moderate'):\n",
    "        return 'PASS'\n",
    "    else:\n",
    "        return '.'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Curation data\n",
    "ajhg_path = 'AJHG_ClinGenLoF_genes.txt.gz'\n",
    "hi_path = 'ClinGen_haploinsufficiency_gene_GRCh37.bed'\n",
    "gnomad_path = 'gnomad.v2.1.1.lof_metrics.by_gene.txt.gz'\n",
    "phaplo_path = 'Collins_rCNV_2022.dosage_sensitivity_scores.tsv.gz'\n",
    "\n",
    "# G2P data\n",
    "dd = './DDG2P.csv.gz'\n",
    "eye = './EyeG2P.csv.gz'\n",
    "skin = './SkinG2P.csv.gz'\n",
    "cancer = './CancerG2P.csv.gz'\n",
    "cardiac = './CardiacG2P.csv.gz'\n",
    "skeletal = './SkeletalG2P.csv.gz'\n",
    "\n",
    "# Validation set\n",
    "am_path = 'AlphaMissense_gene_hg19.tsv.gz'\n",
    "genovo_path = 'GenovoLOEUF.txt.gz'\n",
    "gnocchi_path = 'EnhancerGnocchi.txt.gz'\n",
    "\n",
    "# Load data\n",
    "ajhg = pd.read_csv(\n",
    "    ajhg_path, sep='\\t', header=None, skiprows=1,\n",
    "    names=['gene', 'pLI', 'RefSeq', 'SpliceAcceptor', 'SpliceDonor'],\n",
    "    usecols=['gene', 'RefSeq']) \n",
    "\n",
    "gnomad = pd.read_csv(\n",
    "    gnomad_path, sep='\\t', compression='gzip',\n",
    "    usecols=['gene', 'pLI', 'oe_lof_upper', 'transcript', 'cds_length']\n",
    "    )\n",
    "\n",
    "higene = pd.read_csv(\n",
    "    hi_path, sep='\\t', header=None, skiprows=1,\n",
    "    names=['CHROM', 'start', 'end', 'gene', 'HIscore'],\n",
    "    usecols=['gene', 'HIscore']\n",
    "    ) \n",
    "\n",
    "phaplo = pd.read_csv(\n",
    "    phaplo_path, sep='\\t', header=None, compression='gzip', skiprows=1,\n",
    "    names=['gene', 'pHaplo', 'pTriplo'],\n",
    "    usecols=['gene', 'pHaplo']\n",
    "    ) \n",
    "\n",
    "genovo = pd.read_csv(\n",
    "    genovo_path, sep='\\t', compression='gzip',\n",
    "    usecols=['enstID', 'enstVersion', 'genovo_LOEUF_corrected']\n",
    "    )\n",
    "\n",
    "gnocchi = pd.read_csv(\n",
    "    gnocchi_path, sep='\\t', compression='gzip',\n",
    "    usecols=['gene', 'enhancer_Gnocchi']\n",
    "    )\n",
    "\n",
    "am = pd.read_csv(am_path, sep='\\t', compression='gzip', skiprows=3)\n",
    "am['enstID'] = am['transcript_id'].str.split('.').str[0]\n",
    "am['version'] = am['transcript_id'].str.split('.').str[1]\n",
    "am.drop(columns=['transcript_id'], inplace=True)\n",
    "am = am[['enstID', 'version', 'mean_am_pathogenicity']]\n",
    "\n",
    "dddf = pd.read_csv(dd, sep=',', header=0, compression='infer')\n",
    "eyedf = pd.read_csv(eye, sep=',', header=0, compression='infer')\n",
    "skidf = pd.read_csv(skin, sep=',', header=0, compression='infer')\n",
    "candf = pd.read_csv(cancer, sep=',', header=0, compression='infer')\n",
    "cardf = pd.read_csv(cardiac, sep=',', header=0, compression='infer')\n",
    "skedf = pd.read_csv(skeletal, sep=',', header=0, compression='infer')\n",
    "dfs = [dddf, eyedf, skidf, candf, cardf, skedf]\n",
    "\n",
    "\n",
    "rename_cols = {\n",
    "    'gene symbol': 'gene',\n",
    "    'prev symbols': 'prev_symbols', \n",
    "    'hgnc id': 'hgnc_id',\n",
    "    'gene mim': 'gene_mim', \n",
    "    'mutation consequence': 'mutation_consequence',\n",
    "    'confidence category': 'confidence_category',\n",
    "    'allelic requirement': 'allelic_requirement', \n",
    "    'disease name': 'disease_name'\n",
    "    }\n",
    "\n",
    "pickup_cols = [\n",
    "    'gene', 'prev_symbols', 'hgnc_id',  \n",
    "    'mutation_consequence', 'confidence_category', \n",
    "    'allelic_requirement', 'panel', 'disease_name', \n",
    "    'g2pLoF', 'definitiveFILTER', 'strongFILTER','moderateFILTER'\n",
    "    ]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for df in dfs:\n",
    "    df['g2pLoF'] = df['mutation consequence'].apply(extract_g2pLoF_genes)\n",
    "\n",
    "for df in dfs:\n",
    "    df['definitiveFILTER'] = df['confidence category'].apply(extract_definitive)\n",
    "    df['strongFILTER'] = df['confidence category'].apply(extract_strong)\n",
    "    df['moderateFILTER'] = df['confidence category'].apply(extract_moderate)\n",
    "\n",
    "### Rename columns\n",
    "dddf = dddf.rename(columns=rename_cols)\n",
    "eyedf = eyedf.rename(columns=rename_cols)\n",
    "skidf = skidf.rename(columns=rename_cols)\n",
    "candf = candf.rename(columns=rename_cols)\n",
    "cardf = cardf.rename(columns=rename_cols)\n",
    "skedf = skedf.rename(columns=rename_cols)\n",
    "\n",
    "### Extract columns for analysis\n",
    "dddf = dddf[pickup_cols]\n",
    "eyedf = eyedf[pickup_cols]\n",
    "skidf = skidf[pickup_cols]\n",
    "candf = candf[pickup_cols]\n",
    "cardf = cardf[pickup_cols]\n",
    "skedf = skedf[pickup_cols]\n",
    "\n",
    "### Extract LoF genes\n",
    "lof_dd = dddf[dddf['g2pLoF'] == 'PASS']\n",
    "lof_eye = eyedf[eyedf['g2pLoF'] == 'PASS']\n",
    "lof_ski = skidf[skidf['g2pLoF'] == 'PASS']\n",
    "lof_can = candf[candf['g2pLoF'] == 'PASS']\n",
    "lof_car = cardf[cardf['g2pLoF'] == 'PASS']\n",
    "lof_ske = skedf[skedf['g2pLoF'] == 'PASS']\n",
    "\n",
    "\n",
    "### Concatnate dataframes\n",
    "alldf = pd.concat([lof_dd, lof_eye, lof_ski, lof_can, lof_car, lof_ske])\n",
    "alldf.fillna(value='.', inplace=True)\n",
    "\n",
    "### Extract genes with strong evidence\n",
    "alldf_filterd = alldf[alldf['strongFILTER'] == 'PASS']\n",
    "alldf_filterd = alldf_filterd.drop_duplicates(subset='gene')\n",
    "\n",
    "# Gene lists from 4 sources\n",
    "ajhg_genes = ajhg['gene']\n",
    "dcpr_genes = alldf_filterd['gene']\n",
    "hi_genes = hi['gene']\n",
    "phaplo_genes = phaplo['gene']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Curation set\n",
      "--- AJHG ---\n",
      "   gene       RefSeq\n",
      "0  AAAS  NM_015665.5\n",
      "1  AARS  NM_001605.2\n",
      "2  AASS  NM_005763.3\n",
      "1043\n",
      "\n",
      "--- gnomAD ---\n",
      "    gene       transcript  pLI  oe_lof_upper  cds_length\n",
      "0  MED13  ENST00000397786  1.0         0.030        6522\n",
      "1  NIPBL  ENST00000282516  1.0         0.032        8412\n",
      "2   SMC3  ENST00000361804  1.0         0.037        3651\n",
      "19658\n",
      "\n",
      "--- HI ---\n",
      "      gene  HIscore\n",
      "0     AGRN       30\n",
      "1  B3GALT6       30\n",
      "2     GNB1        1\n",
      "1500\n",
      "\n",
      "--- Phaplo ---\n",
      "      gene    pHaplo\n",
      "0  CACNA1C  0.998982\n",
      "1   ZNF462  1.000000\n",
      "2     CHD8  0.991650\n",
      "18641\n"
     ]
    }
   ],
   "source": [
    "# Curation set\n",
    "print('Curation set')\n",
    "print('--- AJHG ---')\n",
    "print(ajhg.head(3))\n",
    "print(ajhg['gene'].nunique())\n",
    "print('\\n--- gnomAD ---')\n",
    "print(gnomad.head(3))\n",
    "print(gnomad['gene'].nunique())\n",
    "print('\\n--- HI ---')\n",
    "print(higene.head(3))\n",
    "print(higene['gene'].nunique())\n",
    "print('\\n--- Phaplo ---')\n",
    "print(phaplo.head(3))\n",
    "print(phaplo['gene'].nunique())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validation set\n",
      "--- AlphaMissense ---\n",
      "            enstID version  mean_am_pathogenicity\n",
      "0  ENST00000000233       5               0.742270\n",
      "1  ENST00000000412       3               0.378343\n",
      "2  ENST00000001008       4               0.422290\n",
      "18761\n",
      "\n",
      "--- Gnocchi ---\n",
      "     gene  enhancer_Gnocchi\n",
      "0   NOC2L          1.509945\n",
      "1  SAMD11          1.579239\n",
      "2  RNF223          2.725941\n",
      "15541\n",
      "\n",
      "--- Genovo ---\n",
      "            enstID  enstVersion  genovo_LOEUF_corrected\n",
      "0  ENST00000394484            1                0.000000\n",
      "1  ENST00000329516            3                0.039973\n",
      "2  ENST00000330331            5                0.071441\n",
      "57308\n"
     ]
    }
   ],
   "source": [
    "print('Validation set')\n",
    "print('--- AlphaMissense ---')\n",
    "print(am.head(3))\n",
    "print(am['enstID'].nunique())\n",
    "print('\\n--- Gnocchi ---')\n",
    "print(gnocchi.head(3))\n",
    "print(gnocchi['gene'].nunique())\n",
    "print('\\n--- Genovo ---')\n",
    "print(genovo.head(3))\n",
    "print(genovo['enstID'].nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- AlphaMissense + Genovo ---\n",
      "            enstID version  mean_am_pathogenicity  enstVersion  \\\n",
      "0  ENST00000000233       5               0.742270            5   \n",
      "1  ENST00000000412       3               0.378343            3   \n",
      "2  ENST00000001008       4               0.422290            4   \n",
      "\n",
      "   genovo_LOEUF_corrected  \n",
      "0                0.109528  \n",
      "1                0.787049  \n",
      "2                0.524644  \n",
      "16355\n"
     ]
    }
   ],
   "source": [
    "# Merge AlphaMissense and Genovo data by enstID\n",
    "am_genovo = pd.merge(am, genovo, on='enstID', how='inner')\n",
    "print('\\n--- AlphaMissense + Genovo ---')\n",
    "print(am_genovo.head(3))\n",
    "print(am_genovo['enstID'].nunique())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_gnocchi = gnocchi[gnocchi['tissue'] == 'Merged']\n",
    "genelist = list(merged_gnocchi['gene'] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_genelist = list(gnocchi['gene'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15541\n",
      "15860\n",
      "['DCAF4L2', 'TAS2R20', 'ADAM8', 'SPTLC1', 'CALML5', 'USP17L17', 'C11orf94', 'CNOT3', 'NRG1', 'SPHAR', 'DDX11', 'C3orf56', 'ZNF382', 'SLC25A11', 'ZDHHC24', 'OR2M5', 'OR1N1', 'OR2T8', 'PAGR1', 'KRTAP12-1', 'SUMO4', 'CSF3', 'LCE6A', 'SPATA12', 'CCDC15', 'AGBL3', 'IFNA4', 'TAS2R42', 'LCE4A', 'HUS1B', 'CTAGE4', 'GPR84', 'ZNF497', 'COPS6', 'OR14A16', 'CEMP1', 'DIRAS3', 'TAS2R31', 'FAM170A', 'STH', 'TAS2R14', 'ZNF860', 'DNPH1', 'CCAR2', 'NR2E3', 'CTAGE8', 'IFNA8', 'FITM1', 'HIST2H3C', 'LENG9', 'OR7C1', 'UFC1', 'TIAF1', 'IFNB1', 'TYSND1', 'DVL2', 'KMT2B', 'CHRNB1', 'CARD6', 'OGDHL', 'HIST2H4B', 'OR14K1', 'IFNA21', 'TAS2R1', 'CCT6A', 'SYTL1', 'SRRM5', 'FPR1', 'HRCT1', 'MRGPRD', 'KIR3DX1', 'PTPN6', 'MANEAL', 'TAF1C', 'PTTG2', 'FAM218A', 'NDUFB10', 'TAS2R60', 'VN1R1', 'KRTAP4-3', 'CCDC177', 'KRTAP21-2', 'PM20D1', 'FAM71D', 'KRTDAP', 'PHGR1', 'MRGPRX1', 'PSAPL1', 'TRIM73', 'LRRN4CL', 'MRGPRX2', 'FPR3', 'ANAPC15', 'BLID', 'FRMD1', 'LCN6', 'NDUFA7', 'NPIPB5', 'PARP2', 'CFC1B', 'ZNF219', 'CDX1', 'GIMAP7', 'AL445989.1', 'H1FNT', 'ANXA8', 'KRTAP9-1', 'HIST2H2AA4', 'SPPL2B', 'PRKACG', 'FUT7', 'SIRT2', 'IL23A', 'OR1M1', 'BNIPL', 'CTAGE9', 'OR7A5', 'RNASE11', 'DAGLB', 'FSCB', 'PLEKHN1', 'OR8D1', 'ARIH2OS', 'SPRR2B', 'DYNLL1', 'TPSAB1', 'ANKRD36C', 'ZNF491', 'IFNA6', 'TAS2R8', 'KRTAP12-3', 'TAS2R9', 'IRGM', 'RNF6', 'RFPL4B', 'FAM25G', 'MRGPRX3', 'IVL', 'IFNA13', 'HSH2D', 'CDC42EP5', 'PLA2G3', 'OR2T12', 'IFNA10', 'SPRR2A', 'KRTAP23-1', 'TAOK2', 'VPS35', 'PPIL4', 'HIST2H4A', 'RNASE3', 'OR2T34', 'PPP1R26', 'OR14I1', 'RNASE2', 'MRGPRG', 'CRB3', 'LCE3A', 'KMT2D', 'TRIM60', 'IFNA2', 'RAD51AP1', 'SMCP', 'OR7D2', 'TAS2R50', 'ZNF835', 'WDR5', 'AGAP4', 'HEPN1', 'MISP', 'ERVW-1', 'RAB3D', 'IFRD2', 'HIST2H2AA3', 'IFNA14', 'GPR32', 'SEMA4A', 'NAT8', 'VN1R2', 'NPDC1', 'CSMD2', 'ANXA2R', 'MAS1L', 'C6orf120', 'ZNF781', 'FTL', 'FBXW9', 'PYDC1', 'KRTAP19-1', 'DERL2', 'TAS2R30', 'FES', 'CYP21A2', 'CD177', 'C6orf141', 'USP17L30', 'IFNA1', 'ZNF492', 'HIST2H3A', 'OR7G2', 'HSPB6', 'DIRC1', 'SCYL1', 'PBOV1', 'ZNF600', 'FUT6', 'ZNF26', 'C21orf58', 'PRRT2', 'TAS2R46', 'TTC25', 'OR51C1P', 'MTRNR2L4', 'TEN1', 'ARHGEF35', 'EID2B', 'ADIG', 'PLD4', 'APOC4-APOC2', 'ZNF83', 'CMTM1', 'SH3BGRL3', 'PLK1', 'TRAF7', 'ZNF80', 'NANS', 'KXD1', 'MPI', 'GPR137C', 'ZFP42', 'SULT2A1', 'ALDOC', 'OR14C36', 'OR2M2', 'LMTK3', 'IFNA5', 'GSTT1', 'SPRR3', 'PRR23A', 'OR2L5', 'SLC35G5', 'IFNW1', 'MRGPRX4', 'MAP3K14', 'PRR20E', 'ZNF778', 'SMIM6', 'PSKH1', 'TSPYL6', 'FAM220A', 'PYDC2', 'ERV3-1', 'TAS2R19', 'USP17L19', 'DHH', 'TAS2R43', 'C19orf48', 'OR2L2', 'SNURF', 'IFNA7', 'HIST1H1T', 'C5orf22', 'SLC9A3R2', 'OR2L3', 'OR5A2', 'RAB6C', 'C14orf132', 'DYNC2H1', 'RIMBP3C', 'ZNF843', 'USP29', 'PRR20D', 'KRTAP12-2', 'EDDM3B', 'ZNF699', 'KRTAP15-1', 'TEX19', 'DNLZ', 'USP17L25', 'EBLN2', 'ZNF223', 'EDDM3A', 'DOK3', 'OR14A2', 'OR2AK2', 'OR2T3', 'SPRR2G', 'SLC35G3', 'FAM217B', 'ADAM30', 'MTRNR2L8', 'KBTBD6', 'IGFLR1', 'PCED1B', 'YPEL4', 'H3F3C', 'NHLRC4', 'RNASE8', 'USP17L27', 'CCDC140', 'CTAGE1', 'OR2M3', 'UHRF1', 'CYP4F8', 'KRTAP19-8', 'PPIAL4A', 'C8orf49', 'GABRG1', 'ACTR1A', 'TMA7', 'HSPB9', 'ADAM29', 'MTRNR2L3', 'FBXO2', 'TICAM1', 'TAS2R10', 'ZNF888', 'PRH1', 'PRSS8']\n"
     ]
    }
   ],
   "source": [
    "# comparison of gene lists (genelist vs. raw_genelist)\n",
    "print(len(genelist))\n",
    "print(len(raw_genelist))\n",
    "\n",
    "# substraction of gene lists\n",
    "diff = list(set(raw_genelist) - set(genelist))\n",
    "print(diff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>gene</th>\n",
       "      <th>ENSG</th>\n",
       "      <th>z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2126</th>\n",
       "      <td>ESC</td>\n",
       "      <td>ASCL2</td>\n",
       "      <td>ENSG00000183734</td>\n",
       "      <td>5.063194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31916</th>\n",
       "      <td>iPSC</td>\n",
       "      <td>ASCL2</td>\n",
       "      <td>ENSG00000183734</td>\n",
       "      <td>5.063194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47167</th>\n",
       "      <td>Blood_T-cell</td>\n",
       "      <td>ASCL2</td>\n",
       "      <td>ENSG00000183734</td>\n",
       "      <td>6.084149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47168</th>\n",
       "      <td>Digestive</td>\n",
       "      <td>ASCL2</td>\n",
       "      <td>ENSG00000183734</td>\n",
       "      <td>6.084149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67741</th>\n",
       "      <td>HSC_B-cell</td>\n",
       "      <td>ASCL2</td>\n",
       "      <td>ENSG00000183734</td>\n",
       "      <td>6.084149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108892</th>\n",
       "      <td>Epithelial</td>\n",
       "      <td>ASCL2</td>\n",
       "      <td>ENSG00000183734</td>\n",
       "      <td>5.521794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121319</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>ASCL2</td>\n",
       "      <td>ENSG00000183734</td>\n",
       "      <td>6.084149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182268</th>\n",
       "      <td>PLACENTA</td>\n",
       "      <td>ASCL2</td>\n",
       "      <td>ENSG00000183734</td>\n",
       "      <td>5.521794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192625</th>\n",
       "      <td>PANCREAS</td>\n",
       "      <td>ASCL2</td>\n",
       "      <td>ENSG00000183734</td>\n",
       "      <td>5.569333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199282</th>\n",
       "      <td>LUNG</td>\n",
       "      <td>ASCL2</td>\n",
       "      <td>ENSG00000183734</td>\n",
       "      <td>6.084149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203777</th>\n",
       "      <td>LIVER</td>\n",
       "      <td>ASCL2</td>\n",
       "      <td>ENSG00000183734</td>\n",
       "      <td>4.492174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208641</th>\n",
       "      <td>SPLEEN</td>\n",
       "      <td>ASCL2</td>\n",
       "      <td>ENSG00000183734</td>\n",
       "      <td>6.084149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213417</th>\n",
       "      <td>Merged</td>\n",
       "      <td>ASCL2</td>\n",
       "      <td>ENSG00000183734</td>\n",
       "      <td>5.578892</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              tissue   gene             ENSG         z\n",
       "2126             ESC  ASCL2  ENSG00000183734  5.063194\n",
       "31916           iPSC  ASCL2  ENSG00000183734  5.063194\n",
       "47167   Blood_T-cell  ASCL2  ENSG00000183734  6.084149\n",
       "47168      Digestive  ASCL2  ENSG00000183734  6.084149\n",
       "67741     HSC_B-cell  ASCL2  ENSG00000183734  6.084149\n",
       "108892    Epithelial  ASCL2  ENSG00000183734  5.521794\n",
       "121319        Thymus  ASCL2  ENSG00000183734  6.084149\n",
       "182268      PLACENTA  ASCL2  ENSG00000183734  5.521794\n",
       "192625      PANCREAS  ASCL2  ENSG00000183734  5.569333\n",
       "199282          LUNG  ASCL2  ENSG00000183734  6.084149\n",
       "203777         LIVER  ASCL2  ENSG00000183734  4.492174\n",
       "208641        SPLEEN  ASCL2  ENSG00000183734  6.084149\n",
       "213417        Merged  ASCL2  ENSG00000183734  5.578892"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ascl2 = gnocchi[gnocchi['gene'] == 'ASCL2']\n",
    "ascl2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5.64469795825\n"
     ]
    }
   ],
   "source": [
    "ascl2_without_merged = ascl2[ascl2['tissue'] != 'Merged']\n",
    "mean_ascl2 = ascl2_without_merged['z'].mean()\n",
    "\n",
    "print(mean_ascl2)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "wesanno",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
