{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Import package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import pandas as pd\n",
    "from Bio.Seq import Seq\n",
    "from liftover import get_lifter\n",
    "from pathlib2 import Path\n",
    "from pandarallel import pandarallel\n",
    "from tqdm import tqdm\n",
    "import gffutils\n",
    "import pysam\n",
    "from cyvcf2 import VCF\n",
    "\n",
    "### Logging setup\n",
    "from logging import getLogger, config\n",
    "import yaml\n",
    "parent_directory = os.path.dirname(os.path.dirname('__file__'))\n",
    "config_path: str = os.path.join(parent_directory, '../../../config/logging.yaml')\n",
    "with open(config_path, 'r') as f:\n",
    "    config.dictConfig(yaml.safe_load(f))\n",
    "logger = getLogger(__name__)\n",
    "\n",
    "########   Initialize and setup pandas methods   ########\n",
    "os.environ['JOBLIB_TEMP_FOLDER'] = '/tmp' \n",
    "pandarallel.initialize(nb_workers=3, progress_bar=False, verbose=0, use_memory_fs=False) \n",
    "tqdm.pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Allele': 0, 'Consequence': 1, 'IMPACT': 2, 'SYMBOL': 3, 'Gene': 4, 'Feature_type': 5, 'Feature': 6, 'BIOTYPE': 7, 'EXON': 8, 'INTRON': 9, 'HGVSc': 10, 'HGVSp': 11, 'cDNA_position': 12, 'CDS_position': 13, 'Protein_position': 14, 'Amino_acids': 15, 'Codons': 16, 'Existing_variation': 17, 'DISTANCE': 18, 'STRAND': 19, 'FLAGS': 20, 'VARIANT_CLASS': 21, 'SYMBOL_SOURCE': 22, 'HGNC_ID': 23, 'CANONICAL': 24, 'REFSEQ_MATCH': 25, 'SOURCE': 26, 'REFSEQ_OFFSET': 27, 'HGVS_OFFSET': 28}\n"
     ]
    }
   ],
   "source": [
    "## Parse VCF to simple input table\n",
    "\n",
    "raw_vcf = 'splai_vep_vcfs/hgmd_dm/all_DM_chr1.splai.vep.vcf'\n",
    "\n",
    "vcf = VCF(raw_vcf)\n",
    "header = vcf.header_iter()\n",
    "for h in header:\n",
    "    try:\n",
    "        h['ID']\n",
    "    except KeyError:\n",
    "        continue\n",
    "    else:\n",
    "        if h['ID'] == 'CSQ':\n",
    "            vep_cols_list = h['Description'].split('Format: ')[1].rstrip('\"').split('|')\n",
    "        elif h['ID'] == 'SpliceAI':\n",
    "            splai_cols_list = h['Description'].split('Format: ')[1].rstrip('\"').split('|')\n",
    "        else:\n",
    "            pass\n",
    "\n",
    "vepidx: dict = {col: i for i, col in enumerate(vep_cols_list)}\n",
    "splaidx: dict = {col: i for i, col in enumerate(splai_cols_list)}\n",
    "\n",
    "cols = [\n",
    "    'CHROM', 'POS', 'REF', 'ALT', 'GeneSymbol', 'SymbolSource', 'HGNC_ID', \n",
    "    'ENST', 'HGVSc', 'Consequence', 'EXON', 'INTRON', 'Strand',\n",
    "    'DS_AG', 'DS_AL', 'DS_DG', 'DS_DL', 'DP_AG', 'DP_AL', 'DP_DG', 'DP_DL', 'MaxSpliceAI'\n",
    "]\n",
    "\n",
    "print(vepidx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024/10/16 08:55:23 [WARNING] (__main__) - [1:1855272] HGVSc were not found\n",
      "2024/10/16 08:55:23 [WARNING] (__main__) - [1:1858463] HGVSc were not found\n",
      "2024/10/16 08:55:23 [WARNING] (__main__) - [1:1902161] HGVSc were not found\n",
      "2024/10/16 08:55:23 [WARNING] (__main__) - [1:1905486] HGVSc were not found\n",
      "2024/10/16 08:55:23 [WARNING] (__main__) - [1:5932827] HGVSc were not found\n",
      "2024/10/16 08:55:23 [WARNING] (__main__) - [1:5932827] SpliceAI scores were not found\n",
      "2024/10/16 08:55:23 [WARNING] (__main__) - [1:5951267] HGVSc were not found\n",
      "2024/10/16 08:55:23 [WARNING] (__main__) - [1:5951267] SpliceAI scores were not found\n",
      "2024/10/16 08:55:23 [WARNING] (__main__) - [1:5951835] HGVSc were not found\n",
      "2024/10/16 08:55:23 [WARNING] (__main__) - [1:5951835] SpliceAI scores were not found\n",
      "2024/10/16 08:55:23 [WARNING] (__main__) - [1:6556592] SpliceAI scores were not found\n",
      "2024/10/16 08:55:23 [WARNING] (__main__) - [1:6556613] SpliceAI scores were not found\n",
      "2024/10/16 08:55:23 [WARNING] (__main__) - [1:8018776] HGVSc were not found\n",
      "2024/10/16 08:55:23 [WARNING] (__main__) - [1:8018776] SpliceAI scores were not found\n",
      "2024/10/16 08:55:24 [WARNING] (__main__) - [1:12007910] HGVSc were not found\n",
      "2024/10/16 08:55:24 [WARNING] (__main__) - [1:12007910] SpliceAI scores were not found\n",
      "2024/10/16 08:55:24 [WARNING] (__main__) - [1:12067290] HGVSc were not found\n",
      "2024/10/16 08:55:24 [WARNING] (__main__) - [1:12067290] SpliceAI scores were not found\n",
      "2024/10/16 08:55:24 [WARNING] (__main__) - [1:16450820] HGVSc were not found\n",
      "2024/10/16 08:55:24 [WARNING] (__main__) - [1:16450820] SpliceAI scores were not found\n",
      "2024/10/16 08:55:25 [WARNING] (__main__) - [1:17356164] HGVSc were not found\n",
      "2024/10/16 08:55:25 [WARNING] (__main__) - [1:17356164] SpliceAI scores were not found\n",
      "2024/10/16 08:55:25 [WARNING] (__main__) - [1:17375248] HGVSc were not found\n",
      "2024/10/16 08:55:25 [WARNING] (__main__) - [1:17375248] SpliceAI scores were not found\n",
      "2024/10/16 08:55:25 [WARNING] (__main__) - [1:17378585] HGVSc were not found\n",
      "2024/10/16 08:55:25 [WARNING] (__main__) - [1:17378585] SpliceAI scores were not found\n",
      "2024/10/16 08:55:26 [WARNING] (__main__) - [1:20973462] HGVSc were not found\n",
      "2024/10/16 08:55:26 [WARNING] (__main__) - [1:20973462] SpliceAI scores were not found\n",
      "2024/10/16 08:55:27 [WARNING] (__main__) - [1:24143142] HGVSc were not found\n",
      "2024/10/16 08:55:27 [WARNING] (__main__) - [1:24143142] SpliceAI scores were not found\n",
      "2024/10/16 08:55:27 [WARNING] (__main__) - [1:24186555] HGVSc were not found\n",
      "2024/10/16 08:55:27 [WARNING] (__main__) - [1:24186555] SpliceAI scores were not found\n",
      "2024/10/16 08:55:28 [WARNING] (__main__) - [1:26126660] HGVSc were not found\n",
      "2024/10/16 08:55:28 [WARNING] (__main__) - [1:26126660] SpliceAI scores were not found\n",
      "2024/10/16 08:55:28 [WARNING] (__main__) - [1:27227086] HGVSc were not found\n",
      "2024/10/16 08:55:28 [WARNING] (__main__) - [1:27227086] SpliceAI scores were not found\n",
      "2024/10/16 08:55:28 [WARNING] (__main__) - [1:27227151] HGVSc were not found\n",
      "2024/10/16 08:55:28 [WARNING] (__main__) - [1:27227151] SpliceAI scores were not found\n",
      "2024/10/16 08:55:28 [WARNING] (__main__) - [1:27589831] HGVSc were not found\n",
      "2024/10/16 08:55:28 [WARNING] (__main__) - [1:27589831] SpliceAI scores were not found\n",
      "2024/10/16 08:55:29 [WARNING] (__main__) - [1:29557457] HGVSc were not found\n",
      "2024/10/16 08:55:29 [WARNING] (__main__) - [1:29557457] SpliceAI scores were not found\n",
      "2024/10/16 08:55:29 [WARNING] (__main__) - [1:33473828] HGVSc were not found\n",
      "2024/10/16 08:55:29 [WARNING] (__main__) - [1:33473828] SpliceAI scores were not found\n",
      "2024/10/16 08:55:29 [WARNING] (__main__) - [1:36937173] HGVSc were not found\n",
      "2024/10/16 08:55:29 [WARNING] (__main__) - [1:36937173] SpliceAI scores were not found\n",
      "2024/10/16 08:55:29 [WARNING] (__main__) - [1:38093282] HGVSc were not found\n",
      "2024/10/16 08:55:29 [WARNING] (__main__) - [1:38093282] SpliceAI scores were not found\n",
      "2024/10/16 08:55:29 [WARNING] (__main__) - [1:40732007] HGVSc were not found\n",
      "2024/10/16 08:55:29 [WARNING] (__main__) - [1:40732007] SpliceAI scores were not found\n",
      "2024/10/16 08:55:30 [WARNING] (__main__) - [1:43218171] HGVSc were not found\n",
      "2024/10/16 08:55:30 [WARNING] (__main__) - [1:43218171] SpliceAI scores were not found\n",
      "2024/10/16 08:55:30 [WARNING] (__main__) - [1:43251878] HGVSc were not found\n",
      "2024/10/16 08:55:30 [WARNING] (__main__) - [1:43251878] SpliceAI scores were not found\n",
      "2024/10/16 08:55:33 [WARNING] (__main__) - [1:46660223] HGVSc were not found\n",
      "2024/10/16 08:55:33 [WARNING] (__main__) - [1:46660223] SpliceAI scores were not found\n",
      "2024/10/16 08:55:33 [WARNING] (__main__) - [1:46664205] SpliceAI scores were not found\n",
      "2024/10/16 08:55:34 [WARNING] (__main__) - [1:53673062] HGVSc were not found\n",
      "2024/10/16 08:55:34 [WARNING] (__main__) - [1:53673062] SpliceAI scores were not found\n",
      "2024/10/16 08:55:34 [WARNING] (__main__) - [1:55505180] HGVSc were not found\n",
      "2024/10/16 08:55:34 [WARNING] (__main__) - [1:55505180] SpliceAI scores were not found\n",
      "2024/10/16 08:55:35 [WARNING] (__main__) - [1:66087171] HGVSc were not found\n",
      "2024/10/16 08:55:35 [WARNING] (__main__) - [1:66087171] SpliceAI scores were not found\n",
      "2024/10/16 08:55:35 [WARNING] (__main__) - [1:67509776] HGVSc were not found\n",
      "2024/10/16 08:55:35 [WARNING] (__main__) - [1:67509776] SpliceAI scores were not found\n",
      "2024/10/16 08:55:36 [WARNING] (__main__) - [1:76192737] HGVSc were not found\n",
      "2024/10/16 08:55:36 [WARNING] (__main__) - [1:76192737] SpliceAI scores were not found\n",
      "2024/10/16 08:55:37 [WARNING] (__main__) - [1:78354257] HGVSc were not found\n",
      "2024/10/16 08:55:37 [WARNING] (__main__) - [1:78354257] SpliceAI scores were not found\n",
      "2024/10/16 08:55:38 [WARNING] (__main__) - [1:91487748] HGVSc were not found\n",
      "2024/10/16 08:55:38 [WARNING] (__main__) - [1:91487748] SpliceAI scores were not found\n",
      "2024/10/16 08:55:39 [WARNING] (__main__) - [1:94402741] HGVSc were not found\n",
      "2024/10/16 08:55:39 [WARNING] (__main__) - [1:94402741] SpliceAI scores were not found\n",
      "2024/10/16 08:55:42 [WARNING] (__main__) - [1:94507676] HGVSc were not found\n",
      "2024/10/16 08:55:42 [WARNING] (__main__) - [1:94507676] SpliceAI scores were not found\n",
      "2024/10/16 08:55:43 [WARNING] (__main__) - [1:94514388] HGVSc were not found\n",
      "2024/10/16 08:55:43 [WARNING] (__main__) - [1:94514388] SpliceAI scores were not found\n",
      "2024/10/16 08:55:44 [WARNING] (__main__) - [1:94534443] HGVSc were not found\n",
      "2024/10/16 08:55:44 [WARNING] (__main__) - [1:94534443] SpliceAI scores were not found\n",
      "2024/10/16 08:55:45 [WARNING] (__main__) - [1:94564008] HGVSc were not found\n",
      "2024/10/16 08:55:45 [WARNING] (__main__) - [1:94564008] SpliceAI scores were not found\n",
      "2024/10/16 08:55:45 [WARNING] (__main__) - [1:94564078] HGVSc were not found\n",
      "2024/10/16 08:55:45 [WARNING] (__main__) - [1:94564078] SpliceAI scores were not found\n",
      "2024/10/16 08:55:45 [WARNING] (__main__) - [1:94568029] HGVSc were not found\n",
      "2024/10/16 08:55:45 [WARNING] (__main__) - [1:94568029] SpliceAI scores were not found\n",
      "2024/10/16 08:55:45 [WARNING] (__main__) - [1:94577489] HGVSc were not found\n",
      "2024/10/16 08:55:45 [WARNING] (__main__) - [1:94577489] SpliceAI scores were not found\n",
      "2024/10/16 08:55:46 [WARNING] (__main__) - [1:98515539] HGVSc were not found\n",
      "2024/10/16 08:55:46 [WARNING] (__main__) - [1:98515539] SpliceAI scores were not found\n",
      "2024/10/16 08:55:47 [WARNING] (__main__) - [1:100366307] HGVSc were not found\n",
      "2024/10/16 08:55:47 [WARNING] (__main__) - [1:100366307] SpliceAI scores were not found\n",
      "2024/10/16 08:55:48 [WARNING] (__main__) - [1:100680537] HGVSc were not found\n",
      "2024/10/16 08:55:48 [WARNING] (__main__) - [1:100680537] SpliceAI scores were not found\n",
      "2024/10/16 08:55:48 [WARNING] (__main__) - [1:100706311] HGVSc were not found\n",
      "2024/10/16 08:55:48 [WARNING] (__main__) - [1:100706311] SpliceAI scores were not found\n",
      "2024/10/16 08:55:48 [WARNING] (__main__) - [1:103471297] HGVSc were not found\n",
      "2024/10/16 08:55:48 [WARNING] (__main__) - [1:103471297] SpliceAI scores were not found\n",
      "2024/10/16 08:55:50 [WARNING] (__main__) - [1:113498814] HGVSc were not found\n",
      "2024/10/16 08:55:50 [WARNING] (__main__) - [1:115571593] HGVSc were not found\n",
      "2024/10/16 08:55:50 [WARNING] (__main__) - [1:115571593] SpliceAI scores were not found\n",
      "2024/10/16 08:55:51 [WARNING] (__main__) - [1:116243031] HGVSc were not found\n",
      "2024/10/16 08:55:51 [WARNING] (__main__) - [1:116243031] SpliceAI scores were not found\n",
      "2024/10/16 08:55:51 [WARNING] (__main__) - [1:119591856] HGVSc were not found\n",
      "2024/10/16 08:55:51 [WARNING] (__main__) - [1:119591856] SpliceAI scores were not found\n",
      "2024/10/16 08:55:54 [WARNING] (__main__) - [1:145516400] HGVSc were not found\n",
      "2024/10/16 08:55:54 [WARNING] (__main__) - [1:145516400] SpliceAI scores were not found\n",
      "2024/10/16 08:55:55 [WARNING] (__main__) - [1:150485081] HGVSc were not found\n",
      "2024/10/16 08:55:55 [WARNING] (__main__) - [1:150485081] SpliceAI scores were not found\n",
      "2024/10/16 08:55:56 [WARNING] (__main__) - [1:151259492] HGVSc were not found\n",
      "2024/10/16 08:55:56 [WARNING] (__main__) - [1:151260761] HGVSc were not found\n",
      "2024/10/16 08:55:56 [WARNING] (__main__) - [1:151261626] HGVSc were not found\n",
      "2024/10/16 08:55:56 [WARNING] (__main__) - [1:151262329] HGVSc were not found\n",
      "2024/10/16 08:55:58 [WARNING] (__main__) - [1:153784762] HGVSc were not found\n",
      "2024/10/16 08:55:58 [WARNING] (__main__) - [1:153784762] SpliceAI scores were not found\n",
      "2024/10/16 08:55:58 [WARNING] (__main__) - [1:153788316] HGVSc were not found\n",
      "2024/10/16 08:55:58 [WARNING] (__main__) - [1:153788316] SpliceAI scores were not found\n",
      "2024/10/16 08:55:59 [WARNING] (__main__) - [1:154245500] HGVSc were not found\n",
      "2024/10/16 08:55:59 [WARNING] (__main__) - [1:154245500] SpliceAI scores were not found\n",
      "2024/10/16 08:56:05 [WARNING] (__main__) - [1:155210069] HGVSc were not found\n",
      "2024/10/16 08:56:05 [WARNING] (__main__) - [1:155210069] SpliceAI scores were not found\n",
      "2024/10/16 08:56:05 [WARNING] (__main__) - [1:155211053] SpliceAI scores were not found\n",
      "2024/10/16 08:56:06 [WARNING] (__main__) - [1:155260582] HGVSc were not found\n",
      "2024/10/16 08:56:06 [WARNING] (__main__) - [1:155260582] SpliceAI scores were not found\n",
      "2024/10/16 08:56:06 [WARNING] (__main__) - [1:155261106] HGVSc were not found\n",
      "2024/10/16 08:56:06 [WARNING] (__main__) - [1:155261106] SpliceAI scores were not found\n",
      "2024/10/16 08:56:06 [WARNING] (__main__) - [1:155262969] HGVSc were not found\n",
      "2024/10/16 08:56:06 [WARNING] (__main__) - [1:155262969] SpliceAI scores were not found\n",
      "2024/10/16 08:56:08 [WARNING] (__main__) - [1:155271256] HGVSc were not found\n",
      "2024/10/16 08:56:08 [WARNING] (__main__) - [1:155271256] SpliceAI scores were not found\n",
      "2024/10/16 08:56:08 [WARNING] (__main__) - [1:155271258] HGVSc were not found\n",
      "2024/10/16 08:56:08 [WARNING] (__main__) - [1:155271258] SpliceAI scores were not found\n",
      "2024/10/16 08:56:08 [WARNING] (__main__) - [1:155271259] HGVSc were not found\n",
      "2024/10/16 08:56:08 [WARNING] (__main__) - [1:155271259] SpliceAI scores were not found\n",
      "2024/10/16 08:56:08 [WARNING] (__main__) - [1:155271269] HGVSc were not found\n",
      "2024/10/16 08:56:08 [WARNING] (__main__) - [1:155271269] SpliceAI scores were not found\n",
      "2024/10/16 08:56:08 [WARNING] (__main__) - [1:155271269] HGVSc were not found\n",
      "2024/10/16 08:56:08 [WARNING] (__main__) - [1:155271269] SpliceAI scores were not found\n",
      "2024/10/16 08:56:09 [WARNING] (__main__) - [1:155652848] HGVSc were not found\n",
      "2024/10/16 08:56:09 [WARNING] (__main__) - [1:155652848] SpliceAI scores were not found\n",
      "2024/10/16 08:56:09 [WARNING] (__main__) - [1:156052973] HGVSc were not found\n",
      "2024/10/16 08:56:09 [WARNING] (__main__) - [1:156052973] SpliceAI scores were not found\n",
      "2024/10/16 08:56:15 [WARNING] (__main__) - [1:156837500] HGVSc were not found\n",
      "2024/10/16 08:56:15 [WARNING] (__main__) - [1:156837500] SpliceAI scores were not found\n",
      "2024/10/16 08:56:15 [WARNING] (__main__) - [1:156844809] HGVSc were not found\n",
      "2024/10/16 08:56:15 [WARNING] (__main__) - [1:156844809] SpliceAI scores were not found\n",
      "2024/10/16 08:56:16 [WARNING] (__main__) - [1:158537344] HGVSc were not found\n",
      "2024/10/16 08:56:16 [WARNING] (__main__) - [1:158537344] SpliceAI scores were not found\n",
      "2024/10/16 08:56:16 [WARNING] (__main__) - [1:158639821] HGVSc were not found\n",
      "2024/10/16 08:56:16 [WARNING] (__main__) - [1:158639821] SpliceAI scores were not found\n",
      "2024/10/16 08:56:17 [WARNING] (__main__) - [1:160001799] HGVSc were not found\n",
      "2024/10/16 08:56:17 [WARNING] (__main__) - [1:160001799] SpliceAI scores were not found\n",
      "2024/10/16 08:56:19 [WARNING] (__main__) - [1:161135755] HGVSc were not found\n",
      "2024/10/16 08:56:19 [WARNING] (__main__) - [1:161135755] SpliceAI scores were not found\n",
      "2024/10/16 08:56:19 [WARNING] (__main__) - [1:161137433] HGVSc were not found\n",
      "2024/10/16 08:56:19 [WARNING] (__main__) - [1:161137433] SpliceAI scores were not found\n",
      "2024/10/16 08:56:19 [WARNING] (__main__) - [1:161137575] HGVSc were not found\n",
      "2024/10/16 08:56:19 [WARNING] (__main__) - [1:161137575] SpliceAI scores were not found\n",
      "2024/10/16 08:56:19 [WARNING] (__main__) - [1:161137905] HGVSc were not found\n",
      "2024/10/16 08:56:19 [WARNING] (__main__) - [1:161137905] SpliceAI scores were not found\n",
      "2024/10/16 08:56:25 [WARNING] (__main__) - [1:161326271] HGVSc were not found\n",
      "2024/10/16 08:56:25 [WARNING] (__main__) - [1:161326271] SpliceAI scores were not found\n",
      "2024/10/16 08:56:25 [WARNING] (__main__) - [1:164037507] HGVSc were not found\n",
      "2024/10/16 08:56:25 [WARNING] (__main__) - [1:164037507] SpliceAI scores were not found\n",
      "2024/10/16 08:56:26 [WARNING] (__main__) - [1:169433702] HGVSc were not found\n",
      "2024/10/16 08:56:26 [WARNING] (__main__) - [1:169433702] SpliceAI scores were not found\n",
      "2024/10/16 08:56:28 [WARNING] (__main__) - [1:169511083] HGVSc were not found\n",
      "2024/10/16 08:56:28 [WARNING] (__main__) - [1:169511083] SpliceAI scores were not found\n",
      "2024/10/16 08:56:28 [WARNING] (__main__) - [1:169522729] HGVSc were not found\n",
      "2024/10/16 08:56:28 [WARNING] (__main__) - [1:169522729] SpliceAI scores were not found\n",
      "2024/10/16 08:56:31 [WARNING] (__main__) - [1:172628081] HGVSc were not found\n",
      "2024/10/16 08:56:31 [WARNING] (__main__) - [1:172628081] SpliceAI scores were not found\n",
      "2024/10/16 08:56:32 [WARNING] (__main__) - [1:173873549] HGVSc were not found\n",
      "2024/10/16 08:56:32 [WARNING] (__main__) - [1:173873549] SpliceAI scores were not found\n",
      "2024/10/16 08:56:32 [WARNING] (__main__) - [1:173874581] HGVSc were not found\n",
      "2024/10/16 08:56:32 [WARNING] (__main__) - [1:173874581] SpliceAI scores were not found\n",
      "2024/10/16 08:56:32 [WARNING] (__main__) - [1:173874714] HGVSc were not found\n",
      "2024/10/16 08:56:32 [WARNING] (__main__) - [1:173874714] SpliceAI scores were not found\n",
      "2024/10/16 08:56:34 [WARNING] (__main__) - [1:173881627] HGVSc were not found\n",
      "2024/10/16 08:56:34 [WARNING] (__main__) - [1:173881627] SpliceAI scores were not found\n",
      "2024/10/16 08:56:36 [WARNING] (__main__) - [1:173886568] HGVSc were not found\n",
      "2024/10/16 08:56:36 [WARNING] (__main__) - [1:173886568] SpliceAI scores were not found\n",
      "2024/10/16 08:56:38 [WARNING] (__main__) - [1:179545163] HGVSc were not found\n",
      "2024/10/16 08:56:38 [WARNING] (__main__) - [1:179545163] SpliceAI scores were not found\n",
      "2024/10/16 08:56:39 [WARNING] (__main__) - [1:183542953] HGVSc were not found\n",
      "2024/10/16 08:56:39 [WARNING] (__main__) - [1:183542953] SpliceAI scores were not found\n",
      "2024/10/16 08:56:40 [WARNING] (__main__) - [1:192722486] HGVSc were not found\n",
      "2024/10/16 08:56:40 [WARNING] (__main__) - [1:192722486] SpliceAI scores were not found\n",
      "2024/10/16 08:56:40 [WARNING] (__main__) - [1:192978719] HGVSc were not found\n",
      "2024/10/16 08:56:40 [WARNING] (__main__) - [1:192978719] SpliceAI scores were not found\n",
      "2024/10/16 08:56:44 [WARNING] (__main__) - [1:196797831] HGVSc were not found\n",
      "2024/10/16 08:56:44 [WARNING] (__main__) - [1:196797831] SpliceAI scores were not found\n",
      "2024/10/16 08:56:45 [WARNING] (__main__) - [1:196921850] HGVSc were not found\n",
      "2024/10/16 08:56:45 [WARNING] (__main__) - [1:196921850] SpliceAI scores were not found\n",
      "2024/10/16 08:56:47 [WARNING] (__main__) - [1:197091966] HGVSc were not found\n",
      "2024/10/16 08:56:47 [WARNING] (__main__) - [1:197091966] SpliceAI scores were not found\n",
      "2024/10/16 08:56:49 [WARNING] (__main__) - [1:197390290] HGVSc were not found\n",
      "2024/10/16 08:56:49 [WARNING] (__main__) - [1:197390290] SpliceAI scores were not found\n",
      "2024/10/16 08:56:55 [WARNING] (__main__) - [1:204165663] HGVSc were not found\n",
      "2024/10/16 08:56:55 [WARNING] (__main__) - [1:204165663] SpliceAI scores were not found\n",
      "2024/10/16 08:56:56 [WARNING] (__main__) - [1:205114791] HGVSc were not found\n",
      "2024/10/16 08:56:56 [WARNING] (__main__) - [1:205114791] SpliceAI scores were not found\n",
      "2024/10/16 08:56:57 [WARNING] (__main__) - [1:207976124] HGVSc were not found\n",
      "2024/10/16 08:56:57 [WARNING] (__main__) - [1:207976124] SpliceAI scores were not found\n",
      "2024/10/16 08:56:58 [WARNING] (__main__) - [1:209825713] HGVSc were not found\n",
      "2024/10/16 08:57:01 [WARNING] (__main__) - [1:209979435] HGVSc were not found\n",
      "2024/10/16 08:57:01 [WARNING] (__main__) - [1:209979435] SpliceAI scores were not found\n",
      "2024/10/16 08:57:01 [WARNING] (__main__) - [1:209989478] HGVSc were not found\n",
      "2024/10/16 08:57:01 [WARNING] (__main__) - [1:209989478] SpliceAI scores were not found\n",
      "2024/10/16 08:57:02 [WARNING] (__main__) - [1:214574150] HGVSc were not found\n",
      "2024/10/16 08:57:02 [WARNING] (__main__) - [1:214574150] SpliceAI scores were not found\n",
      "2024/10/16 08:57:02 [WARNING] (__main__) - [1:215713271] HGVSc were not found\n",
      "2024/10/16 08:57:02 [WARNING] (__main__) - [1:215713271] SpliceAI scores were not found\n",
      "2024/10/16 08:57:02 [WARNING] (__main__) - [1:215803084] HGVSc were not found\n",
      "2024/10/16 08:57:03 [WARNING] (__main__) - [1:215803084] SpliceAI scores were not found\n",
      "2024/10/16 08:57:03 [WARNING] (__main__) - [1:215836169] HGVSc were not found\n",
      "2024/10/16 08:57:03 [WARNING] (__main__) - [1:215836169] SpliceAI scores were not found\n",
      "2024/10/16 08:57:06 [WARNING] (__main__) - [1:215934146] HGVSc were not found\n",
      "2024/10/16 08:57:06 [WARNING] (__main__) - [1:215934146] SpliceAI scores were not found\n",
      "2024/10/16 08:57:07 [WARNING] (__main__) - [1:215949318] HGVSc were not found\n",
      "2024/10/16 08:57:07 [WARNING] (__main__) - [1:215949318] SpliceAI scores were not found\n",
      "2024/10/16 08:57:07 [WARNING] (__main__) - [1:215958622] HGVSc were not found\n",
      "2024/10/16 08:57:07 [WARNING] (__main__) - [1:215958622] SpliceAI scores were not found\n",
      "2024/10/16 08:57:07 [WARNING] (__main__) - [1:215963392] HGVSc were not found\n",
      "2024/10/16 08:57:07 [WARNING] (__main__) - [1:215963392] SpliceAI scores were not found\n",
      "2024/10/16 08:57:08 [WARNING] (__main__) - [1:216009682] HGVSc were not found\n",
      "2024/10/16 08:57:08 [WARNING] (__main__) - [1:216009682] SpliceAI scores were not found\n",
      "2024/10/16 08:57:08 [WARNING] (__main__) - [1:216033421] HGVSc were not found\n",
      "2024/10/16 08:57:08 [WARNING] (__main__) - [1:216033421] SpliceAI scores were not found\n",
      "2024/10/16 08:57:09 [WARNING] (__main__) - [1:216072474] HGVSc were not found\n",
      "2024/10/16 08:57:09 [WARNING] (__main__) - [1:216072474] SpliceAI scores were not found\n",
      "2024/10/16 08:57:10 [WARNING] (__main__) - [1:216167536] HGVSc were not found\n",
      "2024/10/16 08:57:10 [WARNING] (__main__) - [1:216167536] SpliceAI scores were not found\n",
      "2024/10/16 08:57:11 [WARNING] (__main__) - [1:216210458] HGVSc were not found\n",
      "2024/10/16 08:57:11 [WARNING] (__main__) - [1:216210458] SpliceAI scores were not found\n",
      "2024/10/16 08:57:11 [WARNING] (__main__) - [1:216240158] HGVSc were not found\n",
      "2024/10/16 08:57:11 [WARNING] (__main__) - [1:216240158] SpliceAI scores were not found\n",
      "2024/10/16 08:57:11 [WARNING] (__main__) - [1:216246071] HGVSc were not found\n",
      "2024/10/16 08:57:11 [WARNING] (__main__) - [1:216246071] SpliceAI scores were not found\n",
      "2024/10/16 08:57:12 [WARNING] (__main__) - [1:216259362] HGVSc were not found\n",
      "2024/10/16 08:57:12 [WARNING] (__main__) - [1:216259362] SpliceAI scores were not found\n",
      "2024/10/16 08:57:12 [WARNING] (__main__) - [1:216261882] HGVSc were not found\n",
      "2024/10/16 08:57:12 [WARNING] (__main__) - [1:216261882] SpliceAI scores were not found\n",
      "2024/10/16 08:57:13 [WARNING] (__main__) - [1:216369891] HGVSc were not found\n",
      "2024/10/16 08:57:13 [WARNING] (__main__) - [1:216369891] SpliceAI scores were not found\n",
      "2024/10/16 08:57:15 [WARNING] (__main__) - [1:216404578] HGVSc were not found\n",
      "2024/10/16 08:57:15 [WARNING] (__main__) - [1:216404578] SpliceAI scores were not found\n",
      "2024/10/16 08:57:15 [WARNING] (__main__) - [1:216419554] HGVSc were not found\n",
      "2024/10/16 08:57:15 [WARNING] (__main__) - [1:216419554] SpliceAI scores were not found\n",
      "2024/10/16 08:57:16 [WARNING] (__main__) - [1:216466303] HGVSc were not found\n",
      "2024/10/16 08:57:16 [WARNING] (__main__) - [1:216466303] SpliceAI scores were not found\n",
      "2024/10/16 08:57:17 [WARNING] (__main__) - [1:216536919] HGVSc were not found\n",
      "2024/10/16 08:57:17 [WARNING] (__main__) - [1:216536919] SpliceAI scores were not found\n",
      "2024/10/16 08:57:19 [WARNING] (__main__) - [1:220325699] HGVSc were not found\n",
      "2024/10/16 08:57:19 [WARNING] (__main__) - [1:220325699] SpliceAI scores were not found\n",
      "2024/10/16 08:57:22 [WARNING] (__main__) - [1:227165829] HGVSc were not found\n",
      "2024/10/16 08:57:22 [WARNING] (__main__) - [1:227165829] SpliceAI scores were not found\n",
      "2024/10/16 08:57:28 [WARNING] (__main__) - [1:230839695] HGVSc were not found\n",
      "2024/10/16 08:57:28 [WARNING] (__main__) - [1:230839695] SpliceAI scores were not found\n",
      "2024/10/16 08:57:29 [WARNING] (__main__) - [1:232144529] SpliceAI scores were not found\n",
      "2024/10/16 08:57:29 [WARNING] (__main__) - [1:232162292] SpliceAI scores were not found\n",
      "2024/10/16 08:57:30 [WARNING] (__main__) - [1:235891373] HGVSc were not found\n",
      "2024/10/16 08:57:30 [WARNING] (__main__) - [1:235891373] SpliceAI scores were not found\n",
      "2024/10/16 08:57:33 [WARNING] (__main__) - [1:237493799] HGVSc were not found\n",
      "2024/10/16 08:57:33 [WARNING] (__main__) - [1:237493799] SpliceAI scores were not found\n",
      "2024/10/16 08:57:33 [WARNING] (__main__) - [1:237493876] HGVSc were not found\n",
      "2024/10/16 08:57:33 [WARNING] (__main__) - [1:237493876] SpliceAI scores were not found\n",
      "2024/10/16 08:57:33 [WARNING] (__main__) - [1:237493949] HGVSc were not found\n",
      "2024/10/16 08:57:33 [WARNING] (__main__) - [1:237493949] SpliceAI scores were not found\n",
      "2024/10/16 08:57:33 [WARNING] (__main__) - [1:237655363] HGVSc were not found\n",
      "2024/10/16 08:57:33 [WARNING] (__main__) - [1:237655363] SpliceAI scores were not found\n",
      "2024/10/16 08:57:34 [WARNING] (__main__) - [1:237693678] HGVSc were not found\n",
      "2024/10/16 08:57:34 [WARNING] (__main__) - [1:237693678] SpliceAI scores were not found\n",
      "2024/10/16 08:57:37 [WARNING] (__main__) - [1:239019922] HGVSc were not found\n",
      "2024/10/16 08:57:37 [WARNING] (__main__) - [1:239019922] SpliceAI scores were not found\n",
      "2024/10/16 08:57:37 [WARNING] (__main__) - [1:240341240] HGVSc were not found\n",
      "2024/10/16 08:57:37 [WARNING] (__main__) - [1:240341240] SpliceAI scores were not found\n",
      "2024/10/16 08:57:37 [WARNING] (__main__) - [1:241461371] HGVSc were not found\n",
      "2024/10/16 08:57:37 [WARNING] (__main__) - [1:241461371] SpliceAI scores were not found\n",
      "2024/10/16 08:57:37 [WARNING] (__main__) - [1:241649904] HGVSc were not found\n",
      "2024/10/16 08:57:37 [WARNING] (__main__) - [1:241649904] SpliceAI scores were not found\n",
      "2024/10/16 08:57:41 [WARNING] (__main__) - [1:241677012] HGVSc were not found\n",
      "2024/10/16 08:57:41 [WARNING] (__main__) - [1:241677012] SpliceAI scores were not found\n",
      "2024/10/16 08:57:41 [WARNING] (__main__) - [1:241682772] HGVSc were not found\n",
      "2024/10/16 08:57:41 [WARNING] (__main__) - [1:241682772] SpliceAI scores were not found\n",
      "2024/10/16 08:57:41 [WARNING] (__main__) - [1:242263612] HGVSc were not found\n",
      "2024/10/16 08:57:41 [WARNING] (__main__) - [1:242263612] SpliceAI scores were not found\n"
     ]
    }
   ],
   "source": [
    "\n",
    "df: pd.DataFrame = pd.DataFrame(columns=cols)\n",
    "for v in VCF(raw_vcf):\n",
    "    vep: list = v.INFO.get('CSQ').split('|')\n",
    "\n",
    "    # Get SpliceAI scores\n",
    "    if v.INFO.get('SpliceAI'):\n",
    "        splai: list = v.INFO.get('SpliceAI').split(',')[0].split('|')\n",
    "    else:\n",
    "        splai = ['NA'] * len(splai_cols_list)\n",
    "\n",
    "    # Get HGVSc from VEP\n",
    "    try:\n",
    "        hgvsc = re.search('(?<=:).*',vep[vepidx['HGVSc']])[0]\n",
    "    except TypeError:\n",
    "        hgvsc = \"NA\"\n",
    "\n",
    "    # Convert strand to +/- \n",
    "    strand = lambda s: '+' if s == 1 else '-'\n",
    "\n",
    "    # Get max SpliceAI scores\n",
    "    ds_ag: float = splai[splaidx['DS_AG']]\n",
    "    ds_al: float = splai[splaidx['DS_AL']]\n",
    "    ds_dg: float = splai[splaidx['DS_DG']]\n",
    "    ds_dl: float = splai[splaidx['DS_DL']]\n",
    "    if splai[splaidx['DP_AG']] == 'NA':\n",
    "        maxsplai: float = \"NA\"\n",
    "    maxsplai: float = max(ds_ag, ds_al, ds_dg, ds_dl)\n",
    "\n",
    "    # Add df row\n",
    "    df = pd.concat([df, pd.DataFrame([[\n",
    "        v.CHROM, v.POS, v.REF, v.ALT[0], \n",
    "        vep[vepidx['SYMBOL']], vep[vepidx['SYMBOL_SOURCE']], vep[vepidx['HGNC_ID']], \n",
    "        vep[vepidx['Feature']], hgvsc, vep[vepidx['Consequence']], \n",
    "        vep[vepidx['EXON']], vep[vepidx['INTRON']],\n",
    "        strand(vep[vepidx['STRAND']]), \n",
    "        ds_ag, ds_al, ds_dg, ds_dl,\n",
    "        splai[splaidx['DP_AG']], splai[splaidx['DP_AL']], \n",
    "        splai[splaidx['DP_DG']], splai[splaidx['DP_DL']],\n",
    "        maxsplai\n",
    "    ]], columns=cols)], ignore_index=True)\n",
    "\n",
    "    if hgvsc == \"NA\":\n",
    "        logger.warning(f\"[{v.CHROM}:{v.POS}] HGVSc were not found\")\n",
    "    if maxsplai == \"NA\":\n",
    "        logger.warning(f\"[{v.CHROM}:{v.POS}] SpliceAI scores were not found\")\n",
    "\n",
    "# ALLELE|SYMBOL|DS_AG|DS_AL|DS_DG|DS_DL|DP_AG|DP_AL|DP_DG|DP_DL\n",
    "# CHROM, POS, REF, ALT, GeneSymbol, NCBI_ID, ENST, ExonIntronNumbers, FLAGS, SYMBOL_SOURCE|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_pickle('splai_vep_vcfs/hgmd_dm/all_DM_chr1.splai.vep.pkl')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20701\n",
      "20687\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_pickle('splai_vep_vcfs/hgmd_dm/all_DM_chr1.splai.vep.pkl')\n",
    "print(len(df))\n",
    "df.drop_duplicates(inplace=True)\n",
    "print(len(df))\n",
    "\n",
    "import sys\n",
    "try: \n",
    "    __file__\n",
    "    sys.path.append(os.path.join(os.path.dirname('__file__')))\n",
    "except NameError:\n",
    "    Path().resolve()\n",
    "    sys.path.append(os.path.join(Path().resolve(), '../../../'))\n",
    "\n",
    "from libs import utils, preprocess, variantfilter, posparser, splaiparser\n",
    "# from libs import predeffect, scoring\n",
    "from libs import anno_spliceai, anno_clinvar\n",
    "from libs.deco import print_filtering_count\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024/10/16 08:35:47 [INFO   ] (__main__) - Classify \"Canonical\" splice site or \"Non-canonical\" splice site...\n",
      "-2: 321\n",
      "-1: 358\n",
      "+1: 648\n",
      "+2: 224\n",
      "Total variants      : 20687\n",
      "Canonical variants  : 1551\n",
      "non-Canon variants  : 19136\n",
      "\n",
      "2024/10/16 08:35:47 [INFO   ] (__main__) - Calculating exonic positions...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 20687/20687 [00:21<00:00, 973.78it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024/10/16 08:36:09 [INFO   ] (__main__) - Annotating splicing information...\n"
     ]
    }
   ],
   "source": [
    "#1.   Classify 'Canonical' splice site or 'Non-canonical' splice site\n",
    "logger.info('Classify \"Canonical\" splice site or \"Non-canonical\" splice site...')\n",
    "df = posparser.classifying_canonical(df, cdot='HGVSc')\n",
    "\n",
    "#2.   Calculate exonic positions\n",
    "logger.info('Calculating exonic positions...')\n",
    "#2-1. Generate 'exonic upstream distance and exonic downstream distance\n",
    "gencode_gff = '../../../Resources/05_GENCODE_v43lift37/gencode.v43lift37.annotation.sort.gff3.gz'\n",
    "tbx_anno = pysam.TabixFile(gencode_gff)\n",
    "df['exon_loc'] = df.progress_apply(\n",
    "    posparser.calc_exon_loc, tabixfile=tbx_anno, enstcolname='ENST', axis=1)\n",
    "df = pd.concat([df, df['exon_loc'].str.split(':', expand=True)], axis=1)\n",
    "df.rename(columns={0: 'ex_up_dist', 1: 'ex_down_dist'}, inplace=True)\n",
    "df.drop(columns=['exon_loc'], inplace=True)\n",
    "#2-2. Select minimum distance from upstream distance and downstream distance\n",
    "df['exon_pos'] = df.parallel_apply(posparser.select_exon_pos, axis=1)\n",
    "#2-3. Decision exonic splice sites (1 nt in acceptor site or 3 nts on Donor site)\n",
    "df['exon_splice_site'] = df.parallel_apply(posparser.extract_splicing_region, axis=1)\n",
    "\n",
    "#3.   Additional Splicing information\n",
    "logger.info('Annotating splicing information...')\n",
    "#3-1. Annotate splicing type ('Exonic Acceptor' etc.)\n",
    "df['SpliceType'] = df.parallel_apply(posparser.select_donor_acceptor, axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 20687/20687 [09:11<00:00, 37.51it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024/10/16 08:45:21 [INFO   ] (__main__) - Annotating ClinVar varaints interpretations...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|██████████| 20687/20687 [00:02<00:00, 7592.77it/s] \n",
      "100%|██████████| 20687/20687 [00:01<00:00, 10444.36it/s]\n"
     ]
    }
   ],
   "source": [
    "#3-2. Annotate rank of exon or intron\n",
    "try:\n",
    "    db_anno_gencode = '../../../Resources/06_gffutilsdb/gencode.v43lift37.annotation.gtf.db'\n",
    "    db_anno_intron = '../../../Resources/06_gffutilsdb/gencode.v43lift37.annotation.intron.gtf.db'\n",
    "    db = gffutils.FeatureDB(db_anno_gencode)\n",
    "    db_intron = gffutils.FeatureDB(db_anno_intron)\n",
    "except ValueError:\n",
    "    db_anno_gencode = '/resources/DBs/gencode.v43lift37.annotation.gtf.db'\n",
    "    db_anno_intron = '/resources/DBs/gencode.v43lift37.annotation.intron.gtf.db'\n",
    "    db = gffutils.FeatureDB(db_anno_gencode)\n",
    "    db_intron = gffutils.FeatureDB(db_anno_intron)\n",
    "\n",
    "df['ENST_Full'] = df.progress_apply(posparser.fetch_enst_full, db=db, axis=1)\n",
    "df.to_pickle('splai_vep_vcfs/hgmd_dm/all_DM_chr1.splai.vep.enst.pkl')\n",
    "\n",
    "#5.   Annotate ClinVar varaints interpretations\n",
    "logger.info('Annotating ClinVar varaints interpretations...')\n",
    "clinvar_file = '../../../Resources/03_ClinVar/variant_summary.snv.grch37.germline.criteria.sort.bed.gz'\n",
    "tbx_clinvar = pysam.TabixFile(clinvar_file)\n",
    "df['clinvar_same_pos'] = df.progress_apply(\n",
    "    anno_clinvar.anno_same_pos_vars, tabixfile=tbx_clinvar, axis=1)\n",
    "df['clinvar_same_motif'] = df.progress_apply(\n",
    "    anno_clinvar.anno_same_motif_vars, tabixfile=tbx_clinvar, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CHROM                                 1\n",
       "POS                             1464710\n",
       "REF                    AGGTGAGTCAGGCTCC\n",
       "ALT                                   A\n",
       "GeneSymbol                       ATAD3A\n",
       "SymbolSource                       HGNC\n",
       "HGNC_ID                           25567\n",
       "ENST                    ENST00000378755\n",
       "HGVSc               c.1758+2_1758+16del\n",
       "EXON                              15/16\n",
       "INTRON                            15/15\n",
       "Strand                                -\n",
       "DS_AG                              0.00\n",
       "DS_AL                              0.64\n",
       "DS_DG                              0.03\n",
       "DS_DL                              0.99\n",
       "DP_AG                                17\n",
       "DP_AL                              -107\n",
       "DP_DG                                16\n",
       "DP_DL                                 1\n",
       "MaxSpliceAI                        0.99\n",
       "is_Canonical                       True\n",
       "Int_loc                              +2\n",
       "ex_up_dist                            1\n",
       "ex_down_dist                        107\n",
       "exon_pos                              1\n",
       "exon_splice_site          ex_donor_site\n",
       "SpliceType                    Donor_int\n",
       "ENST_Full           ENST00000378755.9_5\n",
       "Name: 149, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[149]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024/10/16 08:45:26 [INFO   ] (__main__) - Parsing SpliceAI results...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  1%|          | 150/20687 [00:03<06:54, 49.55it/s]\n"
     ]
    },
    {
     "ename": "StopIteration",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mStopIteration\u001b[0m                             Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 12\u001b[0m\n\u001b[1;32m      9\u001b[0m TH_sAG, TH_sDG \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0.2\u001b[39m, \u001b[38;5;241m0.2\u001b[39m\n\u001b[1;32m     11\u001b[0m \u001b[38;5;66;03m#6-1. Annotate Exon/Intron position information\u001b[39;00m\n\u001b[0;32m---> 12\u001b[0m df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mExInt_INFO\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43mdf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mprogress_apply\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     13\u001b[0m \u001b[43m    \u001b[49m\u001b[43msplaiparser\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcalc_exint_info\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdb\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdb\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdb_intron\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdb_intron\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/wesanno/lib/python3.9/site-packages/tqdm/std.py:920\u001b[0m, in \u001b[0;36mtqdm.pandas.<locals>.inner_generator.<locals>.inner\u001b[0;34m(df, func, *args, **kwargs)\u001b[0m\n\u001b[1;32m    917\u001b[0m \u001b[38;5;66;03m# Apply the provided function (in **kwargs)\u001b[39;00m\n\u001b[1;32m    918\u001b[0m \u001b[38;5;66;03m# on the df using our wrapper (which provides bar updating)\u001b[39;00m\n\u001b[1;32m    919\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 920\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mgetattr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mdf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdf_function\u001b[49m\u001b[43m)\u001b[49m\u001b[43m(\u001b[49m\u001b[43mwrapper\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    921\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    922\u001b[0m     t\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[0;32m~/miniforge3/envs/wesanno/lib/python3.9/site-packages/pandas/core/frame.py:10374\u001b[0m, in \u001b[0;36mDataFrame.apply\u001b[0;34m(self, func, axis, raw, result_type, args, by_row, engine, engine_kwargs, **kwargs)\u001b[0m\n\u001b[1;32m  10360\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mcore\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mapply\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m frame_apply\n\u001b[1;32m  10362\u001b[0m op \u001b[38;5;241m=\u001b[39m frame_apply(\n\u001b[1;32m  10363\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m  10364\u001b[0m     func\u001b[38;5;241m=\u001b[39mfunc,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m  10372\u001b[0m     kwargs\u001b[38;5;241m=\u001b[39mkwargs,\n\u001b[1;32m  10373\u001b[0m )\n\u001b[0;32m> 10374\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mop\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39m__finalize__(\u001b[38;5;28mself\u001b[39m, method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mapply\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/miniforge3/envs/wesanno/lib/python3.9/site-packages/pandas/core/apply.py:916\u001b[0m, in \u001b[0;36mFrameApply.apply\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    913\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mraw:\n\u001b[1;32m    914\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply_raw(engine\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mengine, engine_kwargs\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mengine_kwargs)\n\u001b[0;32m--> 916\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply_standard\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/wesanno/lib/python3.9/site-packages/pandas/core/apply.py:1063\u001b[0m, in \u001b[0;36mFrameApply.apply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1061\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mapply_standard\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m   1062\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mengine \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpython\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m-> 1063\u001b[0m         results, res_index \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply_series_generator\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1064\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1065\u001b[0m         results, res_index \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply_series_numba()\n",
      "File \u001b[0;32m~/miniforge3/envs/wesanno/lib/python3.9/site-packages/pandas/core/apply.py:1081\u001b[0m, in \u001b[0;36mFrameApply.apply_series_generator\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1078\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m option_context(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmode.chained_assignment\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m   1079\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m i, v \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(series_gen):\n\u001b[1;32m   1080\u001b[0m         \u001b[38;5;66;03m# ignore SettingWithCopy here in case the user mutates\u001b[39;00m\n\u001b[0;32m-> 1081\u001b[0m         results[i] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mv\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1082\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(results[i], ABCSeries):\n\u001b[1;32m   1083\u001b[0m             \u001b[38;5;66;03m# If we have a view on v, we need to make a copy because\u001b[39;00m\n\u001b[1;32m   1084\u001b[0m             \u001b[38;5;66;03m#  series_generator will swap out the underlying data\u001b[39;00m\n\u001b[1;32m   1085\u001b[0m             results[i] \u001b[38;5;241m=\u001b[39m results[i]\u001b[38;5;241m.\u001b[39mcopy(deep\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "File \u001b[0;32m~/miniforge3/envs/wesanno/lib/python3.9/site-packages/tqdm/std.py:915\u001b[0m, in \u001b[0;36mtqdm.pandas.<locals>.inner_generator.<locals>.inner.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    909\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwrapper\u001b[39m(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    910\u001b[0m     \u001b[38;5;66;03m# update tbar correctly\u001b[39;00m\n\u001b[1;32m    911\u001b[0m     \u001b[38;5;66;03m# it seems `pandas apply` calls `func` twice\u001b[39;00m\n\u001b[1;32m    912\u001b[0m     \u001b[38;5;66;03m# on the first column/row to decide whether it can\u001b[39;00m\n\u001b[1;32m    913\u001b[0m     \u001b[38;5;66;03m# take a fast or slow code path; so stop when t.total==t.n\u001b[39;00m\n\u001b[1;32m    914\u001b[0m     t\u001b[38;5;241m.\u001b[39mupdate(n\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m t\u001b[38;5;241m.\u001b[39mtotal \u001b[38;5;129;01mor\u001b[39;00m t\u001b[38;5;241m.\u001b[39mn \u001b[38;5;241m<\u001b[39m t\u001b[38;5;241m.\u001b[39mtotal \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;241m0\u001b[39m)\n\u001b[0;32m--> 915\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/Volumes/vol/work/Github/dev/ValidationData/TP/additional/../../../libs/splaiparser.py:64\u001b[0m, in \u001b[0;36mcalc_exint_info\u001b[0;34m(row, db, db_intron)\u001b[0m\n\u001b[1;32m     61\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mis_Canonical\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTrue\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[1;32m     62\u001b[0m     fetched_data \u001b[38;5;241m=\u001b[39m db_intron\u001b[38;5;241m.\u001b[39mchildren(\n\u001b[1;32m     63\u001b[0m         query_enst, limit\u001b[38;5;241m=\u001b[39mregion, featuretype\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mintron\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m---> 64\u001b[0m     d \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mnext\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mfetched_data\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     67\u001b[0m \u001b[38;5;66;03m## Set attributes and current featuretype\u001b[39;00m\n\u001b[1;32m     68\u001b[0m d_attr: \u001b[38;5;28mlist\u001b[39m \u001b[38;5;241m=\u001b[39m d\u001b[38;5;241m.\u001b[39mattributes\n",
      "\u001b[0;31mStopIteration\u001b[0m: "
     ]
    }
   ],
   "source": [
    "df = pd.read_pickle('splai_vep_vcfs/hgmd_dm/all_DM_chr1.splai.vep.enst.pkl')\n",
    "\n",
    "#6.   Parising SpliceAI results (For PVS1 decision tree)\n",
    "logger.info('Parsing SpliceAI results...')\n",
    "## Thresholds configuration\n",
    "TH_min_sALDL, TH_max_sALDL = 0.02, 0.2\n",
    "TH_min_sAGDG, TH_max_sAGDG = 0.01, 0.05\n",
    "TH_min_GExon, TH_max_GExon = 25, 500\n",
    "TH_sAG, TH_sDG = 0.2, 0.2\n",
    "\n",
    "#6-1. Annotate Exon/Intron position information\n",
    "df['ExInt_INFO'] = df.progress_apply(\n",
    "    splaiparser.calc_exint_info, db=db, db_intron=db_intron, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#6-2. Relative exon location\n",
    "df['prc_exon_loc'] = df.parallel_apply(posparser.calc_prc_exon_loc, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_pickle('splai_vep_vcfs/hgmd_dm/all_DM_chr1.splai.vep.enst.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EXON</th>\n",
       "      <th>INTRON</th>\n",
       "      <th>Int_loc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14/14</td>\n",
       "      <td></td>\n",
       "      <td>Exonic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td></td>\n",
       "      <td>1/1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2/2</td>\n",
       "      <td></td>\n",
       "      <td>Exonic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2/2</td>\n",
       "      <td></td>\n",
       "      <td>Exonic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2/2</td>\n",
       "      <td></td>\n",
       "      <td>Exonic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20696</th>\n",
       "      <td>5/9</td>\n",
       "      <td></td>\n",
       "      <td>Exonic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20697</th>\n",
       "      <td>6/9</td>\n",
       "      <td></td>\n",
       "      <td>Exonic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20698</th>\n",
       "      <td>6/9</td>\n",
       "      <td></td>\n",
       "      <td>Exonic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20699</th>\n",
       "      <td>7/9</td>\n",
       "      <td></td>\n",
       "      <td>Exonic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20700</th>\n",
       "      <td>7/9</td>\n",
       "      <td></td>\n",
       "      <td>Exonic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20687 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        EXON INTRON Int_loc\n",
       "0      14/14         Exonic\n",
       "1               1/1      -1\n",
       "2        2/2         Exonic\n",
       "3        2/2         Exonic\n",
       "4        2/2         Exonic\n",
       "...      ...    ...     ...\n",
       "20696    5/9         Exonic\n",
       "20697    6/9         Exonic\n",
       "20698    6/9         Exonic\n",
       "20699    7/9         Exonic\n",
       "20700    7/9         Exonic\n",
       "\n",
       "[20687 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['EXON', 'INTRON', 'Int_loc']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024/10/16 08:23:10 [INFO   ] (__main__) - Annotating ClinVar varaints interpretations...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 20687/20687 [00:02<00:00, 9106.89it/s] \n",
      "100%|██████████| 20687/20687 [00:02<00:00, 10212.42it/s]\n"
     ]
    }
   ],
   "source": [
    "#5.   Annotate ClinVar varaints interpretations\n",
    "logger.info('Annotating ClinVar varaints interpretations...')\n",
    "clinvar_file = '../../../Resources/03_ClinVar/variant_summary.snv.grch37.germline.criteria.sort.bed.gz'\n",
    "tbx_clinvar = pysam.TabixFile(clinvar_file)\n",
    "df['clinvar_same_pos'] = df.progress_apply(anno_clinvar.anno_same_pos_vars, \n",
    "                                           tabixfile=tbx_clinvar, axis=1)\n",
    "df['clinvar_same_motif'] = df.progress_apply(anno_clinvar.anno_same_motif_vars, \n",
    "                                             tabixfile=tbx_clinvar, axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>SymbolSource</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ENST</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>EXON</th>\n",
       "      <th>...</th>\n",
       "      <th>is_Canonical</th>\n",
       "      <th>Int_loc</th>\n",
       "      <th>ex_up_dist</th>\n",
       "      <th>ex_down_dist</th>\n",
       "      <th>exon_pos</th>\n",
       "      <th>exon_splice_site</th>\n",
       "      <th>SpliceType</th>\n",
       "      <th>ENST_Full</th>\n",
       "      <th>clinvar_same_pos</th>\n",
       "      <th>clinvar_same_motif</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>879375</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>28706</td>\n",
       "      <td>ENST00000342066</td>\n",
       "      <td>c.1888C&gt;T</td>\n",
       "      <td>14/14</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>579</td>\n",
       "      <td>87</td>\n",
       "      <td>579</td>\n",
       "      <td>non_SplExon</td>\n",
       "      <td>Acceptor_ex</td>\n",
       "      <td>ENST00000342066.8_7</td>\n",
       "      <td>[Pathogenic_1:879375]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>949363</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>4053</td>\n",
       "      <td>ENST00000379389</td>\n",
       "      <td>c.4-1G&gt;A</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>None</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_with_Ver_not_available</td>\n",
       "      <td>[]</td>\n",
       "      <td>unk_SpliceType</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>949443</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>4053</td>\n",
       "      <td>ENST00000379389</td>\n",
       "      <td>c.83T&gt;A</td>\n",
       "      <td>2/2</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>None</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_with_Ver_not_available</td>\n",
       "      <td>[]</td>\n",
       "      <td>unk_SpliceType</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>949523</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>4053</td>\n",
       "      <td>ENST00000379389</td>\n",
       "      <td>c.163C&gt;T</td>\n",
       "      <td>2/2</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>None</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_with_Ver_not_available</td>\n",
       "      <td>[]</td>\n",
       "      <td>unk_SpliceType</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>949643</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>4053</td>\n",
       "      <td>ENST00000379389</td>\n",
       "      <td>c.285del</td>\n",
       "      <td>2/2</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>None</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_with_Ver_not_available</td>\n",
       "      <td>[]</td>\n",
       "      <td>unk_SpliceType</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20696</th>\n",
       "      <td>1</td>\n",
       "      <td>247597502</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NLRP3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>16400</td>\n",
       "      <td>ENST00000336119</td>\n",
       "      <td>c.2425G&gt;A</td>\n",
       "      <td>5/9</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>73</td>\n",
       "      <td>97</td>\n",
       "      <td>73</td>\n",
       "      <td>non_SplExon</td>\n",
       "      <td>Donor_ex</td>\n",
       "      <td>ENST00000336119.8_4</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20697</th>\n",
       "      <td>1</td>\n",
       "      <td>247599354</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NLRP3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>16400</td>\n",
       "      <td>ENST00000336119</td>\n",
       "      <td>c.2581T&gt;C</td>\n",
       "      <td>6/9</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>88</td>\n",
       "      <td>82</td>\n",
       "      <td>88</td>\n",
       "      <td>non_SplExon</td>\n",
       "      <td>Acceptor_ex</td>\n",
       "      <td>ENST00000336119.8_4</td>\n",
       "      <td>[Likely pathogenic_1:247599354]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20698</th>\n",
       "      <td>1</td>\n",
       "      <td>247599355</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NLRP3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>16400</td>\n",
       "      <td>ENST00000336119</td>\n",
       "      <td>c.2582A&gt;G</td>\n",
       "      <td>6/9</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>87</td>\n",
       "      <td>83</td>\n",
       "      <td>87</td>\n",
       "      <td>non_SplExon</td>\n",
       "      <td>Acceptor_ex</td>\n",
       "      <td>ENST00000336119.8_4</td>\n",
       "      <td>[Pathogenic_1:247599355]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20699</th>\n",
       "      <td>1</td>\n",
       "      <td>247607362</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NLRP3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>16400</td>\n",
       "      <td>ENST00000336119</td>\n",
       "      <td>c.2758C&gt;T</td>\n",
       "      <td>7/9</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>82</td>\n",
       "      <td>88</td>\n",
       "      <td>82</td>\n",
       "      <td>non_SplExon</td>\n",
       "      <td>Donor_ex</td>\n",
       "      <td>ENST00000336119.8_4</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20700</th>\n",
       "      <td>1</td>\n",
       "      <td>247607363</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NLRP3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>16400</td>\n",
       "      <td>ENST00000336119</td>\n",
       "      <td>c.2759G&gt;A</td>\n",
       "      <td>7/9</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>81</td>\n",
       "      <td>89</td>\n",
       "      <td>81</td>\n",
       "      <td>non_SplExon</td>\n",
       "      <td>Donor_ex</td>\n",
       "      <td>ENST00000336119.8_4</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20687 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM        POS REF ALT GeneSymbol SymbolSource HGNC_ID  \\\n",
       "0         1     879375   C   T     SAMD11         HGNC   28706   \n",
       "1         1     949363   G   A      ISG15         HGNC    4053   \n",
       "2         1     949443   T   A      ISG15         HGNC    4053   \n",
       "3         1     949523   C   T      ISG15         HGNC    4053   \n",
       "4         1     949643  AC   A      ISG15         HGNC    4053   \n",
       "...     ...        ...  ..  ..        ...          ...     ...   \n",
       "20696     1  247597502   G   A      NLRP3         HGNC   16400   \n",
       "20697     1  247599354   T   C      NLRP3         HGNC   16400   \n",
       "20698     1  247599355   A   G      NLRP3         HGNC   16400   \n",
       "20699     1  247607362   C   T      NLRP3         HGNC   16400   \n",
       "20700     1  247607363   G   A      NLRP3         HGNC   16400   \n",
       "\n",
       "                  ENST      HGVSc   EXON  ... is_Canonical Int_loc  \\\n",
       "0      ENST00000342066  c.1888C>T  14/14  ...        False  Exonic   \n",
       "1      ENST00000379389   c.4-1G>A         ...         True      -1   \n",
       "2      ENST00000379389    c.83T>A    2/2  ...        False  Exonic   \n",
       "3      ENST00000379389   c.163C>T    2/2  ...        False  Exonic   \n",
       "4      ENST00000379389   c.285del    2/2  ...        False  Exonic   \n",
       "...                ...        ...    ...  ...          ...     ...   \n",
       "20696  ENST00000336119  c.2425G>A    5/9  ...        False  Exonic   \n",
       "20697  ENST00000336119  c.2581T>C    6/9  ...        False  Exonic   \n",
       "20698  ENST00000336119  c.2582A>G    6/9  ...        False  Exonic   \n",
       "20699  ENST00000336119  c.2758C>T    7/9  ...        False  Exonic   \n",
       "20700  ENST00000336119  c.2759G>A    7/9  ...        False  Exonic   \n",
       "\n",
       "           ex_up_dist    ex_down_dist        exon_pos exon_splice_site  \\\n",
       "0                 579              87             579      non_SplExon   \n",
       "1      ENST_not_match  ENST_not_match  ENST_not_match             None   \n",
       "2      ENST_not_match  ENST_not_match  ENST_not_match             None   \n",
       "3      ENST_not_match  ENST_not_match  ENST_not_match             None   \n",
       "4      ENST_not_match  ENST_not_match  ENST_not_match             None   \n",
       "...               ...             ...             ...              ...   \n",
       "20696              73              97              73      non_SplExon   \n",
       "20697              88              82              88      non_SplExon   \n",
       "20698              87              83              87      non_SplExon   \n",
       "20699              82              88              82      non_SplExon   \n",
       "20700              81              89              81      non_SplExon   \n",
       "\n",
       "           SpliceType                    ENST_Full  \\\n",
       "0         Acceptor_ex          ENST00000342066.8_7   \n",
       "1      ENST_not_match  ENST_with_Ver_not_available   \n",
       "2      ENST_not_match  ENST_with_Ver_not_available   \n",
       "3      ENST_not_match  ENST_with_Ver_not_available   \n",
       "4      ENST_not_match  ENST_with_Ver_not_available   \n",
       "...               ...                          ...   \n",
       "20696        Donor_ex          ENST00000336119.8_4   \n",
       "20697     Acceptor_ex          ENST00000336119.8_4   \n",
       "20698     Acceptor_ex          ENST00000336119.8_4   \n",
       "20699        Donor_ex          ENST00000336119.8_4   \n",
       "20700        Donor_ex          ENST00000336119.8_4   \n",
       "\n",
       "                      clinvar_same_pos clinvar_same_motif  \n",
       "0                [Pathogenic_1:879375]                 []  \n",
       "1                                   []     unk_SpliceType  \n",
       "2                                   []     unk_SpliceType  \n",
       "3                                   []     unk_SpliceType  \n",
       "4                                   []     unk_SpliceType  \n",
       "...                                ...                ...  \n",
       "20696                               []                 []  \n",
       "20697  [Likely pathogenic_1:247599354]                 []  \n",
       "20698         [Pathogenic_1:247599355]                 []  \n",
       "20699                               []                 []  \n",
       "20700                               []                 []  \n",
       "\n",
       "[20687 rows x 31 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\tHAVANA\ttranscript\t247579475\t247612119\t.\t+\t.\tgene_id \"ENSG00000162711.18_12\"; transcript_id \"ENST00000336119.8_4\"; gene_type \"protein_coding\"; gene_name \"NLRP3\"; transcript_type \"protein_coding\"; transcript_name \"NLRP3-201\"; level \"2\"; protein_id \"ENSP00000337383.4\"; transcript_support_level \"1\"; hgnc_id \"HGNC:16400\"; tag \"CAGE_supported_TSS\"; tag \"upstream_ATG\"; tag \"basic\"; tag \"Ensembl_canonical\"; tag \"MANE_Select\"; tag \"appris_principal_1\"; havana_gene \"OTTHUMG00000040647.7_12\"; havana_transcript \"OTTHUMT00000097740.1_4\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "ENST00000336119.8_4\n"
     ]
    }
   ],
   "source": [
    "query_chr = 'chr1'\n",
    "query_pos = 247599354\n",
    "query_enst = 'ENST00000336119'\n",
    "query_region = (query_chr, query_pos-1, query_pos)\n",
    "\n",
    "# def fetch_enst_full(enst: str) -> str:\n",
    "\n",
    "for transcript in db.region(region=query_region, featuretype='transcript'):\n",
    "    if transcript.id.startswith(query_enst):\n",
    "        print(transcript)\n",
    "        print(transcript.id)\n",
    "    # print(transcript)\n",
    "    # print(transcript.id)\n",
    "    # break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\tHAVANA\tgene\t12010\t13670\t.\t+\t.\tgene_id \"ENSG00000223972.6_6\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; level \"2\"; hgnc_id \"HGNC:37102\"; havana_gene \"OTTHUMG00000000961.2_6\"; remap_status \"full_contig\"; remap_num_mappings \"1\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\ttranscript\t12010\t13670\t.\t+\t.\tgene_id \"ENSG00000223972.6_6\"; transcript_id \"ENST00000450305.2_2\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"transcribed_unprocessed_pseudogene\"; transcript_name \"DDX11L1-201\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:37102\"; ont \"PGO:0000005\"; ont \"PGO:0000019\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000961.2_6\"; havana_transcript \"OTTHUMT00000002844.2_2\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\texon\t12010\t12057\t.\t+\t.\tgene_id \"ENSG00000223972.6_6\"; transcript_id \"ENST00000450305.2_2\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"transcribed_unprocessed_pseudogene\"; transcript_name \"DDX11L1-201\"; exon_number \"1\"; exon_id \"ENSE00001948541.1_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:37102\"; ont \"PGO:0000005\"; ont \"PGO:0000019\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000961.2_6\"; havana_transcript \"OTTHUMT00000002844.2_2\"; remap_original_location \"chr1:+:12010-12057\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t12179\t12227\t.\t+\t.\tgene_id \"ENSG00000223972.6_6\"; transcript_id \"ENST00000450305.2_2\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"transcribed_unprocessed_pseudogene\"; transcript_name \"DDX11L1-201\"; exon_number \"2\"; exon_id \"ENSE00001671638.2_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:37102\"; ont \"PGO:0000005\"; ont \"PGO:0000019\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000961.2_6\"; havana_transcript \"OTTHUMT00000002844.2_2\"; remap_original_location \"chr1:+:12179-12227\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t12613\t12697\t.\t+\t.\tgene_id \"ENSG00000223972.6_6\"; transcript_id \"ENST00000450305.2_2\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"transcribed_unprocessed_pseudogene\"; transcript_name \"DDX11L1-201\"; exon_number \"3\"; exon_id \"ENSE00001758273.2_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:37102\"; ont \"PGO:0000005\"; ont \"PGO:0000019\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000961.2_6\"; havana_transcript \"OTTHUMT00000002844.2_2\"; remap_original_location \"chr1:+:12613-12697\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t12975\t13052\t.\t+\t.\tgene_id \"ENSG00000223972.6_6\"; transcript_id \"ENST00000450305.2_2\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"transcribed_unprocessed_pseudogene\"; transcript_name \"DDX11L1-201\"; exon_number \"4\"; exon_id \"ENSE00001799933.2_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:37102\"; ont \"PGO:0000005\"; ont \"PGO:0000019\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000961.2_6\"; havana_transcript \"OTTHUMT00000002844.2_2\"; remap_original_location \"chr1:+:12975-13052\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t13221\t13374\t.\t+\t.\tgene_id \"ENSG00000223972.6_6\"; transcript_id \"ENST00000450305.2_2\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"transcribed_unprocessed_pseudogene\"; transcript_name \"DDX11L1-201\"; exon_number \"5\"; exon_id \"ENSE00001746346.2_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:37102\"; ont \"PGO:0000005\"; ont \"PGO:0000019\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000961.2_6\"; havana_transcript \"OTTHUMT00000002844.2_2\"; remap_original_location \"chr1:+:13221-13374\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t13453\t13670\t.\t+\t.\tgene_id \"ENSG00000223972.6_6\"; transcript_id \"ENST00000450305.2_2\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"transcribed_unprocessed_pseudogene\"; transcript_name \"DDX11L1-201\"; exon_number \"6\"; exon_id \"ENSE00001863096.1_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:37102\"; ont \"PGO:0000005\"; ont \"PGO:0000019\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000961.2_6\"; havana_transcript \"OTTHUMT00000002844.2_2\"; remap_original_location \"chr1:+:13453-13670\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\tgene\t14404\t29570\t.\t-\t.\tgene_id \"ENSG00000227232.5_5\"; gene_type \"unprocessed_pseudogene\"; gene_name \"WASH7P\"; level \"2\"; hgnc_id \"HGNC:38034\"; havana_gene \"OTTHUMG00000000958.1_5\"; remap_status \"full_contig\"; remap_num_mappings \"1\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\ttranscript\t14404\t29570\t.\t-\t.\tgene_id \"ENSG00000227232.5_5\"; transcript_id \"ENST00000488147.1_4\"; gene_type \"unprocessed_pseudogene\"; gene_name \"WASH7P\"; transcript_type \"unprocessed_pseudogene\"; transcript_name \"WASH7P-201\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:38034\"; ont \"PGO:0000005\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000958.1_5\"; havana_transcript \"OTTHUMT00000002839.1_4\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\texon\t29534\t29570\t.\t-\t.\tgene_id \"ENSG00000227232.5_5\"; transcript_id \"ENST00000488147.1_4\"; gene_type \"unprocessed_pseudogene\"; gene_name \"WASH7P\"; transcript_type \"unprocessed_pseudogene\"; transcript_name \"WASH7P-201\"; exon_number \"1\"; exon_id \"ENSE00001890219.1_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:38034\"; ont \"PGO:0000005\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000958.1_5\"; havana_transcript \"OTTHUMT00000002839.1_4\"; remap_original_location \"chr1:-:29534-29570\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t24738\t24891\t.\t-\t.\tgene_id \"ENSG00000227232.5_5\"; transcript_id \"ENST00000488147.1_4\"; gene_type \"unprocessed_pseudogene\"; gene_name \"WASH7P\"; transcript_type \"unprocessed_pseudogene\"; transcript_name \"WASH7P-201\"; exon_number \"2\"; exon_id \"ENSE00003507205.1_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:38034\"; ont \"PGO:0000005\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000958.1_5\"; havana_transcript \"OTTHUMT00000002839.1_4\"; remap_original_location \"chr1:-:24738-24891\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t18268\t18366\t.\t-\t.\tgene_id \"ENSG00000227232.5_5\"; transcript_id \"ENST00000488147.1_4\"; gene_type \"unprocessed_pseudogene\"; gene_name \"WASH7P\"; transcript_type \"unprocessed_pseudogene\"; transcript_name \"WASH7P-201\"; exon_number \"3\"; exon_id \"ENSE00003477500.1_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:38034\"; ont \"PGO:0000005\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000958.1_5\"; havana_transcript \"OTTHUMT00000002839.1_4\"; remap_original_location \"chr1:-:18268-18366\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t17915\t18061\t.\t-\t.\tgene_id \"ENSG00000227232.5_5\"; transcript_id \"ENST00000488147.1_4\"; gene_type \"unprocessed_pseudogene\"; gene_name \"WASH7P\"; transcript_type \"unprocessed_pseudogene\"; transcript_name \"WASH7P-201\"; exon_number \"4\"; exon_id \"ENSE00003565697.1_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:38034\"; ont \"PGO:0000005\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000958.1_5\"; havana_transcript \"OTTHUMT00000002839.1_4\"; remap_original_location \"chr1:-:17915-18061\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t17606\t17742\t.\t-\t.\tgene_id \"ENSG00000227232.5_5\"; transcript_id \"ENST00000488147.1_4\"; gene_type \"unprocessed_pseudogene\"; gene_name \"WASH7P\"; transcript_type \"unprocessed_pseudogene\"; transcript_name \"WASH7P-201\"; exon_number \"5\"; exon_id \"ENSE00003475637.1_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:38034\"; ont \"PGO:0000005\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000958.1_5\"; havana_transcript \"OTTHUMT00000002839.1_4\"; remap_original_location \"chr1:-:17606-17742\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t17233\t17368\t.\t-\t.\tgene_id \"ENSG00000227232.5_5\"; transcript_id \"ENST00000488147.1_4\"; gene_type \"unprocessed_pseudogene\"; gene_name \"WASH7P\"; transcript_type \"unprocessed_pseudogene\"; transcript_name \"WASH7P-201\"; exon_number \"6\"; exon_id \"ENSE00003502542.1_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:38034\"; ont \"PGO:0000005\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000958.1_5\"; havana_transcript \"OTTHUMT00000002839.1_4\"; remap_original_location \"chr1:-:17233-17368\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t16858\t17055\t.\t-\t.\tgene_id \"ENSG00000227232.5_5\"; transcript_id \"ENST00000488147.1_4\"; gene_type \"unprocessed_pseudogene\"; gene_name \"WASH7P\"; transcript_type \"unprocessed_pseudogene\"; transcript_name \"WASH7P-201\"; exon_number \"7\"; exon_id \"ENSE00003553898.1_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:38034\"; ont \"PGO:0000005\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000958.1_5\"; havana_transcript \"OTTHUMT00000002839.1_4\"; remap_original_location \"chr1:-:16858-17055\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t16607\t16765\t.\t-\t.\tgene_id \"ENSG00000227232.5_5\"; transcript_id \"ENST00000488147.1_4\"; gene_type \"unprocessed_pseudogene\"; gene_name \"WASH7P\"; transcript_type \"unprocessed_pseudogene\"; transcript_name \"WASH7P-201\"; exon_number \"8\"; exon_id \"ENSE00003621279.1_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:38034\"; ont \"PGO:0000005\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000958.1_5\"; havana_transcript \"OTTHUMT00000002839.1_4\"; remap_original_location \"chr1:-:16607-16765\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t15796\t15947\t.\t-\t.\tgene_id \"ENSG00000227232.5_5\"; transcript_id \"ENST00000488147.1_4\"; gene_type \"unprocessed_pseudogene\"; gene_name \"WASH7P\"; transcript_type \"unprocessed_pseudogene\"; transcript_name \"WASH7P-201\"; exon_number \"9\"; exon_id \"ENSE00002030414.1_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:38034\"; ont \"PGO:0000005\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000958.1_5\"; havana_transcript \"OTTHUMT00000002839.1_4\"; remap_original_location \"chr1:-:15796-15947\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t15005\t15038\t.\t-\t.\tgene_id \"ENSG00000227232.5_5\"; transcript_id \"ENST00000488147.1_4\"; gene_type \"unprocessed_pseudogene\"; gene_name \"WASH7P\"; transcript_type \"unprocessed_pseudogene\"; transcript_name \"WASH7P-201\"; exon_number \"10\"; exon_id \"ENSE00001935574.1_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:38034\"; ont \"PGO:0000005\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000958.1_5\"; havana_transcript \"OTTHUMT00000002839.1_4\"; remap_original_location \"chr1:-:15005-15038\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t14404\t14501\t.\t-\t.\tgene_id \"ENSG00000227232.5_5\"; transcript_id \"ENST00000488147.1_4\"; gene_type \"unprocessed_pseudogene\"; gene_name \"WASH7P\"; transcript_type \"unprocessed_pseudogene\"; transcript_name \"WASH7P-201\"; exon_number \"11\"; exon_id \"ENSE00001843071.1_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:38034\"; ont \"PGO:0000005\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000958.1_5\"; havana_transcript \"OTTHUMT00000002839.1_4\"; remap_original_location \"chr1:-:14404-14501\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\tgene\t29554\t31109\t.\t+\t.\tgene_id \"ENSG00000243485.5_11\"; gene_type \"lncRNA\"; gene_name \"MIR1302-2HG\"; level \"2\"; hgnc_id \"HGNC:52482\"; tag \"ncRNA_host\"; havana_gene \"OTTHUMG00000000959.2_11\"; remap_status \"full_contig\"; remap_num_mappings \"1\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\ttranscript\t29554\t31097\t.\t+\t.\tgene_id \"ENSG00000243485.5_11\"; transcript_id \"ENST00000473358.1_4\"; gene_type \"lncRNA\"; gene_name \"MIR1302-2HG\"; transcript_type \"lncRNA\"; transcript_name \"MIR1302-2HG-202\"; level \"2\"; transcript_support_level \"5\"; hgnc_id \"HGNC:52482\"; tag \"not_best_in_genome_evidence\"; tag \"dotter_confirmed\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000959.2_11\"; havana_transcript \"OTTHUMT00000002840.1_4\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\texon\t29554\t30039\t.\t+\t.\tgene_id \"ENSG00000243485.5_11\"; transcript_id \"ENST00000473358.1_4\"; gene_type \"lncRNA\"; gene_name \"MIR1302-2HG\"; transcript_type \"lncRNA\"; transcript_name \"MIR1302-2HG-202\"; exon_number \"1\"; exon_id \"ENSE00001947070.1_1\"; level \"2\"; transcript_support_level \"5\"; hgnc_id \"HGNC:52482\"; tag \"not_best_in_genome_evidence\"; tag \"dotter_confirmed\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000959.2_11\"; havana_transcript \"OTTHUMT00000002840.1_4\"; remap_original_location \"chr1:+:29554-30039\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t30564\t30667\t.\t+\t.\tgene_id \"ENSG00000243485.5_11\"; transcript_id \"ENST00000473358.1_4\"; gene_type \"lncRNA\"; gene_name \"MIR1302-2HG\"; transcript_type \"lncRNA\"; transcript_name \"MIR1302-2HG-202\"; exon_number \"2\"; exon_id \"ENSE00001922571.1_1\"; level \"2\"; transcript_support_level \"5\"; hgnc_id \"HGNC:52482\"; tag \"not_best_in_genome_evidence\"; tag \"dotter_confirmed\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000959.2_11\"; havana_transcript \"OTTHUMT00000002840.1_4\"; remap_original_location \"chr1:+:30564-30667\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t30976\t31097\t.\t+\t.\tgene_id \"ENSG00000243485.5_11\"; transcript_id \"ENST00000473358.1_4\"; gene_type \"lncRNA\"; gene_name \"MIR1302-2HG\"; transcript_type \"lncRNA\"; transcript_name \"MIR1302-2HG-202\"; exon_number \"3\"; exon_id \"ENSE00001827679.1_1\"; level \"2\"; transcript_support_level \"5\"; hgnc_id \"HGNC:52482\"; tag \"not_best_in_genome_evidence\"; tag \"dotter_confirmed\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000959.2_11\"; havana_transcript \"OTTHUMT00000002840.1_4\"; remap_original_location \"chr1:+:30976-31097\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\ttranscript\t30267\t31109\t.\t+\t.\tgene_id \"ENSG00000243485.5_11\"; transcript_id \"ENST00000469289.1_1\"; gene_type \"lncRNA\"; gene_name \"MIR1302-2HG\"; transcript_type \"lncRNA\"; transcript_name \"MIR1302-2HG-201\"; level \"2\"; transcript_support_level \"5\"; hgnc_id \"HGNC:52482\"; tag \"not_best_in_genome_evidence\"; tag \"basic\"; havana_gene \"OTTHUMG00000000959.2_11\"; havana_transcript \"OTTHUMT00000002841.2_1\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\texon\t30267\t30667\t.\t+\t.\tgene_id \"ENSG00000243485.5_11\"; transcript_id \"ENST00000469289.1_1\"; gene_type \"lncRNA\"; gene_name \"MIR1302-2HG\"; transcript_type \"lncRNA\"; transcript_name \"MIR1302-2HG-201\"; exon_number \"1\"; exon_id \"ENSE00001841699.1_1\"; level \"2\"; transcript_support_level \"5\"; hgnc_id \"HGNC:52482\"; tag \"not_best_in_genome_evidence\"; tag \"basic\"; havana_gene \"OTTHUMG00000000959.2_11\"; havana_transcript \"OTTHUMT00000002841.2_1\"; remap_original_location \"chr1:+:30267-30667\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t30976\t31109\t.\t+\t.\tgene_id \"ENSG00000243485.5_11\"; transcript_id \"ENST00000469289.1_1\"; gene_type \"lncRNA\"; gene_name \"MIR1302-2HG\"; transcript_type \"lncRNA\"; transcript_name \"MIR1302-2HG-201\"; exon_number \"2\"; exon_id \"ENSE00001890064.1_1\"; level \"2\"; transcript_support_level \"5\"; hgnc_id \"HGNC:52482\"; tag \"not_best_in_genome_evidence\"; tag \"basic\"; havana_gene \"OTTHUMG00000000959.2_11\"; havana_transcript \"OTTHUMT00000002841.2_1\"; remap_original_location \"chr1:+:30976-31109\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\tgene\t34554\t36081\t.\t-\t.\tgene_id \"ENSG00000237613.2_6\"; gene_type \"lncRNA\"; gene_name \"FAM138A\"; level \"2\"; hgnc_id \"HGNC:32334\"; havana_gene \"OTTHUMG00000000960.1_6\"; remap_status \"full_contig\"; remap_num_mappings \"1\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\ttranscript\t34554\t36081\t.\t-\t.\tgene_id \"ENSG00000237613.2_6\"; transcript_id \"ENST00000417324.1_4\"; gene_type \"lncRNA\"; gene_name \"FAM138A\"; transcript_type \"lncRNA\"; transcript_name \"FAM138A-201\"; level \"2\"; transcript_support_level \"1\"; hgnc_id \"HGNC:32334\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000960.1_6\"; havana_transcript \"OTTHUMT00000002842.1_4\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\texon\t35721\t36081\t.\t-\t.\tgene_id \"ENSG00000237613.2_6\"; transcript_id \"ENST00000417324.1_4\"; gene_type \"lncRNA\"; gene_name \"FAM138A\"; transcript_type \"lncRNA\"; transcript_name \"FAM138A-201\"; exon_number \"1\"; exon_id \"ENSE00001656588.1_1\"; level \"2\"; transcript_support_level \"1\"; hgnc_id \"HGNC:32334\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000960.1_6\"; havana_transcript \"OTTHUMT00000002842.1_4\"; remap_original_location \"chr1:-:35721-36081\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t35277\t35481\t.\t-\t.\tgene_id \"ENSG00000237613.2_6\"; transcript_id \"ENST00000417324.1_4\"; gene_type \"lncRNA\"; gene_name \"FAM138A\"; transcript_type \"lncRNA\"; transcript_name \"FAM138A-201\"; exon_number \"2\"; exon_id \"ENSE00001669267.1_1\"; level \"2\"; transcript_support_level \"1\"; hgnc_id \"HGNC:32334\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000960.1_6\"; havana_transcript \"OTTHUMT00000002842.1_4\"; remap_original_location \"chr1:-:35277-35481\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t34554\t35174\t.\t-\t.\tgene_id \"ENSG00000237613.2_6\"; transcript_id \"ENST00000417324.1_4\"; gene_type \"lncRNA\"; gene_name \"FAM138A\"; transcript_type \"lncRNA\"; transcript_name \"FAM138A-201\"; exon_number \"3\"; exon_id \"ENSE00001727627.1_1\"; level \"2\"; transcript_support_level \"1\"; hgnc_id \"HGNC:32334\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000000960.1_6\"; havana_transcript \"OTTHUMT00000002842.1_4\"; remap_original_location \"chr1:-:34554-35174\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\ttranscript\t35245\t36073\t.\t-\t.\tgene_id \"ENSG00000237613.2_6\"; transcript_id \"ENST00000461467.1_3\"; gene_type \"lncRNA\"; gene_name \"FAM138A\"; transcript_type \"lncRNA\"; transcript_name \"FAM138A-202\"; level \"2\"; transcript_support_level \"3\"; hgnc_id \"HGNC:32334\"; havana_gene \"OTTHUMG00000000960.1_6\"; havana_transcript \"OTTHUMT00000002843.1_3\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\texon\t35721\t36073\t.\t-\t.\tgene_id \"ENSG00000237613.2_6\"; transcript_id \"ENST00000461467.1_3\"; gene_type \"lncRNA\"; gene_name \"FAM138A\"; transcript_type \"lncRNA\"; transcript_name \"FAM138A-202\"; exon_number \"1\"; exon_id \"ENSE00001618781.2_1\"; level \"2\"; transcript_support_level \"3\"; hgnc_id \"HGNC:32334\"; havana_gene \"OTTHUMG00000000960.1_6\"; havana_transcript \"OTTHUMT00000002843.1_3\"; remap_original_location \"chr1:-:35721-36073\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t35245\t35481\t.\t-\t.\tgene_id \"ENSG00000237613.2_6\"; transcript_id \"ENST00000461467.1_3\"; gene_type \"lncRNA\"; gene_name \"FAM138A\"; transcript_type \"lncRNA\"; transcript_name \"FAM138A-202\"; exon_number \"2\"; exon_id \"ENSE00001874421.1_1\"; level \"2\"; transcript_support_level \"3\"; hgnc_id \"HGNC:32334\"; havana_gene \"OTTHUMG00000000960.1_6\"; havana_transcript \"OTTHUMT00000002843.1_3\"; remap_original_location \"chr1:-:35245-35481\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\tgene\t52473\t53312\t.\t+\t.\tgene_id \"ENSG00000268020.3_5\"; gene_type \"unprocessed_pseudogene\"; gene_name \"OR4G4P\"; level \"2\"; hgnc_id \"HGNC:14822\"; havana_gene \"OTTHUMG00000185779.1_5\"; remap_status \"full_contig\"; remap_num_mappings \"1\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\ttranscript\t52473\t53312\t.\t+\t.\tgene_id \"ENSG00000268020.3_5\"; transcript_id \"ENST00000606857.1_2\"; gene_type \"unprocessed_pseudogene\"; gene_name \"OR4G4P\"; transcript_type \"unprocessed_pseudogene\"; transcript_name \"OR4G4P-201\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:14822\"; ont \"PGO:0000005\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000185779.1_5\"; havana_transcript \"OTTHUMT00000471235.1_2\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\texon\t52473\t53312\t.\t+\t.\tgene_id \"ENSG00000268020.3_5\"; transcript_id \"ENST00000606857.1_2\"; gene_type \"unprocessed_pseudogene\"; gene_name \"OR4G4P\"; transcript_type \"unprocessed_pseudogene\"; transcript_name \"OR4G4P-201\"; exon_number \"1\"; exon_id \"ENSE00003698237.1_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:14822\"; ont \"PGO:0000005\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000185779.1_5\"; havana_transcript \"OTTHUMT00000471235.1_2\"; remap_original_location \"chr1:+:52473-53312\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\tgene\t57598\t64116\t.\t+\t.\tgene_id \"ENSG00000290826.1_1\"; gene_type \"lncRNA\"; gene_name \"ENSG00000290826\"; level \"2\"; tag \"overlaps_pseudogene\"; remap_status \"full_contig\"; remap_num_mappings \"1\"; remap_target_status \"new\";\n",
      "chr1\tHAVANA\ttranscript\t57598\t64116\t.\t+\t.\tgene_id \"ENSG00000290826.1_1\"; transcript_id \"ENST00000642116.1_2\"; gene_type \"lncRNA\"; gene_name \"ENSG00000290826\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000642116\"; level \"2\"; tag \"RNA_Seq_supported_partial\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_transcript \"OTTHUMT00000492680.1_2\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"new\";\n",
      "chr1\tHAVANA\texon\t57598\t57653\t.\t+\t.\tgene_id \"ENSG00000290826.1_1\"; transcript_id \"ENST00000642116.1_2\"; gene_type \"lncRNA\"; gene_name \"ENSG00000290826\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000642116\"; exon_number \"1\"; exon_id \"ENSE00003812686.1_1\"; level \"2\"; tag \"RNA_Seq_supported_partial\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_transcript \"OTTHUMT00000492680.1_2\"; remap_original_location \"chr1:+:57598-57653\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t58700\t58856\t.\t+\t.\tgene_id \"ENSG00000290826.1_1\"; transcript_id \"ENST00000642116.1_2\"; gene_type \"lncRNA\"; gene_name \"ENSG00000290826\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000642116\"; exon_number \"2\"; exon_id \"ENSE00003812505.1_1\"; level \"2\"; tag \"RNA_Seq_supported_partial\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_transcript \"OTTHUMT00000492680.1_2\"; remap_original_location \"chr1:+:58700-58856\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t62916\t64116\t.\t+\t.\tgene_id \"ENSG00000290826.1_1\"; transcript_id \"ENST00000642116.1_2\"; gene_type \"lncRNA\"; gene_name \"ENSG00000290826\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000642116\"; exon_number \"3\"; exon_id \"ENSE00003811818.1_1\"; level \"2\"; tag \"RNA_Seq_supported_partial\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_transcript \"OTTHUMT00000492680.1_2\"; remap_original_location \"chr1:+:62916-64116\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\tgene\t62949\t63887\t.\t+\t.\tgene_id \"ENSG00000240361.3_7\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"OR4G11P\"; level \"2\"; hgnc_id \"HGNC:31276\"; havana_gene \"OTTHUMG00000001095.3_7\"; remap_status \"full_contig\"; remap_num_mappings \"1\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\ttranscript\t62949\t63887\t.\t+\t.\tgene_id \"ENSG00000240361.3_7\"; transcript_id \"ENST00000492842.2_3\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"OR4G11P\"; transcript_type \"transcribed_unprocessed_pseudogene\"; transcript_name \"OR4G11P-201\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:31276\"; ont \"PGO:0000005\"; ont \"PGO:0000019\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000001095.3_7\"; havana_transcript \"OTTHUMT00000003224.3_3\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\texon\t62949\t63887\t.\t+\t.\tgene_id \"ENSG00000240361.3_7\"; transcript_id \"ENST00000492842.2_3\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"OR4G11P\"; transcript_type \"transcribed_unprocessed_pseudogene\"; transcript_name \"OR4G11P-201\"; exon_number \"1\"; exon_id \"ENSE00001830178.2_1\"; level \"2\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:31276\"; ont \"PGO:0000005\"; ont \"PGO:0000019\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000001095.3_7\"; havana_transcript \"OTTHUMT00000003224.3_3\"; remap_original_location \"chr1:+:62949-63887\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\tgene\t65419\t71585\t.\t+\t.\tgene_id \"ENSG00000186092.7_8\"; gene_type \"protein_coding\"; gene_name \"OR4F5\"; level \"2\"; hgnc_id \"HGNC:14825\"; havana_gene \"OTTHUMG00000001094.4_8\"; remap_status \"full_contig\"; remap_num_mappings \"1\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\ttranscript\t65419\t71585\t.\t+\t.\tgene_id \"ENSG00000186092.7_8\"; transcript_id \"ENST00000641515.2_4\"; gene_type \"protein_coding\"; gene_name \"OR4F5\"; transcript_type \"protein_coding\"; transcript_name \"OR4F5-201\"; level \"2\"; protein_id \"ENSP00000493376.2\"; hgnc_id \"HGNC:14825\"; tag \"RNA_Seq_supported_partial\"; tag \"basic\"; tag \"Ensembl_canonical\"; tag \"MANE_Select\"; tag \"appris_principal_1\"; havana_gene \"OTTHUMG00000001094.4_8\"; havana_transcript \"OTTHUMT00000003223.4_4\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"new\";\n",
      "chr1\tHAVANA\texon\t65419\t65433\t.\t+\t.\tgene_id \"ENSG00000186092.7_8\"; transcript_id \"ENST00000641515.2_4\"; gene_type \"protein_coding\"; gene_name \"OR4F5\"; transcript_type \"protein_coding\"; transcript_name \"OR4F5-201\"; exon_number \"1\"; exon_id \"ENSE00003812156.1_1\"; level \"2\"; protein_id \"ENSP00000493376.2\"; hgnc_id \"HGNC:14825\"; tag \"RNA_Seq_supported_partial\"; tag \"basic\"; tag \"Ensembl_canonical\"; tag \"MANE_Select\"; tag \"appris_principal_1\"; havana_gene \"OTTHUMG00000001094.4_8\"; havana_transcript \"OTTHUMT00000003223.4_4\"; remap_original_location \"chr1:+:65419-65433\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\tUTR\t65419\t65433\t.\t+\t.\tgene_id \"ENSG00000186092.7_8\"; transcript_id \"ENST00000641515.2_4\"; gene_type \"protein_coding\"; gene_name \"OR4F5\"; transcript_type \"protein_coding\"; transcript_name \"OR4F5-201\"; exon_number \"1\"; exon_id \"ENSE00003812156.1\"; level \"2\"; protein_id \"ENSP00000493376.2\"; hgnc_id \"HGNC:14825\"; tag \"RNA_Seq_supported_partial\"; tag \"basic\"; tag \"Ensembl_canonical\"; tag \"MANE_Select\"; tag \"appris_principal_1\"; havana_gene \"OTTHUMG00000001094.4_8\"; havana_transcript \"OTTHUMT00000003223.4_4\"; remap_original_location \"chr1:+:65419-65433\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t65520\t65573\t.\t+\t.\tgene_id \"ENSG00000186092.7_8\"; transcript_id \"ENST00000641515.2_4\"; gene_type \"protein_coding\"; gene_name \"OR4F5\"; transcript_type \"protein_coding\"; transcript_name \"OR4F5-201\"; exon_number \"2\"; exon_id \"ENSE00003813641.1_1\"; level \"2\"; protein_id \"ENSP00000493376.2\"; hgnc_id \"HGNC:14825\"; tag \"RNA_Seq_supported_partial\"; tag \"basic\"; tag \"Ensembl_canonical\"; tag \"MANE_Select\"; tag \"appris_principal_1\"; havana_gene \"OTTHUMG00000001094.4_8\"; havana_transcript \"OTTHUMT00000003223.4_4\"; remap_original_location \"chr1:+:65520-65573\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\tUTR\t65520\t65564\t.\t+\t.\tgene_id \"ENSG00000186092.7_8\"; transcript_id \"ENST00000641515.2_4\"; gene_type \"protein_coding\"; gene_name \"OR4F5\"; transcript_type \"protein_coding\"; transcript_name \"OR4F5-201\"; exon_number \"2\"; exon_id \"ENSE00003813641.1\"; level \"2\"; protein_id \"ENSP00000493376.2\"; hgnc_id \"HGNC:14825\"; tag \"RNA_Seq_supported_partial\"; tag \"basic\"; tag \"Ensembl_canonical\"; tag \"MANE_Select\"; tag \"appris_principal_1\"; havana_gene \"OTTHUMG00000001094.4_8\"; havana_transcript \"OTTHUMT00000003223.4_4\"; remap_original_location \"chr1:+:65520-65564\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\tCDS\t65565\t65573\t.\t+\t0\tgene_id \"ENSG00000186092.7_8\"; transcript_id \"ENST00000641515.2_4\"; gene_type \"protein_coding\"; gene_name \"OR4F5\"; transcript_type \"protein_coding\"; transcript_name \"OR4F5-201\"; exon_number \"2\"; exon_id \"ENSE00003813641.1\"; level \"2\"; protein_id \"ENSP00000493376.2\"; hgnc_id \"HGNC:14825\"; tag \"RNA_Seq_supported_partial\"; tag \"basic\"; tag \"Ensembl_canonical\"; tag \"MANE_Select\"; tag \"appris_principal_1\"; havana_gene \"OTTHUMG00000001094.4_8\"; havana_transcript \"OTTHUMT00000003223.4_4\"; remap_original_location \"chr1:+:65565-65573\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\tstart_codon\t65565\t65567\t.\t+\t0\tgene_id \"ENSG00000186092.7_8\"; transcript_id \"ENST00000641515.2_4\"; gene_type \"protein_coding\"; gene_name \"OR4F5\"; transcript_type \"protein_coding\"; transcript_name \"OR4F5-201\"; exon_number \"2\"; exon_id \"ENSE00003813641.1\"; level \"2\"; protein_id \"ENSP00000493376.2\"; hgnc_id \"HGNC:14825\"; tag \"RNA_Seq_supported_partial\"; tag \"basic\"; tag \"Ensembl_canonical\"; tag \"MANE_Select\"; tag \"appris_principal_1\"; havana_gene \"OTTHUMG00000001094.4_8\"; havana_transcript \"OTTHUMT00000003223.4_4\"; remap_original_location \"chr1:+:65565-65567\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t69037\t71585\t.\t+\t.\tgene_id \"ENSG00000186092.7_8\"; transcript_id \"ENST00000641515.2_4\"; gene_type \"protein_coding\"; gene_name \"OR4F5\"; transcript_type \"protein_coding\"; transcript_name \"OR4F5-201\"; exon_number \"3\"; exon_id \"ENSE00003813949.1_1\"; level \"2\"; protein_id \"ENSP00000493376.2\"; hgnc_id \"HGNC:14825\"; tag \"RNA_Seq_supported_partial\"; tag \"basic\"; tag \"Ensembl_canonical\"; tag \"MANE_Select\"; tag \"appris_principal_1\"; havana_gene \"OTTHUMG00000001094.4_8\"; havana_transcript \"OTTHUMT00000003223.4_4\"; remap_original_location \"chr1:+:69037-71585\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\tCDS\t69037\t70005\t.\t+\t0\tgene_id \"ENSG00000186092.7_8\"; transcript_id \"ENST00000641515.2_4\"; gene_type \"protein_coding\"; gene_name \"OR4F5\"; transcript_type \"protein_coding\"; transcript_name \"OR4F5-201\"; exon_number \"3\"; exon_id \"ENSE00003813949.1\"; level \"2\"; protein_id \"ENSP00000493376.2\"; hgnc_id \"HGNC:14825\"; tag \"RNA_Seq_supported_partial\"; tag \"basic\"; tag \"Ensembl_canonical\"; tag \"MANE_Select\"; tag \"appris_principal_1\"; havana_gene \"OTTHUMG00000001094.4_8\"; havana_transcript \"OTTHUMT00000003223.4_4\"; remap_original_location \"chr1:+:69037-70005\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\tUTR\t70006\t71585\t.\t+\t.\tgene_id \"ENSG00000186092.7_8\"; transcript_id \"ENST00000641515.2_4\"; gene_type \"protein_coding\"; gene_name \"OR4F5\"; transcript_type \"protein_coding\"; transcript_name \"OR4F5-201\"; exon_number \"3\"; exon_id \"ENSE00003813949.1\"; level \"2\"; protein_id \"ENSP00000493376.2\"; hgnc_id \"HGNC:14825\"; tag \"RNA_Seq_supported_partial\"; tag \"basic\"; tag \"Ensembl_canonical\"; tag \"MANE_Select\"; tag \"appris_principal_1\"; havana_gene \"OTTHUMG00000001094.4_8\"; havana_transcript \"OTTHUMT00000003223.4_4\"; remap_original_location \"chr1:+:70006-71585\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\tstop_codon\t70006\t70008\t.\t+\t0\tgene_id \"ENSG00000186092.7_8\"; transcript_id \"ENST00000641515.2_4\"; gene_type \"protein_coding\"; gene_name \"OR4F5\"; transcript_type \"protein_coding\"; transcript_name \"OR4F5-201\"; exon_number \"3\"; exon_id \"ENSE00003813949.1\"; level \"2\"; protein_id \"ENSP00000493376.2\"; hgnc_id \"HGNC:14825\"; tag \"RNA_Seq_supported_partial\"; tag \"basic\"; tag \"Ensembl_canonical\"; tag \"MANE_Select\"; tag \"appris_principal_1\"; havana_gene \"OTTHUMG00000001094.4_8\"; havana_transcript \"OTTHUMT00000003223.4_4\"; remap_original_location \"chr1:+:70006-70008\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\tgene\t89295\t133723\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; level \"2\"; tag \"overlapping_locus\"; havana_gene \"OTTHUMG00000001096.2_14\"; remap_status \"full_contig\"; remap_num_mappings \"1\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\ttranscript\t89295\t120932\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000466430.5_3\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000466430\"; level \"2\"; transcript_support_level \"5\"; tag \"not_best_in_genome_evidence\"; tag \"basic\"; havana_gene \"OTTHUMG00000001096.2_14\"; havana_transcript \"OTTHUMT00000003225.1_3\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\texon\t120775\t120932\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000466430.5_3\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000466430\"; exon_number \"1\"; exon_id \"ENSE00001606755.2_1\"; level \"2\"; transcript_support_level \"5\"; tag \"not_best_in_genome_evidence\"; tag \"basic\"; havana_gene \"OTTHUMG00000001096.2_14\"; havana_transcript \"OTTHUMT00000003225.1_3\"; remap_original_location \"chr1:-:120775-120932\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t112700\t112804\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000466430.5_3\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000466430\"; exon_number \"2\"; exon_id \"ENSE00001957285.1_1\"; level \"2\"; transcript_support_level \"5\"; tag \"not_best_in_genome_evidence\"; tag \"basic\"; havana_gene \"OTTHUMG00000001096.2_14\"; havana_transcript \"OTTHUMT00000003225.1_3\"; remap_original_location \"chr1:-:112700-112804\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t92091\t92240\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000466430.5_3\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000466430\"; exon_number \"3\"; exon_id \"ENSE00001944529.1_1\"; level \"2\"; transcript_support_level \"5\"; tag \"not_best_in_genome_evidence\"; tag \"basic\"; havana_gene \"OTTHUMG00000001096.2_14\"; havana_transcript \"OTTHUMT00000003225.1_3\"; remap_original_location \"chr1:-:92091-92240\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t89295\t91629\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000466430.5_3\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000466430\"; exon_number \"4\"; exon_id \"ENSE00001846804.1_1\"; level \"2\"; transcript_support_level \"5\"; tag \"not_best_in_genome_evidence\"; tag \"basic\"; havana_gene \"OTTHUMG00000001096.2_14\"; havana_transcript \"OTTHUMT00000003225.1_3\"; remap_original_location \"chr1:-:89295-91629\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\ttranscript\t92230\t129217\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000477740.5_4\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000477740\"; level \"2\"; transcript_support_level \"5\"; tag \"not_best_in_genome_evidence\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000001096.2_14\"; havana_transcript \"OTTHUMT00000003688.1_4\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\texon\t129055\t129217\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000477740.5_4\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000477740\"; exon_number \"1\"; exon_id \"ENSE00001919246.1_1\"; level \"2\"; transcript_support_level \"5\"; tag \"not_best_in_genome_evidence\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000001096.2_14\"; havana_transcript \"OTTHUMT00000003688.1_4\"; remap_original_location \"chr1:-:129055-129217\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t120721\t120932\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000477740.5_4\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000477740\"; exon_number \"2\"; exon_id \"ENSE00001171005.3_1\"; level \"2\"; transcript_support_level \"5\"; tag \"not_best_in_genome_evidence\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000001096.2_14\"; havana_transcript \"OTTHUMT00000003688.1_4\"; remap_original_location \"chr1:-:120721-120932\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t112700\t112804\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000477740.5_4\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000477740\"; exon_number \"3\"; exon_id \"ENSE00001957285.1_1\"; level \"2\"; transcript_support_level \"5\"; tag \"not_best_in_genome_evidence\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000001096.2_14\"; havana_transcript \"OTTHUMT00000003688.1_4\"; remap_original_location \"chr1:-:112700-112804\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t92230\t92240\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000477740.5_4\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000477740\"; exon_number \"4\"; exon_id \"ENSE00001896976.1_1\"; level \"2\"; transcript_support_level \"5\"; tag \"not_best_in_genome_evidence\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000001096.2_14\"; havana_transcript \"OTTHUMT00000003688.1_4\"; remap_original_location \"chr1:-:92230-92240\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\ttranscript\t110953\t129173\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000471248.1_7\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000471248\"; level \"2\"; transcript_support_level \"5\"; tag \"not_best_in_genome_evidence\"; tag \"dotter_confirmed\"; havana_gene \"OTTHUMG00000001096.2_14\"; havana_transcript \"OTTHUMT00000003687.1_7\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\texon\t129055\t129173\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000471248.1_7\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000471248\"; exon_number \"1\"; exon_id \"ENSE00001934975.1_1\"; level \"2\"; transcript_support_level \"5\"; tag \"not_best_in_genome_evidence\"; tag \"dotter_confirmed\"; havana_gene \"OTTHUMG00000001096.2_14\"; havana_transcript \"OTTHUMT00000003687.1_7\"; remap_original_location \"chr1:-:129055-129173\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t112700\t112804\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000471248.1_7\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000471248\"; exon_number \"2\"; exon_id \"ENSE00001957285.1_1\"; level \"2\"; transcript_support_level \"5\"; tag \"not_best_in_genome_evidence\"; tag \"dotter_confirmed\"; havana_gene \"OTTHUMG00000001096.2_14\"; havana_transcript \"OTTHUMT00000003687.1_7\"; remap_original_location \"chr1:-:112700-112804\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t110953\t111357\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000471248.1_7\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000471248\"; exon_number \"3\"; exon_id \"ENSE00001879696.1_1\"; level \"2\"; transcript_support_level \"5\"; tag \"not_best_in_genome_evidence\"; tag \"dotter_confirmed\"; havana_gene \"OTTHUMG00000001096.2_14\"; havana_transcript \"OTTHUMT00000003687.1_7\"; remap_original_location \"chr1:-:110953-111357\"; remap_status \"full_contig\";\n",
      "chr1\tENSEMBL\ttranscript\t120725\t133723\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000610542.1_3\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000610542\"; level \"3\"; transcript_support_level \"5\"; tag \"basic\"; havana_gene \"OTTHUMG00000001096.2_14\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"new\";\n",
      "chr1\tENSEMBL\texon\t133374\t133723\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000610542.1_3\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000610542\"; exon_number \"1\"; exon_id \"ENSE00003748456.1_1\"; level \"3\"; transcript_support_level \"5\"; tag \"basic\"; havana_gene \"OTTHUMG00000001096.2_14\"; remap_original_location \"chr1:-:133374-133723\"; remap_status \"full_contig\";\n",
      "chr1\tENSEMBL\texon\t129055\t129223\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000610542.1_3\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000610542\"; exon_number \"2\"; exon_id \"ENSE00003734824.1_1\"; level \"3\"; transcript_support_level \"5\"; tag \"basic\"; havana_gene \"OTTHUMG00000001096.2_14\"; remap_original_location \"chr1:-:129055-129223\"; remap_status \"full_contig\";\n",
      "chr1\tENSEMBL\texon\t120725\t120932\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000610542.1_3\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000610542\"; exon_number \"3\"; exon_id \"ENSE00003749543.1_1\"; level \"3\"; transcript_support_level \"5\"; tag \"basic\"; havana_gene \"OTTHUMG00000001096.2_14\"; remap_original_location \"chr1:-:120725-120869\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\ttranscript\t129081\t133566\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000453576.2_3\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000453576\"; level \"2\"; transcript_support_level \"2\"; havana_gene \"OTTHUMG00000001096.2_14\"; havana_transcript \"OTTHUMT00000003689.1_3\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\texon\t133374\t133566\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000453576.2_3\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000453576\"; exon_number \"1\"; exon_id \"ENSE00001737600.2_1\"; level \"2\"; transcript_support_level \"2\"; havana_gene \"OTTHUMG00000001096.2_14\"; havana_transcript \"OTTHUMT00000003689.1_3\"; remap_original_location \"chr1:-:133374-133566\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t129081\t129223\t.\t-\t.\tgene_id \"ENSG00000238009.6_14\"; transcript_id \"ENST00000453576.2_3\"; gene_type \"lncRNA\"; gene_name \"ENSG00000238009\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000453576\"; exon_number \"2\"; exon_id \"ENSE00001827073.1_1\"; level \"2\"; transcript_support_level \"2\"; havana_gene \"OTTHUMG00000001096.2_14\"; havana_transcript \"OTTHUMT00000003689.1_3\"; remap_original_location \"chr1:-:129081-129223\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\tgene\t89551\t91105\t.\t-\t.\tgene_id \"ENSG00000239945.1_12\"; gene_type \"lncRNA\"; gene_name \"ENSG00000239945\"; level \"2\"; tag \"overlapping_locus\"; havana_gene \"OTTHUMG00000001097.2_12\"; remap_status \"full_contig\"; remap_num_mappings \"1\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\ttranscript\t89551\t91105\t.\t-\t.\tgene_id \"ENSG00000239945.1_12\"; transcript_id \"ENST00000495576.1_4\"; gene_type \"lncRNA\"; gene_name \"ENSG00000239945\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000495576\"; level \"2\"; transcript_support_level \"5\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000001097.2_12\"; havana_transcript \"OTTHUMT00000003226.2_4\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\texon\t90287\t91105\t.\t-\t.\tgene_id \"ENSG00000239945.1_12\"; transcript_id \"ENST00000495576.1_4\"; gene_type \"lncRNA\"; gene_name \"ENSG00000239945\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000495576\"; exon_number \"1\"; exon_id \"ENSE00001907785.1_1\"; level \"2\"; transcript_support_level \"5\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000001097.2_12\"; havana_transcript \"OTTHUMT00000003226.2_4\"; remap_original_location \"chr1:-:90287-91105\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t89551\t90050\t.\t-\t.\tgene_id \"ENSG00000239945.1_12\"; transcript_id \"ENST00000495576.1_4\"; gene_type \"lncRNA\"; gene_name \"ENSG00000239945\"; transcript_type \"lncRNA\"; transcript_name \"ENST00000495576\"; exon_number \"2\"; exon_id \"ENSE00001927725.1_1\"; level \"2\"; transcript_support_level \"5\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000001097.2_12\"; havana_transcript \"OTTHUMT00000003226.2_4\"; remap_original_location \"chr1:-:89551-90050\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\tgene\t131025\t134836\t.\t+\t.\tgene_id \"ENSG00000233750.3_7\"; gene_type \"processed_pseudogene\"; gene_name \"CICP27\"; level \"1\"; hgnc_id \"HGNC:48835\"; tag \"pseudo_consens\"; havana_gene \"OTTHUMG00000001257.3_7\"; remap_status \"full_contig\"; remap_num_mappings \"1\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\ttranscript\t131025\t134836\t.\t+\t.\tgene_id \"ENSG00000233750.3_7\"; transcript_id \"ENST00000442987.3_2\"; gene_type \"processed_pseudogene\"; gene_name \"CICP27\"; transcript_type \"processed_pseudogene\"; transcript_name \"CICP27-201\"; level \"1\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:48835\"; ont \"PGO:0000004\"; tag \"pseudo_consens\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000001257.3_7\"; havana_transcript \"OTTHUMT00000003691.3_2\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\texon\t131025\t134836\t.\t+\t.\tgene_id \"ENSG00000233750.3_7\"; transcript_id \"ENST00000442987.3_2\"; gene_type \"processed_pseudogene\"; gene_name \"CICP27\"; transcript_type \"processed_pseudogene\"; transcript_name \"CICP27-201\"; exon_number \"1\"; exon_id \"ENSE00001625118.3_1\"; level \"1\"; transcript_support_level \"NA\"; hgnc_id \"HGNC:48835\"; ont \"PGO:0000004\"; tag \"pseudo_consens\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000001257.3_7\"; havana_transcript \"OTTHUMT00000003691.3_2\"; remap_original_location \"chr1:+:131025-134836\"; remap_status \"full_contig\";\n",
      "chr1\tENSEMBL\tgene\t134901\t139379\t.\t-\t.\tgene_id \"ENSG00000237683.5\"; gene_type \"protein_coding\"; gene_status \"KNOWN\"; gene_name \"AL627309.1\"; level \"3\"; remap_substituted_missing_target \"V19\";\n",
      "chr1\tENSEMBL\ttranscript\t134901\t139379\t.\t-\t.\tgene_id \"ENSG00000237683.5\"; transcript_id \"ENST00000423372.3\"; gene_type \"protein_coding\"; gene_status \"KNOWN\"; gene_name \"AL627309.1\"; transcript_type \"protein_coding\"; transcript_status \"KNOWN\"; transcript_name \"AL627309.1-201\"; level \"3\"; protein_id \"ENSP00000473460.1\"; tag \"basic\"; tag \"appris_principal\"; remap_substituted_missing_target \"V19\";\n",
      "chr1\tENSEMBL\texon\t137621\t139379\t.\t-\t.\tgene_id \"ENSG00000237683.5\"; transcript_id \"ENST00000423372.3\"; gene_type \"protein_coding\"; gene_status \"KNOWN\"; gene_name \"AL627309.1\"; transcript_type \"protein_coding\"; transcript_status \"KNOWN\"; transcript_name \"AL627309.1-201\"; exon_number \"1\"; exon_id \"ENSE00002221580.1\"; level \"3\"; protein_id \"ENSP00000473460.1\"; tag \"basic\"; tag \"appris_principal\"; remap_substituted_missing_target \"V19\";\n",
      "chr1\tENSEMBL\tUTR\t139310\t139379\t.\t-\t.\tgene_id \"ENSG00000237683.5\"; transcript_id \"ENST00000423372.3\"; gene_type \"protein_coding\"; gene_status \"KNOWN\"; gene_name \"AL627309.1\"; transcript_type \"protein_coding\"; transcript_status \"KNOWN\"; transcript_name \"AL627309.1-201\"; level \"3\"; protein_id \"ENSP00000473460.1\"; tag \"basic\"; tag \"appris_principal\"; remap_substituted_missing_target \"V19\";\n",
      "chr1\tENSEMBL\tCDS\t138533\t139309\t.\t-\t0\tgene_id \"ENSG00000237683.5\"; transcript_id \"ENST00000423372.3\"; gene_type \"protein_coding\"; gene_status \"KNOWN\"; gene_name \"AL627309.1\"; transcript_type \"protein_coding\"; transcript_status \"KNOWN\"; transcript_name \"AL627309.1-201\"; exon_number \"1\"; exon_id \"ENSE00002221580.1\"; level \"3\"; protein_id \"ENSP00000473460.1\"; tag \"basic\"; tag \"appris_principal\"; remap_substituted_missing_target \"V19\";\n",
      "chr1\tENSEMBL\tUTR\t137621\t138532\t.\t-\t.\tgene_id \"ENSG00000237683.5\"; transcript_id \"ENST00000423372.3\"; gene_type \"protein_coding\"; gene_status \"KNOWN\"; gene_name \"AL627309.1\"; transcript_type \"protein_coding\"; transcript_status \"KNOWN\"; transcript_name \"AL627309.1-201\"; level \"3\"; protein_id \"ENSP00000473460.1\"; tag \"basic\"; tag \"appris_principal\"; remap_substituted_missing_target \"V19\";\n",
      "chr1\tENSEMBL\tstop_codon\t138530\t138532\t.\t-\t0\tgene_id \"ENSG00000237683.5\"; transcript_id \"ENST00000423372.3\"; gene_type \"protein_coding\"; gene_status \"KNOWN\"; gene_name \"AL627309.1\"; transcript_type \"protein_coding\"; transcript_status \"KNOWN\"; transcript_name \"AL627309.1-201\"; exon_number \"1\"; exon_id \"ENSE00002221580.1\"; level \"3\"; protein_id \"ENSP00000473460.1\"; tag \"basic\"; tag \"appris_principal\"; remap_substituted_missing_target \"V19\";\n",
      "chr1\tENSEMBL\texon\t134901\t135802\t.\t-\t.\tgene_id \"ENSG00000237683.5\"; transcript_id \"ENST00000423372.3\"; gene_type \"protein_coding\"; gene_status \"KNOWN\"; gene_name \"AL627309.1\"; transcript_type \"protein_coding\"; transcript_status \"KNOWN\"; transcript_name \"AL627309.1-201\"; exon_number \"2\"; exon_id \"ENSE00002314092.1\"; level \"3\"; protein_id \"ENSP00000473460.1\"; tag \"basic\"; tag \"appris_principal\"; remap_substituted_missing_target \"V19\";\n",
      "chr1\tENSEMBL\tUTR\t134901\t135802\t.\t-\t.\tgene_id \"ENSG00000237683.5\"; transcript_id \"ENST00000423372.3\"; gene_type \"protein_coding\"; gene_status \"KNOWN\"; gene_name \"AL627309.1\"; transcript_type \"protein_coding\"; transcript_status \"KNOWN\"; transcript_name \"AL627309.1-201\"; level \"3\"; protein_id \"ENSP00000473460.1\"; tag \"basic\"; tag \"appris_principal\"; remap_substituted_missing_target \"V19\";\n",
      "chr1\tHAVANA\tgene\t135141\t135895\t.\t-\t.\tgene_id \"ENSG00000268903.1_9\"; gene_type \"processed_pseudogene\"; gene_name \"ENSG00000268903\"; level \"2\"; havana_gene \"OTTHUMG00000182518.2_9\"; remap_status \"full_contig\"; remap_num_mappings \"1\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\ttranscript\t135141\t135895\t.\t-\t.\tgene_id \"ENSG00000268903.1_9\"; transcript_id \"ENST00000494149.2_3\"; gene_type \"processed_pseudogene\"; gene_name \"ENSG00000268903\"; transcript_type \"processed_pseudogene\"; transcript_name \"ENST00000494149\"; level \"2\"; transcript_support_level \"NA\"; ont \"PGO:0000004\"; tag \"basic\"; tag \"Ensembl_canonical\"; havana_gene \"OTTHUMG00000182518.2_9\"; havana_transcript \"OTTHUMT00000461982.2_3\"; remap_num_mappings \"1\"; remap_status \"full_contig\"; remap_target_status \"overlap\";\n"
     ]
    }
   ],
   "source": [
    "i = 0\n",
    "ts = db.all_features()\n",
    "while i < 100:\n",
    "    t = next(ts)\n",
    "    # t = next(db.featuretypes())\n",
    "    print(t)\n",
    "\n",
    "    i += 1\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n"
     ]
    }
   ],
   "source": [
    "introns = list(introns)\n",
    "print(introns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 0\n",
    "while i < 5:\n",
    "    t = next(db_intron.features_of_type('transcript'))\n",
    "    print(t.id)\n",
    "    i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>SymbolSource</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ENST</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>EXON</th>\n",
       "      <th>...</th>\n",
       "      <th>DP_DG</th>\n",
       "      <th>DP_DL</th>\n",
       "      <th>MaxSpliceAI</th>\n",
       "      <th>is_Canonical</th>\n",
       "      <th>Int_loc</th>\n",
       "      <th>ex_up_dist</th>\n",
       "      <th>ex_down_dist</th>\n",
       "      <th>exon_pos</th>\n",
       "      <th>exon_splice_site</th>\n",
       "      <th>SpliceType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>879375</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>28706</td>\n",
       "      <td>ENST00000342066</td>\n",
       "      <td>c.1888C&gt;T</td>\n",
       "      <td>14/14</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1307</td>\n",
       "      <td>0.03</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>579</td>\n",
       "      <td>87</td>\n",
       "      <td>579</td>\n",
       "      <td>non_SplExon</td>\n",
       "      <td>Acceptor_ex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>949363</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>4053</td>\n",
       "      <td>ENST00000379389</td>\n",
       "      <td>c.4-1G&gt;A</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>288</td>\n",
       "      <td>-88</td>\n",
       "      <td>0.99</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>None</td>\n",
       "      <td>ENST_not_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>949443</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>4053</td>\n",
       "      <td>ENST00000379389</td>\n",
       "      <td>c.83T&gt;A</td>\n",
       "      <td>2/2</td>\n",
       "      <td>...</td>\n",
       "      <td>-294</td>\n",
       "      <td>208</td>\n",
       "      <td>0.03</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>None</td>\n",
       "      <td>ENST_not_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>949523</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>4053</td>\n",
       "      <td>ENST00000379389</td>\n",
       "      <td>c.163C&gt;T</td>\n",
       "      <td>2/2</td>\n",
       "      <td>...</td>\n",
       "      <td>167</td>\n",
       "      <td>128</td>\n",
       "      <td>0.01</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>None</td>\n",
       "      <td>ENST_not_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>949643</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>4053</td>\n",
       "      <td>ENST00000379389</td>\n",
       "      <td>c.285del</td>\n",
       "      <td>2/2</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>43</td>\n",
       "      <td>0.02</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>ENST_not_match</td>\n",
       "      <td>None</td>\n",
       "      <td>ENST_not_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20696</th>\n",
       "      <td>1</td>\n",
       "      <td>247597502</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NLRP3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>16400</td>\n",
       "      <td>ENST00000336119</td>\n",
       "      <td>c.2425G&gt;A</td>\n",
       "      <td>5/9</td>\n",
       "      <td>...</td>\n",
       "      <td>1940</td>\n",
       "      <td>-180</td>\n",
       "      <td>0.01</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>73</td>\n",
       "      <td>97</td>\n",
       "      <td>73</td>\n",
       "      <td>non_SplExon</td>\n",
       "      <td>Donor_ex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20697</th>\n",
       "      <td>1</td>\n",
       "      <td>247599354</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NLRP3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>16400</td>\n",
       "      <td>ENST00000336119</td>\n",
       "      <td>c.2581T&gt;C</td>\n",
       "      <td>6/9</td>\n",
       "      <td>...</td>\n",
       "      <td>88</td>\n",
       "      <td>1608</td>\n",
       "      <td>0.02</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>88</td>\n",
       "      <td>82</td>\n",
       "      <td>88</td>\n",
       "      <td>non_SplExon</td>\n",
       "      <td>Acceptor_ex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20698</th>\n",
       "      <td>1</td>\n",
       "      <td>247599355</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NLRP3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>16400</td>\n",
       "      <td>ENST00000336119</td>\n",
       "      <td>c.2582A&gt;G</td>\n",
       "      <td>6/9</td>\n",
       "      <td>...</td>\n",
       "      <td>-2201</td>\n",
       "      <td>87</td>\n",
       "      <td>0.14</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>87</td>\n",
       "      <td>83</td>\n",
       "      <td>87</td>\n",
       "      <td>non_SplExon</td>\n",
       "      <td>Acceptor_ex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20699</th>\n",
       "      <td>1</td>\n",
       "      <td>247607362</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NLRP3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>16400</td>\n",
       "      <td>ENST00000336119</td>\n",
       "      <td>c.2758C&gt;T</td>\n",
       "      <td>7/9</td>\n",
       "      <td>...</td>\n",
       "      <td>-89</td>\n",
       "      <td>82</td>\n",
       "      <td>0.18</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>82</td>\n",
       "      <td>88</td>\n",
       "      <td>82</td>\n",
       "      <td>non_SplExon</td>\n",
       "      <td>Donor_ex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20700</th>\n",
       "      <td>1</td>\n",
       "      <td>247607363</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NLRP3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>16400</td>\n",
       "      <td>ENST00000336119</td>\n",
       "      <td>c.2759G&gt;A</td>\n",
       "      <td>7/9</td>\n",
       "      <td>...</td>\n",
       "      <td>-90</td>\n",
       "      <td>81</td>\n",
       "      <td>0.10</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>81</td>\n",
       "      <td>89</td>\n",
       "      <td>81</td>\n",
       "      <td>non_SplExon</td>\n",
       "      <td>Donor_ex</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20687 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM        POS REF ALT GeneSymbol SymbolSource HGNC_ID  \\\n",
       "0         1     879375   C   T     SAMD11         HGNC   28706   \n",
       "1         1     949363   G   A      ISG15         HGNC    4053   \n",
       "2         1     949443   T   A      ISG15         HGNC    4053   \n",
       "3         1     949523   C   T      ISG15         HGNC    4053   \n",
       "4         1     949643  AC   A      ISG15         HGNC    4053   \n",
       "...     ...        ...  ..  ..        ...          ...     ...   \n",
       "20696     1  247597502   G   A      NLRP3         HGNC   16400   \n",
       "20697     1  247599354   T   C      NLRP3         HGNC   16400   \n",
       "20698     1  247599355   A   G      NLRP3         HGNC   16400   \n",
       "20699     1  247607362   C   T      NLRP3         HGNC   16400   \n",
       "20700     1  247607363   G   A      NLRP3         HGNC   16400   \n",
       "\n",
       "                  ENST      HGVSc   EXON  ...  DP_DG  DP_DL MaxSpliceAI  \\\n",
       "0      ENST00000342066  c.1888C>T  14/14  ...     -2  -1307        0.03   \n",
       "1      ENST00000379389   c.4-1G>A         ...    288    -88        0.99   \n",
       "2      ENST00000379389    c.83T>A    2/2  ...   -294    208        0.03   \n",
       "3      ENST00000379389   c.163C>T    2/2  ...    167    128        0.01   \n",
       "4      ENST00000379389   c.285del    2/2  ...      8     43        0.02   \n",
       "...                ...        ...    ...  ...    ...    ...         ...   \n",
       "20696  ENST00000336119  c.2425G>A    5/9  ...   1940   -180        0.01   \n",
       "20697  ENST00000336119  c.2581T>C    6/9  ...     88   1608        0.02   \n",
       "20698  ENST00000336119  c.2582A>G    6/9  ...  -2201     87        0.14   \n",
       "20699  ENST00000336119  c.2758C>T    7/9  ...    -89     82        0.18   \n",
       "20700  ENST00000336119  c.2759G>A    7/9  ...    -90     81        0.10   \n",
       "\n",
       "      is_Canonical Int_loc      ex_up_dist    ex_down_dist        exon_pos  \\\n",
       "0            False  Exonic             579              87             579   \n",
       "1             True      -1  ENST_not_match  ENST_not_match  ENST_not_match   \n",
       "2            False  Exonic  ENST_not_match  ENST_not_match  ENST_not_match   \n",
       "3            False  Exonic  ENST_not_match  ENST_not_match  ENST_not_match   \n",
       "4            False  Exonic  ENST_not_match  ENST_not_match  ENST_not_match   \n",
       "...            ...     ...             ...             ...             ...   \n",
       "20696        False  Exonic              73              97              73   \n",
       "20697        False  Exonic              88              82              88   \n",
       "20698        False  Exonic              87              83              87   \n",
       "20699        False  Exonic              82              88              82   \n",
       "20700        False  Exonic              81              89              81   \n",
       "\n",
       "      exon_splice_site      SpliceType  \n",
       "0          non_SplExon     Acceptor_ex  \n",
       "1                 None  ENST_not_match  \n",
       "2                 None  ENST_not_match  \n",
       "3                 None  ENST_not_match  \n",
       "4                 None  ENST_not_match  \n",
       "...                ...             ...  \n",
       "20696      non_SplExon        Donor_ex  \n",
       "20697      non_SplExon     Acceptor_ex  \n",
       "20698      non_SplExon     Acceptor_ex  \n",
       "20699      non_SplExon        Donor_ex  \n",
       "20700      non_SplExon        Donor_ex  \n",
       "\n",
       "[20687 rows x 28 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "df['Num_ExInt'] = df.progress_apply(\n",
    "    posparser.calc_ex_int_num, db=db, db_intron=db_intron, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SpliceType\n",
       "Acceptor_ex       9264\n",
       "Donor_ex          8347\n",
       "ENST_not_match    2938\n",
       "Donor_int           49\n",
       "Acceptor_int        45\n",
       "Center_of_Exon      44\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['SpliceType'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: 'ENST_not_match'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRemoteTraceback\u001b[0m                           Traceback (most recent call last)",
      "\u001b[0;31mRemoteTraceback\u001b[0m: \n\"\"\"\nTraceback (most recent call last):\n  File \"/Users/utsu/miniforge3/envs/wesanno/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n    result = (True, func(*args, **kwds))\n  File \"/Users/utsu/miniforge3/envs/wesanno/lib/python3.9/multiprocessing/pool.py\", line 51, in starmapstar\n    return list(itertools.starmap(args[0], args[1]))\n  File \"/Users/utsu/miniforge3/envs/wesanno/lib/python3.9/site-packages/pandarallel/core.py\", line 158, in __call__\n    results = self.work_function(\n  File \"/Users/utsu/miniforge3/envs/wesanno/lib/python3.9/site-packages/pandarallel/data_types/dataframe.py\", line 32, in work\n    return data.apply(\n  File \"/Users/utsu/miniforge3/envs/wesanno/lib/python3.9/site-packages/pandas/core/frame.py\", line 10374, in apply\n    return op.apply().__finalize__(self, method=\"apply\")\n  File \"/Users/utsu/miniforge3/envs/wesanno/lib/python3.9/site-packages/pandas/core/apply.py\", line 916, in apply\n    return self.apply_standard()\n  File \"/Users/utsu/miniforge3/envs/wesanno/lib/python3.9/site-packages/pandas/core/apply.py\", line 1063, in apply_standard\n    results, res_index = self.apply_series_generator()\n  File \"/Users/utsu/miniforge3/envs/wesanno/lib/python3.9/site-packages/pandas/core/apply.py\", line 1081, in apply_series_generator\n    results[i] = self.func(v, *self.args, **self.kwargs)\n  File \"/Volumes/vol/work/Github/dev/ValidationData/TP/additional/../../../libs/posparser.py\", line 214, in extract_splicing_region\n    if int(row['ex_down_dist']) == 0:\nValueError: invalid literal for int() with base 10: 'ENST_not_match'\n\"\"\"",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m#2-3. Decision exonic splice sites (1 nt in acceptor site or 3 nts on Donor site)\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mexon_splice_site\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43mdf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparallel_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mposparser\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mextract_splicing_region\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/wesanno/lib/python3.9/site-packages/pandarallel/core.py:444\u001b[0m, in \u001b[0;36mparallelize_with_pipe.<locals>.closure\u001b[0;34m(data, user_defined_function, *user_defined_function_args, **user_defined_function_kwargs)\u001b[0m\n\u001b[1;32m    441\u001b[0m     \u001b[38;5;28;01melif\u001b[39;00m worker_status \u001b[38;5;241m==\u001b[39m WorkerStatus\u001b[38;5;241m.\u001b[39mError:\n\u001b[1;32m    442\u001b[0m         progress_bars\u001b[38;5;241m.\u001b[39mset_error(worker_index)\n\u001b[0;32m--> 444\u001b[0m results \u001b[38;5;241m=\u001b[39m \u001b[43mresults_promise\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    446\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m data_type\u001b[38;5;241m.\u001b[39mreduce(results, reduce_extra)\n",
      "File \u001b[0;32m~/miniforge3/envs/wesanno/lib/python3.9/multiprocessing/pool.py:771\u001b[0m, in \u001b[0;36mApplyResult.get\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    769\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_value\n\u001b[1;32m    770\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 771\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_value\n",
      "\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: 'ENST_not_match'"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "tbx = tbx_anno.fetch(\"chr1\", 949442, 949443, parser=pysam.asGFF3())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\tHAVANA\ttranscript\t936525\t949815\t.\t+\t.\tID=ENST00000624652.1;Parent=ENSG00000187608.10;gene_id=ENSG00000187608.10_10;transcript_id=ENST00000624652.1_4;gene_type=protein_coding;gene_name=ISG15;transcript_type=protein_coding;transcript_name=ISG15-201;level=2;protein_id=ENSP00000485313.1;transcript_support_level=3;hgnc_id=HGNC:4053;tag=mRNA_end_NF,cds_end_NF;havana_gene=OTTHUMG00000040777.4_10;havana_transcript=OTTHUMT00000479385.2_4;remap_num_mappings=1;remap_status=full_contig;remap_target_status=new\n",
      "ENST00000624652\n",
      "chr1\tHAVANA\ttranscript\t948877\t949920\t.\t+\t.\tID=ENST00000649529.1;Parent=ENSG00000187608.10;gene_id=ENSG00000187608.10_10;transcript_id=ENST00000649529.1_4;gene_type=protein_coding;gene_name=ISG15;transcript_type=protein_coding;transcript_name=ISG15-203;level=2;protein_id=ENSP00000496832.1;hgnc_id=HGNC:4053;tag=CAGE_supported_TSS,basic,Ensembl_canonical,MANE_Select,appris_principal_1,CCDS;ccdsid=CCDS6.1;havana_gene=OTTHUMG00000040777.4_10;havana_transcript=OTTHUMT00000501480.1_4;remap_num_mappings=1;remap_status=full_contig;remap_target_status=new\n",
      "ENST00000649529\n",
      "chr1\tHAVANA\texon\t949364\t949920\t.\t+\t.\tID=exon:ENST00000624697.4:3;Parent=ENST00000624697.4;gene_id=ENSG00000187608.10_10;transcript_id=ENST00000624697.4_5;gene_type=protein_coding;gene_name=ISG15;transcript_type=protein_coding;transcript_name=ISG15-202;exon_number=3;exon_id=ENSE00003760003.2_1;level=2;protein_id=ENSP00000485643.1;transcript_support_level=3;hgnc_id=HGNC:4053;tag=basic;havana_gene=OTTHUMG00000040777.4_10;havana_transcript=OTTHUMT00000479384.2_5;remap_original_location=chr1:+:1013984-1014540;remap_status=full_contig\n",
      "ENST00000624697\n",
      "chr1\tHAVANA\texon\t949364\t949920\t.\t+\t.\tID=exon:ENST00000649529.1:2;Parent=ENST00000649529.1;gene_id=ENSG00000187608.10_10;transcript_id=ENST00000649529.1_4;gene_type=protein_coding;gene_name=ISG15;transcript_type=protein_coding;transcript_name=ISG15-203;exon_number=2;exon_id=ENSE00001480799.4_1;level=2;protein_id=ENSP00000496832.1;hgnc_id=HGNC:4053;tag=CAGE_supported_TSS,basic,Ensembl_canonical,MANE_Select,appris_principal_1,CCDS;ccdsid=CCDS6.1;havana_gene=OTTHUMG00000040777.4_10;havana_transcript=OTTHUMT00000501480.1_4;remap_original_location=chr1:+:1013984-1014540;remap_status=full_contig\n",
      "ENST00000649529\n",
      "chr1\tHAVANA\tCDS\t949364\t949858\t.\t+\t0\tID=CDS:ENST00000649529.1;Parent=ENST00000649529.1;gene_id=ENSG00000187608.10_10;transcript_id=ENST00000649529.1_4;gene_type=protein_coding;gene_name=ISG15;transcript_type=protein_coding;transcript_name=ISG15-203;exon_number=2;exon_id=ENSE00001480799.4;level=2;protein_id=ENSP00000496832.1;hgnc_id=HGNC:4053;tag=CAGE_supported_TSS,basic,Ensembl_canonical,MANE_Select,appris_principal_1,CCDS;ccdsid=CCDS6.1;havana_gene=OTTHUMG00000040777.4_10;havana_transcript=OTTHUMT00000501480.1_4;remap_original_location=chr1:+:1013984-1014478;remap_status=full_contig\n",
      "ENST00000649529\n",
      "chr1\tHAVANA\texon\t949364\t949815\t.\t+\t.\tID=exon:ENST00000624652.1:3;Parent=ENST00000624652.1;gene_id=ENSG00000187608.10_10;transcript_id=ENST00000624652.1_4;gene_type=protein_coding;gene_name=ISG15;transcript_type=protein_coding;transcript_name=ISG15-201;exon_number=3;exon_id=ENSE00003756535.1_1;level=2;protein_id=ENSP00000485313.1;transcript_support_level=3;hgnc_id=HGNC:4053;tag=mRNA_end_NF,cds_end_NF;havana_gene=OTTHUMG00000040777.4_10;havana_transcript=OTTHUMT00000479385.2_4;remap_original_location=chr1:+:1013984-1014435;remap_status=full_contig\n",
      "ENST00000624652\n",
      "chr1\tHAVANA\tCDS\t949385\t949815\t.\t+\t0\tID=CDS:ENST00000624652.1;Parent=ENST00000624652.1;gene_id=ENSG00000187608.10_10;transcript_id=ENST00000624652.1_4;gene_type=protein_coding;gene_name=ISG15;transcript_type=protein_coding;transcript_name=ISG15-201;exon_number=3;exon_id=ENSE00003756535.1;level=2;protein_id=ENSP00000485313.1;transcript_support_level=3;hgnc_id=HGNC:4053;tag=mRNA_end_NF,cds_end_NF;havana_gene=OTTHUMG00000040777.4_10;havana_transcript=OTTHUMT00000479385.2_4;remap_original_location=chr1:+:1014005-1014435;remap_status=full_contig\n",
      "ENST00000624652\n",
      "chr1\tHAVANA\tCDS\t949385\t949858\t.\t+\t0\tID=CDS:ENST00000624697.4;Parent=ENST00000624697.4;gene_id=ENSG00000187608.10_10;transcript_id=ENST00000624697.4_5;gene_type=protein_coding;gene_name=ISG15;transcript_type=protein_coding;transcript_name=ISG15-202;exon_number=3;exon_id=ENSE00003760003.2;level=2;protein_id=ENSP00000485643.1;transcript_support_level=3;hgnc_id=HGNC:4053;tag=basic;havana_gene=OTTHUMG00000040777.4_10;havana_transcript=OTTHUMT00000479384.2_5;remap_original_location=chr1:+:1014005-1014478;remap_status=full_contig\n",
      "ENST00000624697\n"
     ]
    }
   ],
   "source": [
    "for t in tbx:\n",
    "    print(t)\n",
    "    enst = re.match(r'ENST\\d+', t.transcript_id).group()\n",
    "    print(enst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>SymbolSource</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ENST</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>EXON</th>\n",
       "      <th>...</th>\n",
       "      <th>DS_DG</th>\n",
       "      <th>DS_DL</th>\n",
       "      <th>DP_AG</th>\n",
       "      <th>DP_AL</th>\n",
       "      <th>DP_DG</th>\n",
       "      <th>DP_DL</th>\n",
       "      <th>MaxSpliceAI</th>\n",
       "      <th>is_Canonical</th>\n",
       "      <th>Int_loc</th>\n",
       "      <th>exon_loc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>879375</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>28706</td>\n",
       "      <td>ENST00000342066</td>\n",
       "      <td>c.1888C&gt;T</td>\n",
       "      <td>14/14</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-102</td>\n",
       "      <td>-79</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1307</td>\n",
       "      <td>0.03</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>579:87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>949443</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>4053</td>\n",
       "      <td>ENST00000379389</td>\n",
       "      <td>c.83T&gt;A</td>\n",
       "      <td>2/2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2</td>\n",
       "      <td>-34</td>\n",
       "      <td>-294</td>\n",
       "      <td>208</td>\n",
       "      <td>0.03</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>enst_not_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>949523</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>4053</td>\n",
       "      <td>ENST00000379389</td>\n",
       "      <td>c.163C&gt;T</td>\n",
       "      <td>2/2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-33</td>\n",
       "      <td>-114</td>\n",
       "      <td>167</td>\n",
       "      <td>128</td>\n",
       "      <td>0.01</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>enst_not_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>949643</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>4053</td>\n",
       "      <td>ENST00000379389</td>\n",
       "      <td>c.285del</td>\n",
       "      <td>2/2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-203</td>\n",
       "      <td>-305</td>\n",
       "      <td>8</td>\n",
       "      <td>43</td>\n",
       "      <td>0.02</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>enst_not_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>949648</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>4053</td>\n",
       "      <td>ENST00000379389</td>\n",
       "      <td>c.288C&gt;G</td>\n",
       "      <td>2/2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-239</td>\n",
       "      <td>-310</td>\n",
       "      <td>-33</td>\n",
       "      <td>3</td>\n",
       "      <td>0.01</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>enst_not_match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20696</th>\n",
       "      <td>1</td>\n",
       "      <td>247597502</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NLRP3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>16400</td>\n",
       "      <td>ENST00000336119</td>\n",
       "      <td>c.2425G&gt;A</td>\n",
       "      <td>5/9</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1770</td>\n",
       "      <td>-40</td>\n",
       "      <td>1940</td>\n",
       "      <td>-180</td>\n",
       "      <td>0.01</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>73:97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20697</th>\n",
       "      <td>1</td>\n",
       "      <td>247599354</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NLRP3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>16400</td>\n",
       "      <td>ENST00000336119</td>\n",
       "      <td>c.2581T&gt;C</td>\n",
       "      <td>6/9</td>\n",
       "      <td>...</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-82</td>\n",
       "      <td>-2786</td>\n",
       "      <td>88</td>\n",
       "      <td>1608</td>\n",
       "      <td>0.02</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>88:82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20698</th>\n",
       "      <td>1</td>\n",
       "      <td>247599355</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NLRP3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>16400</td>\n",
       "      <td>ENST00000336119</td>\n",
       "      <td>c.2582A&gt;G</td>\n",
       "      <td>6/9</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.10</td>\n",
       "      <td>-5</td>\n",
       "      <td>-83</td>\n",
       "      <td>-2201</td>\n",
       "      <td>87</td>\n",
       "      <td>0.14</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>87:83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20699</th>\n",
       "      <td>1</td>\n",
       "      <td>247607362</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NLRP3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>16400</td>\n",
       "      <td>ENST00000336119</td>\n",
       "      <td>c.2758C&gt;T</td>\n",
       "      <td>7/9</td>\n",
       "      <td>...</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-257</td>\n",
       "      <td>-88</td>\n",
       "      <td>-89</td>\n",
       "      <td>82</td>\n",
       "      <td>0.18</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>82:88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20700</th>\n",
       "      <td>1</td>\n",
       "      <td>247607363</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NLRP3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>16400</td>\n",
       "      <td>ENST00000336119</td>\n",
       "      <td>c.2759G&gt;A</td>\n",
       "      <td>7/9</td>\n",
       "      <td>...</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.04</td>\n",
       "      <td>3</td>\n",
       "      <td>-89</td>\n",
       "      <td>-90</td>\n",
       "      <td>81</td>\n",
       "      <td>0.10</td>\n",
       "      <td>False</td>\n",
       "      <td>Exonic</td>\n",
       "      <td>81:89</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18468 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM        POS REF ALT GeneSymbol SymbolSource HGNC_ID  \\\n",
       "0         1     879375   C   T     SAMD11         HGNC   28706   \n",
       "2         1     949443   T   A      ISG15         HGNC    4053   \n",
       "3         1     949523   C   T      ISG15         HGNC    4053   \n",
       "4         1     949643  AC   A      ISG15         HGNC    4053   \n",
       "5         1     949648   C   G      ISG15         HGNC    4053   \n",
       "...     ...        ...  ..  ..        ...          ...     ...   \n",
       "20696     1  247597502   G   A      NLRP3         HGNC   16400   \n",
       "20697     1  247599354   T   C      NLRP3         HGNC   16400   \n",
       "20698     1  247599355   A   G      NLRP3         HGNC   16400   \n",
       "20699     1  247607362   C   T      NLRP3         HGNC   16400   \n",
       "20700     1  247607363   G   A      NLRP3         HGNC   16400   \n",
       "\n",
       "                  ENST      HGVSc   EXON  ... DS_DG DS_DL DP_AG  DP_AL  DP_DG  \\\n",
       "0      ENST00000342066  c.1888C>T  14/14  ...  0.00  0.01  -102    -79     -2   \n",
       "2      ENST00000379389    c.83T>A    2/2  ...  0.00  0.00     2    -34   -294   \n",
       "3      ENST00000379389   c.163C>T    2/2  ...  0.00  0.00   -33   -114    167   \n",
       "4      ENST00000379389   c.285del    2/2  ...  0.02  0.00  -203   -305      8   \n",
       "5      ENST00000379389   c.288C>G    2/2  ...  0.00  0.01  -239   -310    -33   \n",
       "...                ...        ...    ...  ...   ...   ...   ...    ...    ...   \n",
       "20696  ENST00000336119  c.2425G>A    5/9  ...  0.00  0.00  1770    -40   1940   \n",
       "20697  ENST00000336119  c.2581T>C    6/9  ...  0.02  0.00   -82  -2786     88   \n",
       "20698  ENST00000336119  c.2582A>G    6/9  ...  0.00  0.10    -5    -83  -2201   \n",
       "20699  ENST00000336119  c.2758C>T    7/9  ...  0.18  0.12  -257    -88    -89   \n",
       "20700  ENST00000336119  c.2759G>A    7/9  ...  0.10  0.04     3    -89    -90   \n",
       "\n",
       "       DP_DL MaxSpliceAI is_Canonical Int_loc        exon_loc  \n",
       "0      -1307        0.03        False  Exonic          579:87  \n",
       "2        208        0.03        False  Exonic  enst_not_match  \n",
       "3        128        0.01        False  Exonic  enst_not_match  \n",
       "4         43        0.02        False  Exonic  enst_not_match  \n",
       "5          3        0.01        False  Exonic  enst_not_match  \n",
       "...      ...         ...          ...     ...             ...  \n",
       "20696   -180        0.01        False  Exonic           73:97  \n",
       "20697   1608        0.02        False  Exonic           88:82  \n",
       "20698     87        0.14        False  Exonic           87:83  \n",
       "20699     82        0.18        False  Exonic           82:88  \n",
       "20700     81        0.10        False  Exonic           81:89  \n",
       "\n",
       "[18468 rows x 24 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df.loc[(df['Int_loc'] == 'Exonic') & (df['exon_loc'] is None)]\n",
    "df.loc[(df['Int_loc'] == 'Exonic')]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A total of 253018 DM mutations are found in allmut.\n"
     ]
    }
   ],
   "source": [
    "all_mut_default_colnames: list = [\n",
    "    \"disase\", \"gene\", \"chrom\", \"genename\", \"gdbid\", \"omimid\", \"amino\", \n",
    "    \"deletion\", \"insertion\", \"codon\", \"codonAff\", \"descr\", \"refseq\", \"hgvs\", \n",
    "    \"hgvsAll\", \"dbsnp\", \"chromosome\", \"startCoord\", \"endCoord\", \n",
    "    \"expected_inheritance\", \"gnomad_AC\", \"gnomad_AF\", \"gnomad_AN\", \"tag\", \n",
    "    \"dmsupport\", \"rankscore\", \"mutype\", \"author\", \"title\", \"fullname\", \n",
    "    \"allname\", \"vol\", \"page\", \"year\", \"pmid\", \"pmidAll\", \"reftag\", \"comments\", \n",
    "    \"acc_num\", \"new_date\", \"base\", \"clinvarID\", \"clinvar_clnsig\"\n",
    "]\n",
    "allmut: pd.DataFrame = pd.read_csv(\n",
    "    'allmut.csv', sep=';', encoding='cp1252', names=all_mut_default_colnames, \n",
    "    skiprows=1,low_memory=False)\n",
    "\n",
    "allmut = allmut[\n",
    "    [\"gene\", \"genename\", \"mutype\", \"clinvar_clnsig\", \"tag\",\n",
    "     \"refseq\", \"hgvs\", \"hgvsAll\", \"chromosome\", \"startCoord\", \"endCoord\", \n",
    "     \"amino\", \"deletion\", \"insertion\", \"expected_inheritance\", \"gnomad_AF\"]]\n",
    "\n",
    "# Drop non-numeric values in 'startCoord'\n",
    "allmut = allmut.dropna(subset=['startCoord'])\n",
    "\n",
    "# Drop duplicates in 'chrom', 'startCoord', and 'endCoord'\n",
    "allmut = allmut.drop_duplicates(subset=['chromosome', 'startCoord', 'endCoord'])\n",
    "\n",
    "# Extract tag == \"DM\" from allmut\n",
    "allmut_dm = allmut[allmut.tag == \"DM\"].copy()\n",
    "print(f\"A total of {len(allmut_dm)} DM mutations are found in allmut.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A total of 231981 DM mutations are found in allmut with MAF 0.\n",
      "A total of 154107 DM mutations are found in allmut with MAF 0 and non-deletion or non-insertion.\n",
      "Splicing_DM: 23777, Non-splicing_DM: 229241\n"
     ]
    }
   ],
   "source": [
    "# Fillna with empty string in \"gnomad_AF\" colmun in allmut_dm\n",
    "# Extratct MAF 0 from allmut_dm\n",
    "allmut_dm['gnomad_AF'].fillna(0, inplace=True)\n",
    "allmut_dm_maf0 = allmut_dm[allmut_dm['gnomad_AF'] == 0].copy()\n",
    "print(f\"A total of {len(allmut_dm_maf0)} DM mutations are found in allmut with MAF 0.\")\n",
    "\n",
    "# Extract non-deletion or non-insertion from allmut_dm\n",
    "allmut_dm_maf0_snv = allmut_dm_maf0[(allmut_dm_maf0['deletion'].isnull()) & (allmut_dm_maf0['insertion'].isnull())]\n",
    "print(f\"A total of {len(allmut_dm_maf0_snv)} DM mutations are found in allmut with MAF 0 and non-deletion or non-insertion.\")\n",
    "\n",
    "# Extract the mutation type from the mutype column\n",
    "splice_mutations = allmut_dm[allmut_dm[\"mutype\"].str.contains(\"splice\")].copy()\n",
    "non_splice_mutations = allmut_dm[~allmut_dm[\"mutype\"].str.contains(\"splice\")]\n",
    "print(f\"Splicing_DM: {len(splice_mutations)}, Non-splicing_DM: {len(non_splice_mutations)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['missense' 'initiation' 'nonsense' 'synonymous' 'gross'\n",
      " 'canonical-splice' 'exonic-splice' 'nonstop' 'regulatory' 'splice']\n",
      "[0.]\n"
     ]
    }
   ],
   "source": [
    "print(allmut_dm_maf0_snv['mutype'].unique())\n",
    "print(allmut_dm_maf0_snv['gnomad_AF'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>genename</th>\n",
       "      <th>mutype</th>\n",
       "      <th>clinvar_clnsig</th>\n",
       "      <th>tag</th>\n",
       "      <th>refseq</th>\n",
       "      <th>hgvs</th>\n",
       "      <th>hgvsAll</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>startCoord</th>\n",
       "      <th>endCoord</th>\n",
       "      <th>amino</th>\n",
       "      <th>deletion</th>\n",
       "      <th>insertion</th>\n",
       "      <th>expected_inheritance</th>\n",
       "      <th>gnomad_AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96651</th>\n",
       "      <td>AP3B1</td>\n",
       "      <td>Adaptor related protein complex 3 subunit beta 1</td>\n",
       "      <td>canonical-splice</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_003664.5</td>\n",
       "      <td>1168-1G&gt;C</td>\n",
       "      <td>1168minus1GtoC</td>\n",
       "      <td>5</td>\n",
       "      <td>78165673.0</td>\n",
       "      <td>78165673.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AR</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126648</th>\n",
       "      <td>NF1</td>\n",
       "      <td>Neurofibromin 1</td>\n",
       "      <td>exonic-splice</td>\n",
       "      <td>Pathogenic/Likely_pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000267.3</td>\n",
       "      <td>1748A&gt;G</td>\n",
       "      <td>1748AtoG | K583R</td>\n",
       "      <td>17</td>\n",
       "      <td>31223470.0</td>\n",
       "      <td>31223470.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AD</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129700</th>\n",
       "      <td>TSC2</td>\n",
       "      <td>TSC complex subunit 2</td>\n",
       "      <td>exonic-splice</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000548.5</td>\n",
       "      <td>1255C&gt;T</td>\n",
       "      <td>1255CtoT | P419S</td>\n",
       "      <td>16</td>\n",
       "      <td>2062006.0</td>\n",
       "      <td>2062006.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AD</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130620</th>\n",
       "      <td>COL4A5</td>\n",
       "      <td>Collagen type IV alpha 5 chain</td>\n",
       "      <td>exonic-splice</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000495.5</td>\n",
       "      <td>1856C&gt;T</td>\n",
       "      <td>1856CtoT | P619L</td>\n",
       "      <td>X</td>\n",
       "      <td>108598778.0</td>\n",
       "      <td>108598778.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XLD</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134411</th>\n",
       "      <td>DMD</td>\n",
       "      <td>Dystrophin</td>\n",
       "      <td>exonic-splice</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_004006.3</td>\n",
       "      <td>10279C&gt;T</td>\n",
       "      <td>10279CtoT | Q3427*</td>\n",
       "      <td>X</td>\n",
       "      <td>31173588.0</td>\n",
       "      <td>31173588.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XLR</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510785</th>\n",
       "      <td>PPT1</td>\n",
       "      <td>Palmitoyl-protein thioesterase 1</td>\n",
       "      <td>canonical-splice</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000310.4</td>\n",
       "      <td>536+1G&gt;A</td>\n",
       "      <td>536plus1GtoA</td>\n",
       "      <td>1</td>\n",
       "      <td>40089409.0</td>\n",
       "      <td>40089409.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AR</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510786</th>\n",
       "      <td>APC</td>\n",
       "      <td>APC regulator of WNT signaling pathway</td>\n",
       "      <td>splice</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000038.6</td>\n",
       "      <td>1409-6A&gt;G</td>\n",
       "      <td>1409minus6AtoG</td>\n",
       "      <td>5</td>\n",
       "      <td>112827102.0</td>\n",
       "      <td>112827102.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AD</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510789</th>\n",
       "      <td>UROD</td>\n",
       "      <td>Uroporphyrinogen decarboxylase</td>\n",
       "      <td>canonical-splice</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000374.5</td>\n",
       "      <td>876-1G&gt;C</td>\n",
       "      <td>876minus1GtoC</td>\n",
       "      <td>1</td>\n",
       "      <td>45014939.0</td>\n",
       "      <td>45014939.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510791</th>\n",
       "      <td>PTS</td>\n",
       "      <td>6-pyruvoyltetrahydropterin synthase</td>\n",
       "      <td>canonical-splice</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000317.3</td>\n",
       "      <td>186+1G&gt;A</td>\n",
       "      <td>186plus1GtoA</td>\n",
       "      <td>11</td>\n",
       "      <td>112230231.0</td>\n",
       "      <td>112230231.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AR</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510792</th>\n",
       "      <td>IDS</td>\n",
       "      <td>Iduronate 2-sulfatase</td>\n",
       "      <td>splice</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000202.8</td>\n",
       "      <td>880-8A&gt;G</td>\n",
       "      <td>880minus8AtoG</td>\n",
       "      <td>X</td>\n",
       "      <td>149490448.0</td>\n",
       "      <td>149490448.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XLR</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23777 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene                                          genename  \\\n",
       "96651    AP3B1  Adaptor related protein complex 3 subunit beta 1   \n",
       "126648     NF1                                   Neurofibromin 1   \n",
       "129700    TSC2                             TSC complex subunit 2   \n",
       "130620  COL4A5                    Collagen type IV alpha 5 chain   \n",
       "134411     DMD                                        Dystrophin   \n",
       "...        ...                                               ...   \n",
       "510785    PPT1                  Palmitoyl-protein thioesterase 1   \n",
       "510786     APC            APC regulator of WNT signaling pathway   \n",
       "510789    UROD                    Uroporphyrinogen decarboxylase   \n",
       "510791     PTS               6-pyruvoyltetrahydropterin synthase   \n",
       "510792     IDS                             Iduronate 2-sulfatase   \n",
       "\n",
       "                  mutype                clinvar_clnsig tag       refseq  \\\n",
       "96651   canonical-splice                           NaN  DM  NM_003664.5   \n",
       "126648     exonic-splice  Pathogenic/Likely_pathogenic  DM  NM_000267.3   \n",
       "129700     exonic-splice                    Pathogenic  DM  NM_000548.5   \n",
       "130620     exonic-splice             Likely_pathogenic  DM  NM_000495.5   \n",
       "134411     exonic-splice                    Pathogenic  DM  NM_004006.3   \n",
       "...                  ...                           ...  ..          ...   \n",
       "510785  canonical-splice             Likely_pathogenic  DM  NM_000310.4   \n",
       "510786            splice             Likely_pathogenic  DM  NM_000038.6   \n",
       "510789  canonical-splice                           NaN  DM  NM_000374.5   \n",
       "510791  canonical-splice                    Pathogenic  DM  NM_000317.3   \n",
       "510792            splice                           NaN  DM  NM_000202.8   \n",
       "\n",
       "             hgvs             hgvsAll chromosome   startCoord     endCoord  \\\n",
       "96651   1168-1G>C      1168minus1GtoC          5   78165673.0   78165673.0   \n",
       "126648    1748A>G    1748AtoG | K583R         17   31223470.0   31223470.0   \n",
       "129700    1255C>T    1255CtoT | P419S         16    2062006.0    2062006.0   \n",
       "130620    1856C>T    1856CtoT | P619L          X  108598778.0  108598778.0   \n",
       "134411   10279C>T  10279CtoT | Q3427*          X   31173588.0   31173588.0   \n",
       "...           ...                 ...        ...          ...          ...   \n",
       "510785   536+1G>A        536plus1GtoA          1   40089409.0   40089409.0   \n",
       "510786  1409-6A>G      1409minus6AtoG          5  112827102.0  112827102.0   \n",
       "510789   876-1G>C       876minus1GtoC          1   45014939.0   45014939.0   \n",
       "510791   186+1G>A        186plus1GtoA         11  112230231.0  112230231.0   \n",
       "510792   880-8A>G       880minus8AtoG          X  149490448.0  149490448.0   \n",
       "\n",
       "       amino deletion insertion expected_inheritance  gnomad_AF  \n",
       "96651    NaN      NaN       NaN                   AR        0.0  \n",
       "126648   NaN      NaN       NaN                   AD        0.0  \n",
       "129700   NaN      NaN       NaN                   AD        0.0  \n",
       "130620   NaN      NaN       NaN                  XLD        0.0  \n",
       "134411   NaN      NaN       NaN                  XLR        0.0  \n",
       "...      ...      ...       ...                  ...        ...  \n",
       "510785   NaN      NaN       NaN                   AR        0.0  \n",
       "510786   NaN      NaN       NaN                   AD        0.0  \n",
       "510789   NaN      NaN       NaN                 ADAR        0.0  \n",
       "510791   NaN      NaN       NaN                   AR        0.0  \n",
       "510792   NaN      NaN       NaN                  XLR        0.0  \n",
       "\n",
       "[23777 rows x 16 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splice_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>genename</th>\n",
       "      <th>mutype</th>\n",
       "      <th>clinvarID</th>\n",
       "      <th>clinvar_clnsig</th>\n",
       "      <th>tag</th>\n",
       "      <th>refseq</th>\n",
       "      <th>hgvs</th>\n",
       "      <th>hgvsAll</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>startCoord</th>\n",
       "      <th>endCoord</th>\n",
       "      <th>amino</th>\n",
       "      <th>deletion</th>\n",
       "      <th>insertion</th>\n",
       "      <th>expected_inheritance</th>\n",
       "      <th>gnomad_AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene, chrom, genename, mutype, clinvarID, clinvar_clnsig, tag, refseq, hgvs, hgvsAll, chromosome, startCoord, endCoord, amino, deletion, insertion, expected_inheritance, gnomad_AF]\n",
       "Index: []"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def remove_dot_ver(x):\n",
    "    if x == '.':\n",
    "        pass\n",
    "    else:\n",
    "        return re.match(r'[a-zA-Z_]+\\d+', x).group()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "510804"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(allmut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _liftover_to_hg19(chrom, pos):\n",
    "    converter = get_lifter('hg38', 'hg19')\n",
    "    result = converter.query(chrom, pos)\n",
    "    if result:\n",
    "        return result[0]\n",
    "    else:\n",
    "        return None\n",
    "    \n",
    "def anno_vars_id(row):\n",
    "    variant_id = f'{row[\"CHROM\"]}:{row[\"POS_hg19\"]}-{row[\"REF\"]}-{row[\"ALT\"]}'\n",
    "    return variant_id\n",
    "\n",
    "def anno_hg19_pos(row):\n",
    "    converted = _liftover_to_hg19(row['CHROM'], row['POS_hg38'])\n",
    "    return converted[1]\n",
    "\n",
    "def remove_dot_ver(x):\n",
    "    if x == '.':\n",
    "        pass\n",
    "    else:\n",
    "        return re.match(r'[a-zA-Z_]+\\d+', x).group()\n",
    "    \n",
    "def remove_non_canon(x):\n",
    "    if x in enst_set:\n",
    "        return True\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "def create_refalt(row, nt):\n",
    "    if row['Strand'] == '+':\n",
    "        return row[nt]\n",
    "    elif row['Strand'] == '-':\n",
    "        return str(Seq(row[nt]).complement())\n",
    "    else:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "canonlist = '/Github/Projects/DeNovo/data/CanonicalTranscripts_COMP/CanonicalTranscripts.exoncount.tsv'\n",
    "refseq = '/work/CanonicalTrasncripts/gencode.v43lift37.metadata.RefSeq.gz'\n",
    "\n",
    "df_canon = pd.read_table(canonlist, header=0)\n",
    "df_canon = df_canon[(df_canon['ENST'] != 'ENST00000649912')\n",
    "                    & (df_canon['ENST'] != 'ENST00000609375')]\n",
    "df_enst = df_canon.drop_duplicates(subset='ENST')\n",
    "enst_set = set(df_enst['ENST'])\n",
    "df_refseq = pd.read_table(refseq, header=None, \n",
    "                          names=['ENST_refseq', 'RefSeq_RNA', 'RefSeq_Pro'])\n",
    "df_refseq.fillna(value='.', inplace=True)\n",
    "\n",
    "df_refseq['ENST_refseq'] = df_refseq['ENST_refseq'].apply(remove_dot_ver)\n",
    "df_refseq['RefSeq_RNA'] = df_refseq['RefSeq_RNA'].apply(remove_dot_ver)\n",
    "df_refseq['RefSeq_Pro'] = df_refseq['RefSeq_Pro'].apply(remove_dot_ver)\n",
    "\n",
    "df_refseq['is_Canonical'] =  df_refseq['ENST_refseq'].apply(remove_non_canon)\n",
    "df_refseq_canon = df_refseq[df_refseq['is_Canonical'] == True].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Generate VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading HGMD splicing variants list (tsv)\n",
    "hgmd_file = '/Github/MyProjects/DeNovo/data/ValidationData/Positive/allmut.trim.colfixed.maf0.tsv.txt'\n",
    "df = pd.read_table(hgmd_file, header=0)\n",
    "\n",
    "# Pre-processing\n",
    "df = df.dropna(subset=['startCoord'])\n",
    "df['startCoord'] = df['startCoord'].astype(int)\n",
    "df['refseq'] = df['refseq'].apply(remove_dot_ver)\n",
    "\n",
    "# Extract REF and ALT from HGVS descriptions\n",
    "sr_alt_nt = df['hgvs'].str[-1:].rename('alt_nt')\n",
    "sr_ref_nt = df['hgvs'].str[-3:-2].rename('ref_nt')\n",
    "df = pd.concat([df, sr_ref_nt, sr_alt_nt], axis=1)\n",
    "\n",
    "# Insert cols as VCF\n",
    "df.loc[:,'ID'] = '.'\n",
    "df.loc[:,'QUAL'] = '.'\n",
    "df.loc[:,'FILTER'] = '.'\n",
    "df.loc[:,'INFO'] = '.'\n",
    "\n",
    "# Annotate ENST IDs\n",
    "df = pd.merge(df, df_enst, how='left', \n",
    "              left_on='gene', right_on='GeneSymbol')\n",
    "\n",
    "# Rename cols for downstream processing\n",
    "df = df.rename(columns={'CHROM': 'Chr',\n",
    "                        'chromosome': 'CHROM', \n",
    "                        'startCoord': 'POS_hg38'})\n",
    "\n",
    "# Create REF and ALT columns\n",
    "df['REF'] = df.apply(create_refalt, nt='ref_nt', axis=1)\n",
    "df['ALT'] = df.apply(create_refalt, nt='alt_nt', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Liftover to hg19\n",
    "df['POS_hg19'] = df.apply(anno_hg19_pos, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = df.copy()\n",
    "df2 = df2.astype({'POS_hg19': str})\n",
    "\n",
    "# Annotate variant IDs\n",
    "df2['variant_id'] = df2.apply(anno_vars_id, axis=1)\n",
    "\n",
    "# Extract columns for VCF\n",
    "df_19 = df2[['CHROM', 'POS_hg19', \n",
    "             'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO']]\n",
    "df_38 = df2[['CHROM', 'POS_hg38', \n",
    "             'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_19 = df_19.dropna(subset=['REF'])\n",
    "df_38 = df_38.dropna(subset=['REF'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output as VCF\n",
    "df_19.to_csv('./patho.hg19.vcf', sep='\\t', index=False, header=False)\n",
    "df_38.to_csv('./patho.hg38.vcf', sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Concatenate header and variant list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "metadata": {},
   "outputs": [],
   "source": [
    "!cat ../header_for_VCF.tsv ./patho.hg19.vcf > patho.hg19.header.vcf\n",
    "!cat ../header_for_VCF.tsv ./patho.hg38.vcf > patho.hg38.header.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Strand</th>\n",
       "      <th>ref_nt</th>\n",
       "      <th>REF</th>\n",
       "      <th>alt_nt</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10553</th>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>None</td>\n",
       "      <td>A</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10554</th>\n",
       "      <td>+</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10555</th>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10556</th>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10557</th>\n",
       "      <td>+</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10558 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Strand ref_nt   REF alt_nt   ALT\n",
       "0          +      A     A      G     G\n",
       "1          -      T     A      G     C\n",
       "2          -      G     C      A     T\n",
       "3          +      A     A      G     G\n",
       "4          -      A     T      G     C\n",
       "...      ...    ...   ...    ...   ...\n",
       "10553    NaN      G  None      A  None\n",
       "10554      +      G     G      A     A\n",
       "10555      -      G     C      A     T\n",
       "10556      -      G     C      A     T\n",
       "10557      +      G     G      A     A\n",
       "\n",
       "[10558 rows x 5 columns]"
      ]
     },
     "execution_count": 263,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['Strand', 'ref_nt', 'REF', 'alt_nt', 'ALT']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 =df2[['gene', 'gdbid', 'omimid', 'amino', 'codon', 'codonAff', 'descr',\n",
    "       'refseq', 'hgvs', 'CHROM', 'POS_hg38', 'endCoord', \n",
    "       'expected_inheritance', 'dmsupport', 'mutype', 'acc_num', 'new_date', \n",
    "       'clinvarID', 'clinvar_clnsig', 'Chr', 'Start', 'End', 'Strand', \n",
    "       'GeneSymbol', 'HGNC_ID', 'ENSG', 'ENST', 'GeneType', 'Tag', \n",
    "       'REF', 'ALT', 'POS_hg19', 'variant_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2.to_csv('./patho2.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "wesanno",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
