{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Import package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "from Bio.Seq import Seq\n",
    "from liftover import get_lifter\n",
    "# from concurrent.futures import ThreadPoolExecutor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A total of 253018 DM mutations are found in allmut.\n"
     ]
    }
   ],
   "source": [
    "all_mut_default_colnames: list = [\n",
    "    \"disase\", \"gene\", \"chrom\", \"genename\", \"gdbid\", \"omimid\", \"amino\", \n",
    "    \"deletion\", \"insertion\", \"codon\", \"codonAff\", \"descr\", \"refseq\", \"hgvs\", \n",
    "    \"hgvsAll\", \"dbsnp\", \"chromosome\", \"startCoord\", \"endCoord\", \n",
    "    \"expected_inheritance\", \"gnomad_AC\", \"gnomad_AF\", \"gnomad_AN\", \"tag\", \n",
    "    \"dmsupport\", \"rankscore\", \"mutype\", \"author\", \"title\", \"fullname\", \n",
    "    \"allname\", \"vol\", \"page\", \"year\", \"pmid\", \"pmidAll\", \"reftag\", \"comments\", \n",
    "    \"acc_num\", \"new_date\", \"base\", \"clinvarID\", \"clinvar_clnsig\"\n",
    "]\n",
    "allmut: pd.DataFrame = pd.read_csv(\n",
    "    'allmut.csv', sep=';', encoding='cp1252', names=all_mut_default_colnames, \n",
    "    skiprows=1,low_memory=False)\n",
    "\n",
    "allmut = allmut[\n",
    "    [\"gene\", \"genename\", \"mutype\", \"clinvar_clnsig\", \"tag\",\n",
    "     \"refseq\", \"hgvs\", \"hgvsAll\", \"chromosome\", \"startCoord\", \"endCoord\", \n",
    "     \"amino\", \"deletion\", \"insertion\", \"expected_inheritance\", \"gnomad_AF\"]]\n",
    "\n",
    "# Drop non-numeric values in 'startCoord'\n",
    "allmut = allmut.dropna(subset=['startCoord'])\n",
    "\n",
    "# Drop duplicates in 'chrom', 'startCoord', and 'endCoord'\n",
    "allmut = allmut.drop_duplicates(subset=['chromosome', 'startCoord', 'endCoord'])\n",
    "\n",
    "# Extract tag == \"DM\" from allmut\n",
    "allmut_dm = allmut[allmut.tag == \"DM\"]\n",
    "print(f\"A total of {len(allmut_dm)} DM mutations are found in allmut.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A total of 231981 DM mutations are found in allmut with MAF 0.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/b_/1qjd6n853xbbctck9_20_nt40000gn/T/ipykernel_28612/1764070414.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  allmut_dm['gnomad_AF'].fillna(0, inplace=True)\n",
      "/var/folders/b_/1qjd6n853xbbctck9_20_nt40000gn/T/ipykernel_28612/1764070414.py:8: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  allmut_dm_maf0_snv = allmut_dm_maf0[(allmut_dm_maf0['deletion'].isnull()) & (allmut_dm['insertion'].isnull())].copy()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A total of 154107 DM mutations are found in allmut with MAF 0 and non-deletion or non-insertion.\n",
      "23777 229241\n"
     ]
    }
   ],
   "source": [
    "# Fillna with empty string in \"gnomad_AF\" colmun in allmut_dm\n",
    "# Extratct MAF 0 from allmut_dm\n",
    "allmut_dm['gnomad_AF'].fillna(0, inplace=True)\n",
    "allmut_dm_maf0 = allmut_dm[allmut_dm['gnomad_AF'] == 0]\n",
    "print(f\"A total of {len(allmut_dm_maf0)} DM mutations are found in allmut with MAF 0.\")\n",
    "\n",
    "# Extract non-deletion or non-insertion from allmut_dm\n",
    "allmut_dm_maf0_snv = allmut_dm_maf0[(allmut_dm_maf0['deletion'].isnull()) & (allmut_dm['insertion'].isnull())].copy()\n",
    "print(f\"A total of {len(allmut_dm_maf0_snv)} DM mutations are found in allmut with MAF 0 and non-deletion or non-insertion.\")\n",
    "\n",
    "# Extract the mutation type from the mutype column\n",
    "splice_mutations = allmut_dm[allmut_dm[\"mutype\"].str.contains(\"splice\")]\n",
    "non_splice_mutations = allmut_dm[~allmut_dm[\"mutype\"].str.contains(\"splice\")]\n",
    "print(len(splice_mutations), len(non_splice_mutations))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>genename</th>\n",
       "      <th>mutype</th>\n",
       "      <th>clinvar_clnsig</th>\n",
       "      <th>tag</th>\n",
       "      <th>refseq</th>\n",
       "      <th>hgvs</th>\n",
       "      <th>hgvsAll</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>startCoord</th>\n",
       "      <th>endCoord</th>\n",
       "      <th>amino</th>\n",
       "      <th>deletion</th>\n",
       "      <th>insertion</th>\n",
       "      <th>expected_inheritance</th>\n",
       "      <th>gnomad_AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96651</th>\n",
       "      <td>AP3B1</td>\n",
       "      <td>Adaptor related protein complex 3 subunit beta 1</td>\n",
       "      <td>canonical-splice</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_003664.5</td>\n",
       "      <td>1168-1G&gt;C</td>\n",
       "      <td>1168minus1GtoC</td>\n",
       "      <td>5</td>\n",
       "      <td>78165673.0</td>\n",
       "      <td>78165673.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AR</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126648</th>\n",
       "      <td>NF1</td>\n",
       "      <td>Neurofibromin 1</td>\n",
       "      <td>exonic-splice</td>\n",
       "      <td>Pathogenic/Likely_pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000267.3</td>\n",
       "      <td>1748A&gt;G</td>\n",
       "      <td>1748AtoG | K583R</td>\n",
       "      <td>17</td>\n",
       "      <td>31223470.0</td>\n",
       "      <td>31223470.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AD</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129700</th>\n",
       "      <td>TSC2</td>\n",
       "      <td>TSC complex subunit 2</td>\n",
       "      <td>exonic-splice</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000548.5</td>\n",
       "      <td>1255C&gt;T</td>\n",
       "      <td>1255CtoT | P419S</td>\n",
       "      <td>16</td>\n",
       "      <td>2062006.0</td>\n",
       "      <td>2062006.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AD</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130620</th>\n",
       "      <td>COL4A5</td>\n",
       "      <td>Collagen type IV alpha 5 chain</td>\n",
       "      <td>exonic-splice</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000495.5</td>\n",
       "      <td>1856C&gt;T</td>\n",
       "      <td>1856CtoT | P619L</td>\n",
       "      <td>X</td>\n",
       "      <td>108598778.0</td>\n",
       "      <td>108598778.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XLD</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134411</th>\n",
       "      <td>DMD</td>\n",
       "      <td>Dystrophin</td>\n",
       "      <td>exonic-splice</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_004006.3</td>\n",
       "      <td>10279C&gt;T</td>\n",
       "      <td>10279CtoT | Q3427*</td>\n",
       "      <td>X</td>\n",
       "      <td>31173588.0</td>\n",
       "      <td>31173588.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XLR</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510785</th>\n",
       "      <td>PPT1</td>\n",
       "      <td>Palmitoyl-protein thioesterase 1</td>\n",
       "      <td>canonical-splice</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000310.4</td>\n",
       "      <td>536+1G&gt;A</td>\n",
       "      <td>536plus1GtoA</td>\n",
       "      <td>1</td>\n",
       "      <td>40089409.0</td>\n",
       "      <td>40089409.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AR</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510786</th>\n",
       "      <td>APC</td>\n",
       "      <td>APC regulator of WNT signaling pathway</td>\n",
       "      <td>splice</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000038.6</td>\n",
       "      <td>1409-6A&gt;G</td>\n",
       "      <td>1409minus6AtoG</td>\n",
       "      <td>5</td>\n",
       "      <td>112827102.0</td>\n",
       "      <td>112827102.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AD</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510789</th>\n",
       "      <td>UROD</td>\n",
       "      <td>Uroporphyrinogen decarboxylase</td>\n",
       "      <td>canonical-splice</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000374.5</td>\n",
       "      <td>876-1G&gt;C</td>\n",
       "      <td>876minus1GtoC</td>\n",
       "      <td>1</td>\n",
       "      <td>45014939.0</td>\n",
       "      <td>45014939.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510791</th>\n",
       "      <td>PTS</td>\n",
       "      <td>6-pyruvoyltetrahydropterin synthase</td>\n",
       "      <td>canonical-splice</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000317.3</td>\n",
       "      <td>186+1G&gt;A</td>\n",
       "      <td>186plus1GtoA</td>\n",
       "      <td>11</td>\n",
       "      <td>112230231.0</td>\n",
       "      <td>112230231.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AR</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510792</th>\n",
       "      <td>IDS</td>\n",
       "      <td>Iduronate 2-sulfatase</td>\n",
       "      <td>splice</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000202.8</td>\n",
       "      <td>880-8A&gt;G</td>\n",
       "      <td>880minus8AtoG</td>\n",
       "      <td>X</td>\n",
       "      <td>149490448.0</td>\n",
       "      <td>149490448.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XLR</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23777 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene                                          genename  \\\n",
       "96651    AP3B1  Adaptor related protein complex 3 subunit beta 1   \n",
       "126648     NF1                                   Neurofibromin 1   \n",
       "129700    TSC2                             TSC complex subunit 2   \n",
       "130620  COL4A5                    Collagen type IV alpha 5 chain   \n",
       "134411     DMD                                        Dystrophin   \n",
       "...        ...                                               ...   \n",
       "510785    PPT1                  Palmitoyl-protein thioesterase 1   \n",
       "510786     APC            APC regulator of WNT signaling pathway   \n",
       "510789    UROD                    Uroporphyrinogen decarboxylase   \n",
       "510791     PTS               6-pyruvoyltetrahydropterin synthase   \n",
       "510792     IDS                             Iduronate 2-sulfatase   \n",
       "\n",
       "                  mutype                clinvar_clnsig tag       refseq  \\\n",
       "96651   canonical-splice                           NaN  DM  NM_003664.5   \n",
       "126648     exonic-splice  Pathogenic/Likely_pathogenic  DM  NM_000267.3   \n",
       "129700     exonic-splice                    Pathogenic  DM  NM_000548.5   \n",
       "130620     exonic-splice             Likely_pathogenic  DM  NM_000495.5   \n",
       "134411     exonic-splice                    Pathogenic  DM  NM_004006.3   \n",
       "...                  ...                           ...  ..          ...   \n",
       "510785  canonical-splice             Likely_pathogenic  DM  NM_000310.4   \n",
       "510786            splice             Likely_pathogenic  DM  NM_000038.6   \n",
       "510789  canonical-splice                           NaN  DM  NM_000374.5   \n",
       "510791  canonical-splice                    Pathogenic  DM  NM_000317.3   \n",
       "510792            splice                           NaN  DM  NM_000202.8   \n",
       "\n",
       "             hgvs             hgvsAll chromosome   startCoord     endCoord  \\\n",
       "96651   1168-1G>C      1168minus1GtoC          5   78165673.0   78165673.0   \n",
       "126648    1748A>G    1748AtoG | K583R         17   31223470.0   31223470.0   \n",
       "129700    1255C>T    1255CtoT | P419S         16    2062006.0    2062006.0   \n",
       "130620    1856C>T    1856CtoT | P619L          X  108598778.0  108598778.0   \n",
       "134411   10279C>T  10279CtoT | Q3427*          X   31173588.0   31173588.0   \n",
       "...           ...                 ...        ...          ...          ...   \n",
       "510785   536+1G>A        536plus1GtoA          1   40089409.0   40089409.0   \n",
       "510786  1409-6A>G      1409minus6AtoG          5  112827102.0  112827102.0   \n",
       "510789   876-1G>C       876minus1GtoC          1   45014939.0   45014939.0   \n",
       "510791   186+1G>A        186plus1GtoA         11  112230231.0  112230231.0   \n",
       "510792   880-8A>G       880minus8AtoG          X  149490448.0  149490448.0   \n",
       "\n",
       "       amino deletion insertion expected_inheritance  gnomad_AF  \n",
       "96651    NaN      NaN       NaN                   AR        0.0  \n",
       "126648   NaN      NaN       NaN                   AD        0.0  \n",
       "129700   NaN      NaN       NaN                   AD        0.0  \n",
       "130620   NaN      NaN       NaN                  XLD        0.0  \n",
       "134411   NaN      NaN       NaN                  XLR        0.0  \n",
       "...      ...      ...       ...                  ...        ...  \n",
       "510785   NaN      NaN       NaN                   AR        0.0  \n",
       "510786   NaN      NaN       NaN                   AD        0.0  \n",
       "510789   NaN      NaN       NaN                 ADAR        0.0  \n",
       "510791   NaN      NaN       NaN                   AR        0.0  \n",
       "510792   NaN      NaN       NaN                  XLR        0.0  \n",
       "\n",
       "[23777 rows x 16 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splice_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['missense' 'initiation' 'nonsense' 'synonymous' 'gross'\n",
      " 'canonical-splice' 'exonic-splice' 'nonstop' 'regulatory' 'splice']\n",
      "[0.]\n"
     ]
    }
   ],
   "source": [
    "print(allmut_dm['mutype'].unique())\n",
    "print(allmut_dm['gnomad_AF'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>genename</th>\n",
       "      <th>mutype</th>\n",
       "      <th>clinvar_clnsig</th>\n",
       "      <th>tag</th>\n",
       "      <th>refseq</th>\n",
       "      <th>hgvs</th>\n",
       "      <th>hgvsAll</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>startCoord</th>\n",
       "      <th>endCoord</th>\n",
       "      <th>amino</th>\n",
       "      <th>deletion</th>\n",
       "      <th>insertion</th>\n",
       "      <th>expected_inheritance</th>\n",
       "      <th>gnomad_AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>462252</th>\n",
       "      <td>NF1</td>\n",
       "      <td>Neurofibromin 1</td>\n",
       "      <td>splice</td>\n",
       "      <td>Pathogenic/Likely_pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000267.3</td>\n",
       "      <td>2410-16A&gt;G</td>\n",
       "      <td>2410minus16AtoG</td>\n",
       "      <td>17</td>\n",
       "      <td>31229009.0</td>\n",
       "      <td>31229009.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AD</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462258</th>\n",
       "      <td>NF1</td>\n",
       "      <td>Neurofibromin 1</td>\n",
       "      <td>splice</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000267.3</td>\n",
       "      <td>1393-9T&gt;A</td>\n",
       "      <td>1393minus9TtoA</td>\n",
       "      <td>17</td>\n",
       "      <td>31214442.0</td>\n",
       "      <td>31214442.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AD</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462270</th>\n",
       "      <td>NTRK1</td>\n",
       "      <td>Neurotrophic receptor tyrosine kinase 1</td>\n",
       "      <td>splice</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_001012331.2</td>\n",
       "      <td>851-33T&gt;A</td>\n",
       "      <td>851minus33TtoA</td>\n",
       "      <td>1</td>\n",
       "      <td>156873600.0</td>\n",
       "      <td>156873600.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AR</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462274</th>\n",
       "      <td>SLC25A20</td>\n",
       "      <td>Solute carrier family 25 member 20</td>\n",
       "      <td>splice</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000387.6</td>\n",
       "      <td>199-10T&gt;G</td>\n",
       "      <td>199minus10TtoG</td>\n",
       "      <td>3</td>\n",
       "      <td>48884134.0</td>\n",
       "      <td>48884134.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AR</td>\n",
       "      <td>0.000065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462325</th>\n",
       "      <td>MAPT</td>\n",
       "      <td>Microtubule associated protein tau</td>\n",
       "      <td>splice</td>\n",
       "      <td>not_provided</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_005910.6</td>\n",
       "      <td>915+12C&gt;T</td>\n",
       "      <td>915plus12CtoT</td>\n",
       "      <td>17</td>\n",
       "      <td>46010414.0</td>\n",
       "      <td>46010414.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AD</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510767</th>\n",
       "      <td>SERPING1</td>\n",
       "      <td>Serpin family G member 1</td>\n",
       "      <td>splice</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000062.3</td>\n",
       "      <td>1250-13G&gt;A</td>\n",
       "      <td>1250minus13GtoA</td>\n",
       "      <td>11</td>\n",
       "      <td>57614315.0</td>\n",
       "      <td>57614315.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AD</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510773</th>\n",
       "      <td>PYGM</td>\n",
       "      <td>Glycogen phosphorylase, muscle associated</td>\n",
       "      <td>splice</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_005609.4</td>\n",
       "      <td>529-8G&gt;A</td>\n",
       "      <td>529minus8GtoA</td>\n",
       "      <td>11</td>\n",
       "      <td>64757918.0</td>\n",
       "      <td>64757918.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AR</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510783</th>\n",
       "      <td>MYBPC3</td>\n",
       "      <td>Myosin binding protein C, cardiac</td>\n",
       "      <td>splice</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000256.3</td>\n",
       "      <td>927-9G&gt;A</td>\n",
       "      <td>927minus9GtoA</td>\n",
       "      <td>11</td>\n",
       "      <td>47346379.0</td>\n",
       "      <td>47346379.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AD</td>\n",
       "      <td>0.000032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510786</th>\n",
       "      <td>APC</td>\n",
       "      <td>APC regulator of WNT signaling pathway</td>\n",
       "      <td>splice</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000038.6</td>\n",
       "      <td>1409-6A&gt;G</td>\n",
       "      <td>1409minus6AtoG</td>\n",
       "      <td>5</td>\n",
       "      <td>112827102.0</td>\n",
       "      <td>112827102.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AD</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510792</th>\n",
       "      <td>IDS</td>\n",
       "      <td>Iduronate 2-sulfatase</td>\n",
       "      <td>splice</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DM</td>\n",
       "      <td>NM_000202.8</td>\n",
       "      <td>880-8A&gt;G</td>\n",
       "      <td>880minus8AtoG</td>\n",
       "      <td>X</td>\n",
       "      <td>149490448.0</td>\n",
       "      <td>149490448.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XLR</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2768 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene                                   genename  mutype  \\\n",
       "462252       NF1                            Neurofibromin 1  splice   \n",
       "462258       NF1                            Neurofibromin 1  splice   \n",
       "462270     NTRK1    Neurotrophic receptor tyrosine kinase 1  splice   \n",
       "462274  SLC25A20         Solute carrier family 25 member 20  splice   \n",
       "462325      MAPT         Microtubule associated protein tau  splice   \n",
       "...          ...                                        ...     ...   \n",
       "510767  SERPING1                   Serpin family G member 1  splice   \n",
       "510773      PYGM  Glycogen phosphorylase, muscle associated  splice   \n",
       "510783    MYBPC3          Myosin binding protein C, cardiac  splice   \n",
       "510786       APC     APC regulator of WNT signaling pathway  splice   \n",
       "510792       IDS                      Iduronate 2-sulfatase  splice   \n",
       "\n",
       "                      clinvar_clnsig tag          refseq        hgvs  \\\n",
       "462252  Pathogenic/Likely_pathogenic  DM     NM_000267.3  2410-16A>G   \n",
       "462258                           NaN  DM     NM_000267.3   1393-9T>A   \n",
       "462270                    Pathogenic  DM  NM_001012331.2   851-33T>A   \n",
       "462274                    Pathogenic  DM     NM_000387.6   199-10T>G   \n",
       "462325                  not_provided  DM     NM_005910.6   915+12C>T   \n",
       "...                              ...  ..             ...         ...   \n",
       "510767                           NaN  DM     NM_000062.3  1250-13G>A   \n",
       "510773                           NaN  DM     NM_005609.4    529-8G>A   \n",
       "510783                    Pathogenic  DM     NM_000256.3    927-9G>A   \n",
       "510786             Likely_pathogenic  DM     NM_000038.6   1409-6A>G   \n",
       "510792                           NaN  DM     NM_000202.8    880-8A>G   \n",
       "\n",
       "                hgvsAll chromosome   startCoord     endCoord amino deletion  \\\n",
       "462252  2410minus16AtoG         17   31229009.0   31229009.0   NaN      NaN   \n",
       "462258   1393minus9TtoA         17   31214442.0   31214442.0   NaN      NaN   \n",
       "462270   851minus33TtoA          1  156873600.0  156873600.0   NaN      NaN   \n",
       "462274   199minus10TtoG          3   48884134.0   48884134.0   NaN      NaN   \n",
       "462325    915plus12CtoT         17   46010414.0   46010414.0   NaN      NaN   \n",
       "...                 ...        ...          ...          ...   ...      ...   \n",
       "510767  1250minus13GtoA         11   57614315.0   57614315.0   NaN      NaN   \n",
       "510773    529minus8GtoA         11   64757918.0   64757918.0   NaN      NaN   \n",
       "510783    927minus9GtoA         11   47346379.0   47346379.0   NaN      NaN   \n",
       "510786   1409minus6AtoG          5  112827102.0  112827102.0   NaN      NaN   \n",
       "510792    880minus8AtoG          X  149490448.0  149490448.0   NaN      NaN   \n",
       "\n",
       "       insertion expected_inheritance  gnomad_AF  \n",
       "462252       NaN                   AD   0.000000  \n",
       "462258       NaN                   AD   0.000000  \n",
       "462270       NaN                   AR   0.000000  \n",
       "462274       NaN                   AR   0.000065  \n",
       "462325       NaN                   AD   0.000000  \n",
       "...          ...                  ...        ...  \n",
       "510767       NaN                   AD   0.000000  \n",
       "510773       NaN                   AR   0.000000  \n",
       "510783       NaN                   AD   0.000032  \n",
       "510786       NaN                   AD   0.000000  \n",
       "510792       NaN                  XLR   0.000000  \n",
       "\n",
       "[2768 rows x 16 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allmut_dm[allmut_dm.mutype == 'splice']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>genename</th>\n",
       "      <th>mutype</th>\n",
       "      <th>clinvarID</th>\n",
       "      <th>clinvar_clnsig</th>\n",
       "      <th>tag</th>\n",
       "      <th>refseq</th>\n",
       "      <th>hgvs</th>\n",
       "      <th>hgvsAll</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>startCoord</th>\n",
       "      <th>endCoord</th>\n",
       "      <th>amino</th>\n",
       "      <th>deletion</th>\n",
       "      <th>insertion</th>\n",
       "      <th>expected_inheritance</th>\n",
       "      <th>gnomad_AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene, chrom, genename, mutype, clinvarID, clinvar_clnsig, tag, refseq, hgvs, hgvsAll, chromosome, startCoord, endCoord, amino, deletion, insertion, expected_inheritance, gnomad_AF]\n",
       "Index: []"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def remove_dot_ver(x):\n",
    "    if x == '.':\n",
    "        pass\n",
    "    else:\n",
    "        return re.match(r'[a-zA-Z_]+\\d+', x).group()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "510804"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(allmut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _liftover_to_hg19(chrom, pos):\n",
    "    converter = get_lifter('hg38', 'hg19')\n",
    "    result = converter.query(chrom, pos)\n",
    "    if result:\n",
    "        return result[0]\n",
    "    else:\n",
    "        return None\n",
    "    \n",
    "def anno_vars_id(row):\n",
    "    variant_id = f'{row[\"CHROM\"]}:{row[\"POS_hg19\"]}-{row[\"REF\"]}-{row[\"ALT\"]}'\n",
    "    return variant_id\n",
    "\n",
    "def anno_hg19_pos(row):\n",
    "    converted = _liftover_to_hg19(row['CHROM'], row['POS_hg38'])\n",
    "    return converted[1]\n",
    "\n",
    "def remove_dot_ver(x):\n",
    "    if x == '.':\n",
    "        pass\n",
    "    else:\n",
    "        return re.match(r'[a-zA-Z_]+\\d+', x).group()\n",
    "    \n",
    "def remove_non_canon(x):\n",
    "    if x in enst_set:\n",
    "        return True\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "def create_refalt(row, nt):\n",
    "    if row['Strand'] == '+':\n",
    "        return row[nt]\n",
    "    elif row['Strand'] == '-':\n",
    "        return str(Seq(row[nt]).complement())\n",
    "    else:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "canonlist = '/Github/Projects/DeNovo/data/CanonicalTranscripts_COMP/CanonicalTranscripts.exoncount.tsv'\n",
    "refseq = '/work/CanonicalTrasncripts/gencode.v43lift37.metadata.RefSeq.gz'\n",
    "\n",
    "df_canon = pd.read_table(canonlist, header=0)\n",
    "df_canon = df_canon[(df_canon['ENST'] != 'ENST00000649912')\n",
    "                    & (df_canon['ENST'] != 'ENST00000609375')]\n",
    "df_enst = df_canon.drop_duplicates(subset='ENST')\n",
    "enst_set = set(df_enst['ENST'])\n",
    "df_refseq = pd.read_table(refseq, header=None, \n",
    "                          names=['ENST_refseq', 'RefSeq_RNA', 'RefSeq_Pro'])\n",
    "df_refseq.fillna(value='.', inplace=True)\n",
    "\n",
    "df_refseq['ENST_refseq'] = df_refseq['ENST_refseq'].apply(remove_dot_ver)\n",
    "df_refseq['RefSeq_RNA'] = df_refseq['RefSeq_RNA'].apply(remove_dot_ver)\n",
    "df_refseq['RefSeq_Pro'] = df_refseq['RefSeq_Pro'].apply(remove_dot_ver)\n",
    "\n",
    "df_refseq['is_Canonical'] =  df_refseq['ENST_refseq'].apply(remove_non_canon)\n",
    "df_refseq_canon = df_refseq[df_refseq['is_Canonical'] == True].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Generate VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading HGMD splicing variants list (tsv)\n",
    "hgmd_file = '/Github/MyProjects/DeNovo/data/ValidationData/Positive/allmut.trim.colfixed.maf0.tsv.txt'\n",
    "df = pd.read_table(hgmd_file, header=0)\n",
    "\n",
    "# Pre-processing\n",
    "df = df.dropna(subset=['startCoord'])\n",
    "df['startCoord'] = df['startCoord'].astype(int)\n",
    "df['refseq'] = df['refseq'].apply(remove_dot_ver)\n",
    "\n",
    "# Extract REF and ALT from HGVS descriptions\n",
    "sr_alt_nt = df['hgvs'].str[-1:].rename('alt_nt')\n",
    "sr_ref_nt = df['hgvs'].str[-3:-2].rename('ref_nt')\n",
    "df = pd.concat([df, sr_ref_nt, sr_alt_nt], axis=1)\n",
    "\n",
    "# Insert cols as VCF\n",
    "df.loc[:,'ID'] = '.'\n",
    "df.loc[:,'QUAL'] = '.'\n",
    "df.loc[:,'FILTER'] = '.'\n",
    "df.loc[:,'INFO'] = '.'\n",
    "\n",
    "# Annotate ENST IDs\n",
    "df = pd.merge(df, df_enst, how='left', \n",
    "              left_on='gene', right_on='GeneSymbol')\n",
    "\n",
    "# Rename cols for downstream processing\n",
    "df = df.rename(columns={'CHROM': 'Chr',\n",
    "                        'chromosome': 'CHROM', \n",
    "                        'startCoord': 'POS_hg38'})\n",
    "\n",
    "# Create REF and ALT columns\n",
    "df['REF'] = df.apply(create_refalt, nt='ref_nt', axis=1)\n",
    "df['ALT'] = df.apply(create_refalt, nt='alt_nt', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Liftover to hg19\n",
    "df['POS_hg19'] = df.apply(anno_hg19_pos, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = df.copy()\n",
    "df2 = df2.astype({'POS_hg19': str})\n",
    "\n",
    "# Annotate variant IDs\n",
    "df2['variant_id'] = df2.apply(anno_vars_id, axis=1)\n",
    "\n",
    "# Extract columns for VCF\n",
    "df_19 = df2[['CHROM', 'POS_hg19', \n",
    "             'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO']]\n",
    "df_38 = df2[['CHROM', 'POS_hg38', \n",
    "             'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_19 = df_19.dropna(subset=['REF'])\n",
    "df_38 = df_38.dropna(subset=['REF'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output as VCF\n",
    "df_19.to_csv('./patho.hg19.vcf', sep='\\t', index=False, header=False)\n",
    "df_38.to_csv('./patho.hg38.vcf', sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Concatenate header and variant list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "metadata": {},
   "outputs": [],
   "source": [
    "!cat ../header_for_VCF.tsv ./patho.hg19.vcf > patho.hg19.header.vcf\n",
    "!cat ../header_for_VCF.tsv ./patho.hg38.vcf > patho.hg38.header.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Strand</th>\n",
       "      <th>ref_nt</th>\n",
       "      <th>REF</th>\n",
       "      <th>alt_nt</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>+</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10553</th>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>None</td>\n",
       "      <td>A</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10554</th>\n",
       "      <td>+</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10555</th>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10556</th>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10557</th>\n",
       "      <td>+</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10558 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Strand ref_nt   REF alt_nt   ALT\n",
       "0          +      A     A      G     G\n",
       "1          -      T     A      G     C\n",
       "2          -      G     C      A     T\n",
       "3          +      A     A      G     G\n",
       "4          -      A     T      G     C\n",
       "...      ...    ...   ...    ...   ...\n",
       "10553    NaN      G  None      A  None\n",
       "10554      +      G     G      A     A\n",
       "10555      -      G     C      A     T\n",
       "10556      -      G     C      A     T\n",
       "10557      +      G     G      A     A\n",
       "\n",
       "[10558 rows x 5 columns]"
      ]
     },
     "execution_count": 263,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['Strand', 'ref_nt', 'REF', 'alt_nt', 'ALT']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 =df2[['gene', 'gdbid', 'omimid', 'amino', 'codon', 'codonAff', 'descr',\n",
    "       'refseq', 'hgvs', 'CHROM', 'POS_hg38', 'endCoord', \n",
    "       'expected_inheritance', 'dmsupport', 'mutype', 'acc_num', 'new_date', \n",
    "       'clinvarID', 'clinvar_clnsig', 'Chr', 'Start', 'End', 'Strand', \n",
    "       'GeneSymbol', 'HGNC_ID', 'ENSG', 'ENST', 'GeneType', 'Tag', \n",
    "       'REF', 'ALT', 'POS_hg19', 'variant_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2.to_csv('./patho2.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "wesanno",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
